---
title: "Figure 5 and supplemental 5"
author: "Ruben Schep"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    theme: journal #cerulean
    highlight: monochrome
    toc: true
    toc_float: true
    code_folding: show
  editor_options:
    chunk_output_type: console
---

# knitr document van Steensel lab
```{r}
StartTime <-Sys.time()
library(knitr)

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8)

## Select outdir
out.dir = paste0("figures/rs", Date, "/")
dir.create(out.dir)
opts_chunk$set(dev=c('png', 'pdf'), fig.path = file.path(out.dir))
pdf.options(useDingbats = FALSE)
```

# Introduction

## Description of Data

# Data processing
## Path, Libraries, Parameters and Useful Functions
```{r setup, message=FALSE, warnings=FALSE}
# INPUT DIRECTORY
in.dir.date = 20200124
in.dir = paste0("/DATA/projects/DSBrepair/data/R/rs", in.dir.date, "/")

# libraries:
library(ggplot2)
library(ggpubr)
# library(reshape2)
# library(tibble)
# library(GenomicRanges)
# library(rtracklayer)
library(corrr)
library(Hmisc)
library(ggbeeswarm)
library(RColorBrewer)
library(data.table)
library(dplyr)
# library(plyr)
library(tidyr)
# library(stringr)
# library(plotly)
library(ggpmisc)
library(glmnet)
library(cowplot)
library(mda)
library(earth)
library(yaml)
library(vip)
library(caret)
library(scales)
library(glmnet)
library(gridExtra)
library(ggcorrplot)
library(bnlearn)
library(pheatmap)
library(ppcor)
library(parallel)

color_redblue <- rev(brewer.pal(11,"RdBu"))
duocolor <- c("#EE756E", "#2BB6BE")
colore <- c("wt" = "#808184", "other" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "SSTR" = "#007A4C")
colores <- c("wt" = "#808184", "other_indels" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "SSTR" = "#007A4C")
colori  <- c("other_indels" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "ssODN" = "#007A4C")
colora <- c("0" = "#808184", "1" = "#E1251B", "-7" = "#223DA0", "-14" = "#223DA0")
coloru <- c("wt" = "#808184", "other" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "MMEJplusNHEJ" = "#EE8A2A", "all" = "black", "MMEJplusNHEJplusHDR" = "#EE8A2A", "HDR" = "#007A4C", "MMEJplusHDR" = "#EE8A2A")
colory <- c("wt" = "#808184", "other" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "MMEJplusNHEJ" = "#EE8A2A", "all" = "black", "SSTR" = "#007A4C")

measure_color = c(NHEJ='#e1251b', MMEJ='#26419a', SSTR='#007A4C', None='grey')

method_color = c(MARS='#00bfc4', LASSO='#f8766d')
lad_color <- c(LAD="#802cdc", iLAD="#dfaf2f")
sign_color <- c(down="#AA6375", up="#7B9A36")

feature_fill = c(insulator="#a6b5a3", repressive="#304441", euchromatin="#cda4cc",
                 transcribing="#ED9114", HDAC="#aac9e0", accessibility='#cebc85',
                 methylation='#7dc98f')

feature_color = c(insulator="black", repressive="white", euchromatin="black",
                  transcribing="black", HDAC="black", accessibility='black',
                  methylation='black')

clustering = data.frame(row.names=c('LMNB1', 'late_replicating', 'H3K9me2',
                                    'H3K9me3', 'H3K27me3', 'EZH2', 'CTCF', 'SMC3',
							        'HDAC3', 'HDAC2', 'HDAC1', 'H3K4me1',
								    'H3K4me2', 'H3K4me3', 'H3K27ac',
								    'H4K5acK8ac', 'H2AFZ', 'DNAse', 'Dam', 'm5C',
								    'H3K79me2', 'TTseq', 'H3K36me3', 'POL2AS2',
								    'POL2'),
					   group=factor(c(rep('repressive', 6), rep('insulator', 2),
				                      rep('HDAC', 3), rep('euchromatin', 6),
                                      rep('accessibility', 2), 'methylation',
						              rep('transcribing', 5)),
									levels=c('transcribing', 'accessibility',
                                             'methylation', 'euchromatin',
								             'HDAC', 'insulator', 'repressive')))

```

### Custom functions
Functions used thoughout this script.
```{r, message=FALSE, warnings=FALSE}
CorrelationDFMean <- function(data, condition_name, feature_name, targets, p.value=0.001){
  hist.corr.list <- data[data$exp_pool==condition_name, ] %>% select(feature_name, targets) %>% as.matrix() %>% rcorr(type = "pearson")

  histone.corr.df <- data.frame(rowname = names(hist.corr.list$r[1, -1]),
                                corr = hist.corr.list$r[1, -1],
                                pvalue = hist.corr.list$P[1, -1])

  return(histone.corr.df=histone.corr.df)
}


CorrelationPlotMean <- function(data, condition_name, feature_name, targets, p.value=0.001){
  cor.df = data.frame()
  for (i in feature_name){
    corr.list <- data %>% filter(exp_pool==condition_name) %>% select(i, as.character(targets)) %>% as.matrix() %>% rcorr(type = "pearson")

    df <- data.frame(rowname = names(corr.list$r[1, -1]),
                                corr = corr.list$r[1, -1],
                                pvalue = corr.list$P[1, -1],
                                feature = i)
    cor.df <- rbind(cor.df, df)
  }
  p <- cor.df %>%
    transform(rowname = factor(rowname, levels = levels(targets)),
              corr = corr,
              pvalue = pvalue) %>%
    ggplot(aes(x = rowname, y = corr, fill = feature, alpha = pvalue < p.value)) +
    geom_bar(stat = "identity", position = "dodge") +
    ylab(paste("Pearson correlation with", feature_name)) +
    xlab("Chromatin Feature") +
    ggtitle(paste("Repair pathways correlation with chromatin features in", condition_name, "exps")) +
    theme_bw(base_size = 16) +
    scale_fill_manual("legend", values = colores) +
    scale_y_continuous(breaks=seq(-1,1, .1)) +
    coord_flip()
  return(list(dataframe=cor.df, plot = p))
}


cor_test <- function(y, X, left, right){
    cor_list = lapply(X, function(x){
        cor.test(x,y)
    })
    p_vec = unlist(lapply(cor_list, function(x){x$p.value}))
    cor_vec = unlist(lapply(cor_list, function(x){x$estimate}))
    p_adj = p.adjust(p_vec)
    color = ifelse(p_adj < 0.05, ifelse(cor_vec > 0, left, right),
                   'None')
    importance = abs(cor_vec)
    target=factor(colnames(X), levels=rownames(clustering))
    feature_group = clustering[colnames(X), ]
    return(list(correlation=cor_vec,importance = importance, p=p_vec,
                p_adj=p_adj, target=target, color=color,
                feature_group=feature_group))
}


get_combinations <- function(domain_vec){
    n_domain = length(domain_vec)
    get_ivec <-function (last_x, n_domain){
    	if (last_x < n_domain){
    		i_vec = c(0, c((last_x+1):n_domain))
    	} else{
    		i_vec = c(0)
    	}
    }

    combine <- function(x_new, x_vec, n_domain){
        new_vec = c(x_vec, x_new)
        if (length(new_vec)==n_domain){
            return(new_vec)
        } else {
            i_vec = get_ivec(max(new_vec), n_domain)
            x_list = lapply(i_vec, combine, x_vec=new_vec,
                            n_domain=n_domain)
            return(do.call(rbind, x_list))
        }
    }

    comb_list = lapply(c(0:n_domain), combine, x_vec=c(), n_domain=n_domain)

    comb = do.call(rbind, comb_list)
    comb_df = unique(t(apply(comb, 1, function(x){x[order(x)]})))

    name_list = apply(comb_df, 1, function(x){
    	name = paste(domain_vec[x], collapse='-')
    	if (name == ''){
    		name = 'iDomain'
    	}
    	return(name)
    })

    opt_list = lapply(domain_vec, function(d){
    	 apply(comb_df, 1, function(x){
    		opt = domain_vec[x]
    		return(d%in%opt)
    	})
    })

    opt_df = do.call(cbind, opt_list)
    colnames(opt_df) = domain_vec
    rownames(opt_df) = name_list

    return(opt_df)
}

plot_comb_grid <- function(indel.dt, opt_df, domains, y_name, min_count=0,
                           max_count=Inf, group_by=NULL, color=NULL){

    opt_melt = reshape2::melt(opt_df)
    colnames(opt_melt) = c('group', 'domain', 'is_in')

    opt_melt$is_in = ifelse(opt_melt$is_in, 'in', 'out')
    opt_melt$domain = factor(opt_melt$domain, levels=domains)


    group_list = lapply(rownames(opt_df), function(opt_name){
    	opt = opt_df[opt_name, ]
    	in_vec = domains[opt]
    	out_vec = domains[!opt]

    	in_sum = rowSums(indel.dt[,..in_vec])
    	out_sum = rowSums(indel.dt[,..out_vec])
    	if (length(in_vec)==0){
    		in_group = which(out_sum == 0)
    	} else if (length(out_vec) == 0){
    		in_group = which(in_sum==length(in_vec))
    	} else  {
    		in_group = which(in_sum == length(in_vec) & out_sum==0)
    	}
    	return(in_group)
    })

    names(group_list) = rownames(opt_df)
    for (opt_name in names(group_list)){
    	i_vec = group_list[[opt_name]]
    	indel.dt[i_vec, group := opt_name]
    }
    if (is.null(group_by)){
        group_count_df = data.frame(indel.dt[,table(group)])
    } else {
        group_count_df = data.frame(indel.dt[,table(group)/length(unique(get(group_by)))])
    }
    group_count_df = group_count_df[order(match(group_count_df$group, rownames(opt_df))),]
    # print(group_count_df)
    group_count_sel = group_count_df[group_count_df$Freq>=min_count,]
    group_count_sel = group_count_sel[group_count_sel$Freq<=max_count,]
    group_levels =  group_count_sel$group
    opt_sel = opt_melt[opt_melt$group %in% group_levels, ]

    opt_sel$group = factor(opt_sel$group, levels=group_levels)
    # print(opt_sel)

    indel_selection = indel.dt[group%in%group_levels, ]
    indel_selection[,group:=factor(group, levels=group_levels)]

    # wilcox_list = lapply(group_levels[group_levels!='iDomain'],
    #                      function(g){
    #                          x = indel_selection[group==g, y_name, with=F]
    #                          y = indel_selection[group=='iDomain', y_name,
    #                                              with=F]
    #                          wilcox.test(unlist(x),unlist(y))
    #                      })

    group_count_sel$group = factor(group_count_sel$group, levels=group_levels)

    count_plot = ggplot(group_count_sel, aes(x=group,y=1, label=Freq)) +
                            geom_label() +
                            theme_void()



    dots = ggplot(opt_sel, aes(x=group, y=domain, color=is_in)) +
    	geom_point(size=2) +
    	geom_line(data=opt_sel[opt_sel$is_in=='in',], aes(group=group)) +
    	theme_bw() +
    	scale_color_manual(values=c('in'='black', 'out'='grey70')) +
    	guides(color=FALSE)  +
    	geom_vline(xintercept=seq(1.5, length(unique(indel.dt$group))-0.5, 1),
    			   lwd=0.5, colour="grey90") +
    	theme(axis.title=element_blank(),
    		  axis.text=element_blank(),
    		  axis.ticks=element_blank(),
    		  panel.grid.major = element_blank())

    symnum = list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1),
                  symbols = c("****", "***", "**", "*", ""))


    if (is.null(group_by)){
        beeswarm = ggplot(indel_selection, aes_string(x='group', y=y_name)) +
        			   geom_quasirandom(width=0.3) +
                       stat_summary(fun.y=median, fun.ymin = median,
                                    fun.ymax = median,
                                    geom="crossbar", width = 0.5,
                                    color='red') +
                       stat_compare_means(aes(label = ..p.signif..),
                                                 symnum.args = symnum,
                                                 ref.group='iDomain')
    } else {
        form = formula(paste(y_name, '~', group_by))
        comp = compare_means(form, group.by = "group", data=indel_selection)

        beeswarm = ggplot(indel_selection,
                          aes_string(x='group', y=y_name, color=group_by)) +
        			   geom_quasirandom(dodge.width=1) +
                       scale_color_manual(values=color) +
                       stat_summary(aes_string(group=group_by), fun.y=median,
                                    fun.ymin = median, fun.ymax = median,
                                    geom="crossbar", width = 0.5,
                                    color='red',
                                    position= position_dodge(width =1)) +
                       geom_text(data=comp, aes(x=group, y=0.8, label=p.signif),
                                 inherit.aes=F)
    }
    beeswarm = beeswarm +
    			  theme_bw() +
    			  geom_vline(xintercept=seq(1.5, length(unique(indel.dt$group))-0.5, 1),
    			             lwd=0.5, colour="grey90") +
    			  theme(axis.title=element_blank(),
    			        axis.text.x=element_blank(),
      				    panel.grid.major.x = element_blank(),
    				    axis.ticks.x=element_blank())


    count = data.frame(domain=gsub('_domain', '', domains),
    	               count=colSums(indel.dt[,..domains], na.rm=T))

    count$domain = factor(count$domain, levels=domain_levels)


    histogram = ggplot(count, aes(x=1, y=domain, label = count)) +
    				geom_label() +
    				theme_bw() +
    				theme(plot.background = element_blank(),
        				  panel.grid.major = element_blank(),
        				  panel.grid.minor = element_blank(),
        				  panel.border = element_blank(),
    					  axis.title=element_blank(),
    	  			      axis.text.x=element_blank(),
    	  				  axis.ticks=element_blank())

    empty = ggplot(data=data.frame(1,1),aes(x=1,y=1)) +
                geom_point() +
                theme_void()

    count_beeswarm = plot_grid(beeswarm, count_plot, ncol=1, rel_heights=c(20,1),
                               align='v', axis='lr')
    return(plot_grid(empty, beeswarm, empty, count_plot,
                     histogram, dots, nrow=3,
                     rel_widths=c(1,10), rel_heights=c(15,1,5), align='vh',
                     axis='tlbr'))
}

```

# Data import
```{r, message=FALSE, warnings=FALSE}
load(paste0(in.dir, "RSTP2_Indel_Chromatin_2kb.RData"))
load(paste0(in.dir, "RSTP2_IndelRatios_Chromatin_2kb.RData"))
```


## Data loading

```{r preprocess experiments}
trip_tib_mut <- copy(trip_tib_2000)
trip_tib_2000 <- trip_tib_2000 %>% distinct(barcode, exp, .keep_all = TRUE)

trip_tib <- trip_tib_2000 %>%
  filter(exp_pool  %in% c("mean_RSTP2_2000_LBR2_64", 
               "mean_RSTP2_2000_LBR2_BIX01294_64",
               "mean_RSTP2_2000_LBR2_DMSO_64", 
               "mean_RSTP2_2000_LBR2_GSK126_64"))

trip_tib_mut <- trip_tib_mut_2000 %>%
  filter(exp_pool %in% c("mean_RSTP2_2000_LBR2_64", 
               "mean_RSTP2_2000_LBR2_BIX01294_64",
               "mean_RSTP2_2000_LBR2_DMSO_64", 
               "mean_RSTP2_2000_LBR2_GSK126_64"))

mean_trip_tib_2000 <- trip_tib_2000 %>%
  gather(NHEJ, MMEJ, other_indels, key = "pathway", value = 'ratios') %>%
  distinct(barcode, exp, pathway, .keep_all = TRUE)

mean_trip_tib_2000$pathway <- factor(mean_trip_tib_2000$pathway, levels=c("other_indels", "MMEJ", "NHEJ"))
mean_trip_tib_2000 <- mean_trip_tib_2000[!is.na(mean_trip_tib_2000$ratios), ]
```
# Figures
## Main Figures
### Panel A

Panel A is the scheme of the ssODN. That has been made.
Add loading here I guess.

### Panel B

### Panel A
```{r pathways vs contribution ssODN, fig.height=8, fig.width=9}
trip_tib_pathways <- trip_tib %>%
  gather(NHEJ, MMEJ, other_indels, key = "pathway", value = 'ratios') %>%
  distinct(barcode, exp, pathway, .keep_all = TRUE)

exp_order <- c("mean_RSTP2_2000_LBR2_64", 
               "mean_RSTP2_2000_LBR2_BIX01294_64",
               "mean_RSTP2_2000_LBR2_DMSO_64", 
               "mean_RSTP2_2000_LBR2_GSK126_64")

trip_tib_pathways$pathway <- factor(trip_tib_pathways$pathway, levels=c("other_indels", "MMEJ", "NHEJ"))
trip_tib_pathways$exp_pool <- factor(trip_tib_pathways$exp_pool, levels=exp_order)
trip_tib_pathways <- trip_tib_pathways[!is.na(trip_tib_pathways$ratios), ]

pathway_contribution <- trip_tib_pathways %>% group_by(exp_pool, pathway) %>% dplyr::summarise(pathway_contr = mean(ratios))

colores <- c("other_indels" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0")

p <- ggplot(pathway_contribution, aes(exp_pool, pathway_contr, fill = pathway, order = pathway)) +
  geom_bar(stat = "identity", position = "fill")  +
  theme_bw(base_size = 16)
p + scale_fill_manual("legend", values = colores) +
  scale_x_discrete(labels=c("indel_mean_LBR2_DMSO_64h" = "DMSO",
                            "indel_mean_LBR2_GSK_64h" = "GSK126",
                            "indel_mean_LBR2_64h" = "-",
                            "indel_mean_LBR2_BIX_64h" = "BIX01294")) +
  labs(x = element_blank(),
       y = "abundance")


```

### Panel C2

```{r ratio DRUG/CTRL per het domain, fig.height=7, fig.width=12}
domain_levels = c('late_replicating', 'LAD', 'H3K9me2', 'H3K9me3', 'H3K27me3')

domains = paste0(domain_levels, '_domain')

indel.dt = trip_tib %>%
  dplyr::select(barcode, drug, MMEJ_MMEJNHEJ, domains) %>%
  spread(drug, MMEJ_MMEJNHEJ) %>%
  filter(!is.na(BIX01294), !is.na(GSK126)) %>%
  mutate(BIX_ratio = BIX01294/`-`,
         GSK_ratio = GSK126/DMSO) %>%
  distinct(barcode, .keep_all = TRUE) %>%
  data.table()


# indel.dt = unique(tib.dt[exp_pool == "indel_mean_LBR2_ssODN_DMSO_64h",  
#                          c('exp_pool', 'barcode', 'SSTR', 'SSTR_MMEJSSTR',
#                            domains), with=F])

indel.dt[indel.dt=='iDomain'] <- 0
indel.dt[indel.dt=='Domain'] <- 1

indel.dt[is.na(LAD_domain), LAD_domain:=0]

indel.dt[, (domains) := lapply(.SD, as.numeric), .SDcols = domains]

colnames(indel.dt) = gsub('_domain', '', colnames(indel.dt))

opt_df = get_combinations(domain_levels)


# Ratio of SSTR vs all indels
plot_comb_grid(indel.dt, opt_df,
                   domain_levels, 'log2(BIX_ratio)', min_count = 20)

# Ratio of SSTR vs MMEJ
plot_comb_grid(indel.dt, opt_df,
                   domain_levels, 'log2(GSK_ratio)', min_count = 20)

### CHRIST ####
# Can you now plot the ratio MMEJ_MMEJNHEJ of these domains that are significant vs the control condition to see the fold change and significance?

bix.dt = melt(indel.dt, measure.vars=c('-', 'BIX01294'),
              variable.name='condition', value.name='MMEJ_MMEJNHEJ')

color_vec = c('-'='grey50', 'BIX01294'='black')

plot_comb_grid(bix.dt, opt_df, domain_levels, 'MMEJ_MMEJNHEJ',
               min_count = 20, group_by='condition', color=color_vec)


gsk.dt = melt(indel.dt, measure.vars=c('DMSO', 'GSK126'),
              variable.name='condition', value.name='MMEJ_MMEJNHEJ')

color_vec = c('DMSO'='grey50', 'GSK126'='black')

plot_comb_grid(gsk.dt, opt_df, domain_levels, 'MMEJ_MMEJNHEJ',
               min_count = 20, group_by='condition', color=color_vec)
```

### Panel D
```{r scatters MMEJ ssODN vs MMEJ -, fig.height=6, fig.width=10}

```
####  Figure legend

### Panel E
```{r LMNA KO, fig.height=6, fig.width=10}
domain_levels = c('late_replicating', 'LAD', 'H3K9me2', 'H3K9me3', 'H3K27me3')

domains = paste0(domain_levels, '_domain')
trip_tib_mut = trip_tib_mut_2000 %>%
  filter(exp_pool %in% c("mean_RSTP2_clone5_LBR2_DMSO_siNT_64", 
                         "indel_LMNAKO_LBRKORep1_clone5_LBR2_LMNA_KO_1_72",
                         "indel_LMNAKO_LBRKORep1_clone5_LBR2_LMNA_KO_2_72",
                         "indel_LMNAKO_LBRKORep1_clone5_LBR2_LBR_KO_1_72",
                         "indel_LMNAKO_LBRKORep1_clone5_LBR2_LBR_KO_2_72"))

trip_tib = trip_tib_2000 %>%
  filter(exp_pool %in% c("mean_RSTP2_clone5_LBR2_DMSO_siNT_64", 
                         "indel_LMNAKO_LBRKORep1_clone5_LBR2_LMNA_KO_1_72",
                         "indel_LMNAKO_LBRKORep1_clone5_LBR2_LMNA_KO_2_72",
                         "indel_LMNAKO_LBRKORep1_clone5_LBR2_LBR_KO_1_72",
                         "indel_LMNAKO_LBRKORep1_clone5_LBR2_LBR_KO_2_72")) %>%
  distinct(barcode, exp_pool, .keep_all = TRUE)

ggplot(filter(trip_tib_mut, mutation != "ssODN"),
       aes(mutation, freq, color = color)) +
  geom_quasirandom() +
  theme_bw(base_size = 16) +
  ylim(0, 1) +
  theme(legend.position = c(0.1, 0.8)) +
  scale_color_manual("legend", values = colore) +
  facet_wrap( ~ drug)

indel.dt = trip_tib %>%
  dplyr::select(barcode, drug, MMEJ_MMEJNHEJ, domains) %>%
  spread(drug, MMEJ_MMEJNHEJ) %>%
  filter(!is.na(LMNA_KO_1), !is.na(LMNA_KO_2)) %>%
  mutate(LMNA1 = LMNA_KO_1-DMSO,
         LMNA2 = LMNA_KO_2-DMSO,
         LBR1 = LBR_KO_1-DMSO,
           LBR2 = LBR_KO_2-DMSO) %>%
  distinct(barcode, .keep_all = TRUE) %>%
  data.table()

indel.dt[indel.dt=='iDomain'] <- 0
indel.dt[indel.dt=='Domain'] <- 1

indel.dt[is.na(LAD_domain), LAD_domain:=0]

indel.dt[, (domains) := lapply(.SD, as.numeric), .SDcols = domains]

colnames(indel.dt) = gsub('_domain', '', colnames(indel.dt))

opt_df = get_combinations(domain_levels)


# Ratio of MMEJ_MMEJNHEJ vs all indels
plot_comb_grid(indel.dt, opt_df,
                   domain_levels, 'log2(LMNA1)')
plot_comb_grid(indel.dt, opt_df,
                   domain_levels, 'log2(LMNA2)')
plot_comb_grid(indel.dt, opt_df,
                   domain_levels, 'log2(LBR1)')
plot_comb_grid(indel.dt, opt_df,
                   domain_levels, 'log2(LBR2)')

# Ratio of SSTR vs MMEJ
plot_comb_grid(indel.dt, opt_df,
                   domain_levels, 'log2(GSK_ratio)')

### CHRIST ####
# Can you now plot the ratio MMEJ_MMEJNHEJ of these domains that are significant vs the control condition to see the fold change and significance?

LMNA_KO1.dt = melt(indel.dt, measure.vars=c('DMSO', 'LMNA_KO_1'),
              variable.name='condition', value.name='MMEJ_MMEJNHEJ')

color_vec = c('DMSO'='grey50', 'LMNA_KO_1'='black')

plot_comb_grid(LMNA_KO1.dt, opt_df, domain_levels, 'MMEJ_MMEJNHEJ', group_by='condition', color=color_vec)


LMNA_KO2.dt = melt(indel.dt, measure.vars=c('DMSO', 'LMNA_KO_2'),
              variable.name='condition', value.name='MMEJ_MMEJNHEJ')

color_vec = c('DMSO'='grey50', 'LMNA_KO_2'='black')

plot_comb_grid(LMNA_KO2.dt, opt_df, domain_levels, 'MMEJ_MMEJNHEJ', group_by='condition', color=color_vec)



LBR_KO_1.dt = melt(indel.dt, measure.vars=c('DMSO', 'LBR_KO_1'),
              variable.name='condition', value.name='MMEJ_MMEJNHEJ')

color_vec = c('DMSO'='grey50', 'LBR_KO_1'='black')

plot_comb_grid(LBR_KO_1.dt, opt_df, domain_levels, 'MMEJ_MMEJNHEJ', group_by='condition', color=color_vec)

LBR_KO_2.dt = melt(indel.dt, measure.vars=c('DMSO', 'LBR_KO_2'),
              variable.name='condition', value.name='MMEJ_MMEJNHEJ')

color_vec = c('DMSO'='grey50', 'LBR_KO_2'='black')

plot_comb_grid(LBR_KO_2.dt, opt_df, domain_levels, 'MMEJ_MMEJNHEJ', group_by='condition', color=color_vec)
```
#### Figure legend

### Panel F
```{r pathways vs chrom ssODN DMSO, fig.height=8, fig.width=9}
trip_tib %>%
  dplyr::select(barcode, drug, MMEJ_MMEJNHEJ, LMNB1) %>%
  spread(drug, MMEJ_MMEJNHEJ) %>%
  filter(!is.na(LMNA_KO_1), !is.na(LMNA_KO_2)) %>%
  mutate(LMNA1 = LMNA_KO_1-DMSO,
         LMNA2 = LMNA_KO_2-DMSO,
         LBR1 = LBR_KO_1-DMSO,
           LBR2 = LBR_KO_2-DMSO) %>%
  gather(LMNA1, LMNA2, LBR1, LBR2, key = "KO", value = "change") %>% 
  distinct(barcode, KO, .keep_all = TRUE) %>% 
  ggplot(., aes(LMNB1, change)) + 
           geom_point() + 
  theme_bw(base_size = 16) + 
  geom_smooth(method = glm)  +
  stat_fit_glance(method = 'lm',
        method.args = list(formula = formula),
        aes(label = paste("P-value = ", signif(..p.value.., digits = 4), sep = "")),
        size = 5) +
  stat_cor(method = "pearson")  +
  facet_wrap(~ KO)

trip_tib %>%
  dplyr::select(barcode, drug, MMEJ_MMEJNHEJ, LAD_domain) %>%
  spread(drug, MMEJ_MMEJNHEJ) %>%
  filter(!is.na(LMNA_KO_1), !is.na(LMNA_KO_2)) %>%
  mutate(LMNA1 = LMNA_KO_1-DMSO,
         LMNA2 = LMNA_KO_2-DMSO,
         LBR1 = LBR_KO_1-DMSO,
         LBR2 = LBR_KO_2-DMSO) %>%
  gather(LMNA1, LMNA2, LBR1, LBR2, key = "KO", value = "change") %>% 
  distinct(barcode, KO, .keep_all = TRUE) %>% 
  ggplot(., aes(LAD_domain, change)) + 
           geom_quasirandom() + 
  theme_bw(base_size = 16) + 
  facet_wrap(~ KO)
```

## Supplementary Figures
### Panel xx
```{r seperate replicates, fig.height=12, fig.width=12}
colores <- c("wt" = "#808184", "other_indels" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "SSTR" = "#007A4C")

pathway_distribution <- trip_tib_2000 %>% 
  filter(cell_line == "RSTP2_clone5", plasmid == "LBR2", time %in% c(64,72), drug %in% c("DMSO", "-"), replicate == "mean") %>%
  gather(NHEJ, MMEJ, SSTR, other_indels, 
         key = "pathway", value = 'ratios') %>% 
  arrange(dplyr::desc(ratios))
barcode_order <- pathway_distribution %>% 
  filter(pathway == "NHEJ")  %>% 
  arrange(dplyr::desc(ratios)) %>% 
  dplyr::select(barcode) %>% unique()

pathway_distribution$barcode <- factor(pathway_distribution$barcode, levels=barcode_order$barcode)
pathway_distribution$pathway <- factor(pathway_distribution$pathway, levels=c("SSTR", "MMEJ", "other_indels",  "NHEJ"))

ggplot(pathway_distribution,
       aes(barcode, ratios, width=1)) +
  geom_bar(stat = "identity", aes(fill = pathway)) +
  theme_bw(base_size = 16) +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  ylab("Pathway Contribution") +
  scale_fill_manual("legend", values = colores) +
  ggtitle("Pathway contribution per barcode") +
  facet_wrap(~ exp_pool)


pathway_distribution <- trip_tib_2000 %>% 
  filter(cell_line == "RSTP2_clone5", plasmid == "LBR2", time %in% c(64,72), siRNA %in% c("siNT", "-")) %>%
  gather(NHEJ, MMEJ, SSTR, other_indels, 
         key = "pathway", value = 'ratios') %>% 
  arrange(dplyr::desc(ratios))
barcode_order <- pathway_distribution %>% 
  filter(pathway == "NHEJ")  %>% 
  arrange(dplyr::desc(ratios)) %>% 
  dplyr::select(barcode)  %>% unique()

pathway_distribution$barcode <- factor(pathway_distribution$barcode, levels=barcode_order$barcode)
pathway_distribution$pathway <- factor(pathway_distribution$pathway, levels=c("SSTR", "MMEJ", "other_indels",  "NHEJ"))

ggplot(pathway_distribution,
       aes(barcode, ratios, width=1)) +
  geom_bar(stat = "identity", aes(fill = pathway)) +
  theme_bw(base_size = 16) +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  ylab("Pathway Contribution") +
  scale_fill_manual("legend", values = colores) +
  ggtitle("Pathway contribution per barcode") +
  facet_wrap(~ drug)
```

### Panel H
```{r scatter NHEJ with and without ssODN, fig.height=6, fig.width=10}

```


# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```
