---
title: "Figure 3 and supplemental 3-4"
author: "Ruben Schep"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    theme: journal #cerulean
    highlight: monochrome
    toc: true
    toc_float: true
    code_folding: show
  editor_options:
    chunk_output_type: console
---

# knitr document van Steensel lab
```{r knitr opt}
StartTime <-Sys.time()
library(knitr)

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8)

## Select outdir
out.dir = paste0("figures/rs", Date, "/")
dir.create(out.dir)
opts_chunk$set(dev=c('png', 'pdf'), fig.path = file.path(out.dir))
pdf.options(useDingbats = FALSE)
```

# Introduction

## Description of Data

# Data processing
## Path, Libraries, Parameters
```{r setup, message=FALSE, warnings=FALSE}


# INPUT DIRECTORY
in.dir.date = 20201029
in.dir = paste0("/DATA/projects/DSBrepair/data/R/rs", in.dir.date, "/")


# libraries:
library(ggplot2)
library(ggpubr)
library(corrr)
library(Hmisc)
library(ggbeeswarm)
library(dplyr)
library(tidyr)
library(tibble)
library(ggpmisc)
library(cowplot)
library(data.table)
library(report)
# library(ggcorrplot)

colore <- c("wt" = "#808184", "other" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "SSTR" = "#007A4C")
colores <- c("wt" = "#808184", "other_indels" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "SSTR" = "#007A4C")
guide_colors <- c("LBR1" = "#c8e3b4", "LBR12" = "#7fccbb", "LBR15" = "#2c7fb9", "LBR2" = "#293990")

feature_fill = c(insulator="#a6b5a3", repressive="#304441", euchromatin="#cda4cc",
                 transcribing="#ED9114", HDAC="#aac9e0", accessibility='#cebc85',
                 methylation='#7dc98f')

feature_color = c(insulator="black", repressive="white", euchromatin="black",
                  transcribing="black", HDAC="black", accessibility='black',
                  methylation='black')

measure_color = c(None='grey', high='#A6C57B', low='#BD6F46')
```

## Custom functions
Functions used thoughout this script.
```{r custom functions, message=FALSE, warnings=FALSE}
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, substr(gsub("-","",Sys.time()),1,8), "_", filename)
  filename
}

CorrelationDFMean <- function(data, condition_name, feature_name, targets, p.value=0.001){
  hist.corr.list <- data[data$exp_pool==condition_name, ] %>%
    dplyr::select(feature_name, targets) %>%
    as.matrix() %>%
    rcorr(type = "pearson")
  
  histone.corr.df <- data.frame(rowname = names(hist.corr.list$r[1, -1]),
                                corr = hist.corr.list$r[1, -1],
                                pvalue = hist.corr.list$P[1, -1])
  
  return(histone.corr.df=histone.corr.df)
}


clustering = data.frame(row.names=c('LMNB1', 'late_replicating', 'H3K9me2',
                                    'H3K27me3', 'EZH2', 'CTCF', 'SMC3',
                                    'HDAC3', 'HDAC2', 'HDAC1', 'H3K4me1',
                                    'H3K4me2', 'H3K4me3', 'H3K27ac',
                                    'H4K5acK8ac', 'H2AFZ', 'DNAse', 'Dam',
                                    'm5C', 'H3K79me2', 'TTseq', 'H3K36me3',
                                    'POL2AS2', 'POL2'),
                        group=factor(c(rep('repressive', 5), rep('insulator', 2),
                                       rep('HDAC', 3), rep('euchromatin', 6),
                                       rep('accessibility', 2), 'methylation',
                                       rep('transcribing', 5)),
                                     levels=c('transcribing', 'accessibility',
                                              'methylation', 'euchromatin',
                                              'HDAC', 'insulator', 'repressive')))

CorrelationPlotMean <- function(data, condition_name, feature_name, targets, p.value=0.001){
  cor.df = data.frame()
  for (i in feature_name){
    corr.list <- data %>% filter(exp_pool==condition_name) %>%
      dplyr::select(i, as.character(targets)) %>%
      as.matrix() %>%
      rcorr(type = "pearson")
    
    df <- data.frame(rowname = names(corr.list$r[1, -1]),
                     corr = corr.list$r[1, -1],
                     pvalue = corr.list$P[1, -1],
                     feature = i)
    cor.df <- rbind(cor.df, df)
  }
  p <- cor.df %>%
    transform(rowname = factor(rowname, levels = levels(targets)),
              corr = corr,
              pvalue = pvalue) %>%
    ggplot(aes(x = rowname, y = corr, fill = feature, alpha = pvalue < p.value)) +
    geom_bar(stat = "identity", position = "dodge") +
    ylab(paste("Chromatin feature correlation with total indel frequency")) +
    xlab("Chromatin Feature") +
    theme_bw(base_size = 16) +
    scale_fill_manual("legend", values = colores) +
    scale_y_continuous(breaks=seq(-1,1, .1)) +
    coord_flip()
  return(list(#dataframe=cor.df,
    plot = p))
}


cor_test <- function(y, X, left, right){
  cor_list = lapply(X, function(x){
    cor.test(x,y)
  })
  p_vec = unlist(lapply(cor_list, function(x){x$p.value}))
  cor_vec = unlist(lapply(cor_list, function(x){x$estimate}))
  p_adj = p.adjust(p_vec)
  color = ifelse(p_adj < 0.05, ifelse(cor_vec > 0, left, right),
                 'None')
  importance = abs(cor_vec)
  target=factor(colnames(X), levels=rownames(clustering))
  feature_group = clustering[colnames(X), ]
  return(list(correlation=cor_vec,importance = importance, p=p_vec,
              p_adj=p_adj, target=target, color=color,
              feature_group=feature_group))
}


get_combinations <- function(domain_vec){
  n_domain = length(domain_vec)
  get_ivec <-function (last_x, n_domain){
    if (last_x < n_domain){
      i_vec = c(0, c((last_x+1):n_domain))
    } else{
      i_vec = c(0)
    }
  }
  
  combine <- function(x_new, x_vec, n_domain){
    new_vec = c(x_vec, x_new)
    if (length(new_vec)==n_domain){
      return(new_vec)
    } else {
      i_vec = get_ivec(max(new_vec), n_domain)
      x_list = lapply(i_vec, combine, x_vec=new_vec,
                      n_domain=n_domain)
      return(do.call(rbind, x_list))
    }
  }
  
  comb_list = lapply(c(0:n_domain), combine, x_vec=c(), n_domain=n_domain)
  
  comb = do.call(rbind, comb_list)
  comb_df = unique(t(apply(comb, 1, function(x){x[order(x)]})))
  
  name_list = apply(comb_df, 1, function(x){
    name = paste(domain_vec[x], collapse='-')
    if (name == ''){
      name = 'iDomain'
    }
    return(name)
  })
  
  opt_list = lapply(domain_vec, function(d){
    apply(comb_df, 1, function(x){
      opt = domain_vec[x]
      return(d%in%opt)
    })
  })
  
  opt_df = do.call(cbind, opt_list)
  colnames(opt_df) = domain_vec
  rownames(opt_df) = name_list
  
  return(opt_df)
}

test_wilcox <- function(dt, group_by='condition', y_name='efficiency'){
  bc_vec = dt[,unique(barcode)]
  name1 = unlist(unique(dt[,group_by, with=F])[1])
  dt_c = dt[get(group_by)==eval(name1), ]
  dt_s = dt[get(group_by)!=eval(name1), ]
  setkey(dt_c, 'barcode')
  setkey(dt_s, 'barcode')
  w = wilcox.test(unlist(dt_c[bc_vec,..y_name]),
                  unlist(dt_s[bc_vec,..y_name]),
                  paired=T)
  return(list(p_value = w$p.value))
}

quantiles_95 <- function(x) {
  r <- quantile(x, probs=c(0.05, 0.25, 0.5, 0.75, 0.95))
  names(r) <- c("ymin", "lower", "middle", "upper", "ymax")
  r
}


plot_comb_grid_beeswarm <- function(bix.dt, opt_df, domains, y_name, min_count=0,
                                    max_count=Inf, group_by=NULL, color=NULL, y_plabel=c(0.15, 0.25)){
  
  opt_melt = reshape2::melt(opt_df)
  colnames(opt_melt) = c('group', 'domain', 'is_in')
  
  opt_melt$is_in = ifelse(opt_melt$is_in, 'in', 'out')
  opt_melt$domain = factor(opt_melt$domain, levels=domains)
  
  
  group_list = lapply(rownames(opt_df), function(opt_name){
    opt = opt_df[opt_name, ]
    in_vec = domains[opt]
    out_vec = domains[!opt]
    
    in_sum = rowSums(bix.dt[,..in_vec])
    out_sum = rowSums(bix.dt[,..out_vec])
    if (length(in_vec)==0){
      in_group = which(out_sum == 0)
    } else if (length(out_vec) == 0){
      in_group = which(in_sum==length(in_vec))
    } else  {
      in_group = which(in_sum == length(in_vec) & out_sum==0)
    }
    return(in_group)
  })
  
  names(group_list) = rownames(opt_df)
  for (opt_name in names(group_list)){
    i_vec = group_list[[opt_name]]
    bix.dt[i_vec, group := opt_name]
  }
  if (is.null(group_by)){
    group_count_df = data.frame(bix.dt[,table(group)])
  } else {
    group_count_df = data.frame(bix.dt[,table(group)/length(unique(get(group_by)))])
  }
  group_count_df = group_count_df[order(match(group_count_df$group, rownames(opt_df))),]
  # print(group_count_df)
  group_count_sel = group_count_df[group_count_df$Freq>=min_count,]
  group_count_sel = group_count_sel[group_count_sel$Freq<=max_count,]
  group_levels =  group_count_sel$group
  opt_sel = opt_melt[opt_melt$group %in% group_levels, ]
  
  opt_sel$group = factor(opt_sel$group, levels=group_levels)
  # print(opt_sel)
  
  indel_selection = bix.dt[group%in%group_levels, ]
  indel_selection[,group:=factor(group, levels=group_levels)]
  
  # wilcox_list = lapply(group_levels[group_levels!='iDomain'],
  #                      function(g){
  #                          x = indel_selection[group==g, y_name, with=F]
  #                          y = indel_selection[group=='iDomain', y_name,
  #                                              with=F]
  #                          wilcox.test(unlist(x),unlist(y))
  #                      })
  
  group_count_sel$group = factor(group_count_sel$group, levels=group_levels)
  
  count_plot = ggplot(group_count_sel, aes(x=group,y=1, label=Freq)) +
    geom_label() +
    theme_void()
  
  
  
  dots = ggplot(opt_sel, aes(x=group, y=domain, color=is_in)) +
    geom_point(size=2) +
    geom_line(data=opt_sel[opt_sel$is_in=='in',], aes(group=group)) +
    theme_bw() +
    scale_color_manual(values=c('in'='black', 'out'='grey70')) +
    guides(color=FALSE)  +
    geom_vline(xintercept=seq(1.5, length(unique(bix.dt$group))-0.5, 1),
               lwd=0.5, colour="grey90") +
    theme(axis.title=element_blank(),
          axis.text=element_blank(),
          axis.ticks=element_blank(),
          panel.grid.major = element_blank())
  
  symnum = list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1),
                symbols = c("****", "***", "**", "*", " "))
  
  if (is.null(group_by)){
    level_vec = indel_selection[, levels(group)]
    wilcox_list = lapply(level_vec[level_vec!='iDomain'],
                         function(g){
                           x = indel_selection[group==g, y_name, with=F]
                           y = indel_selection[group=='iDomain', y_name,
                                               with=F]
                           w = wilcox.test(unlist(x),unlist(y))
                           return(data.table('group1'='iDomain', 'group2'=g,
                                             'p'=w$p.value))
                         })
    wilcox_dt = do.call(rbind, wilcox_list)
    wilcox_dt[,p_adj := p.adjust(p)]
    wilcox_dt[,p_label := sprintf('%.3g', p_adj)]
    wilcox_dt$y.position = seq(y_plabel[1], y_plabel[2],
                               length.out=nrow(wilcox_dt))
    
    comparisons = lapply(level_vec[level_vec!='iDomain'],
                         function(x){return(c('iDomain', x))})
    beeswarm = ggplot(indel_selection, aes_string(x='group', y=y_name)) +
      geom_quasirandom(width=0.3) +
      stat_summary(fun.y=median, fun.ymin = median,
                   fun.ymax = median,
                   geom="crossbar", width = 0.5,
                   color='red') +
      stat_pvalue_manual(wilcox_dt, label = "p_label")
  } else {
    wilcox_dt = indel_selection[,test_wilcox(.SD, group_by, y_name),
                                by='group']
    wilcox_dt[,p_adj:=p.adjust(p_value)]
    wilcox_dt[,p_signif:=symnum(p_adj, cutpoints=symnum$cutpoints,
                                symbols=symnum$symbols)]
    beeswarm = ggplot(indel_selection,
                      aes_string(x='group', y=y_name, color=group_by)) +
      geom_quasirandom(dodge.width=1) +
      scale_color_manual(values=color) +
      stat_summary(aes_string(group=group_by), fun.y=median,
                   fun.ymin = median, fun.ymax = median,
                   geom="crossbar", width = 0.5,
                   color='red',
                   position= position_dodge(width =1)) +
      geom_text(data=wilcox_dt, aes(x=group, y=0.8, label=p_signif),
                inherit.aes=F)
  }
  
  # if (is.null(group_by)){
  #   beeswarm = ggplot(indel_selection, aes_string(x='group', y=y_name)) +
  #     geom_quasirandom(width=0.3) +
  #     stat_summary(fun.y=median, fun.ymin = median,
  #                  fun.ymax = median,
  #                  geom="crossbar", width = 0.5,
  #                  color='red') +
  #     stat_compare_means(aes(label = ..p.signif..),
  #                        symnum.args = symnum,
  #                        method = 'wilcox.test',
  #                        ref.group='iDomain')
  # } else {
  #     wilcox_dt = indel_selection[,test_wilcox(.SD, group_by, y_name),
  #                                 by='group']
  #     wilcox_dt[,p_adj:=p.adjust(p_value)]
  #     wilcox_dt[,p_signif:=symnum(p_adj, cutpoints=symnum$cutpoints,
  #                                 symbols=symnum$symbols)]
  # 
  # 
  #     beeswarm = ggplot(indel_selection,
  #                       aes_string(x='group', y=y_name, color=group_by)) +
  #     			   geom_quasirandom(dodge.width=1) +
  #                    scale_color_manual(values=color) +
  #                    stat_summary(aes_string(group=group_by), fun.y=median,
  #                                 fun.ymin = median, fun.ymax = median,
  #                                 geom="crossbar", width = 0.5,
  #                                 color='red',
  #                                 position= position_dodge(width =1)) +
  #                    geom_text(data=wilcox_dt, aes(x=group, y=0.8, label=p_signif),
  #                              inherit.aes=F)
  # }
  beeswarm = beeswarm +
    theme_bw() +
    geom_vline(xintercept=seq(1.5, length(unique(bix.dt$group))-0.5, 1),
               lwd=0.5, colour="grey90") +
    theme(axis.title=element_blank(),
          axis.text.x=element_blank(),
          panel.grid.major.x = element_blank(),
          axis.ticks.x=element_blank()) +
    ylim(0, 1)
  
  
  count = data.frame(domain=gsub('_domain', '', domains),
                     count=colSums(bix.dt[,..domains], na.rm=T))
  
  count$domain = factor(count$domain, levels=domain_levels)
  
  
  histogram = ggplot(count, aes(x=1, y=domain, label = count)) +
    geom_label() +
    theme_bw() +
    theme(plot.background = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_blank(),
          axis.title=element_blank(),
          axis.text.x=element_blank(),
          axis.ticks=element_blank())
  
  empty = ggplot(data=data.frame(1,1),aes(x=1,y=1)) +
    geom_point() +
    theme_void()
  
  count_beeswarm = plot_grid(beeswarm, count_plot, ncol=1, rel_heights=c(20,1),
                             align='v', axis='lr')
  return(plot_grid(empty, beeswarm, empty, count_plot,
                   histogram, dots, nrow=3,
                   rel_widths=c(1,10), rel_heights=c(15,1,5), align='vh',
                   axis='tlbr'))
}

plot_comb_grid_bothplot <- function(bix.dt, opt_df, domains, y_name, min_count=0,
                                   max_count=Inf, group_by=NULL, color=NULL){
  
  opt_melt = reshape2::melt(opt_df)
  colnames(opt_melt) = c('group', 'domain', 'is_in')
  
  opt_melt$is_in = ifelse(opt_melt$is_in, 'in', 'out')
  opt_melt$domain = factor(opt_melt$domain, levels=domains)
  
  
  group_list = lapply(rownames(opt_df), function(opt_name){
    opt = opt_df[opt_name, ]
    in_vec = domains[opt]
    out_vec = domains[!opt]
    
    in_sum = rowSums(bix.dt[,..in_vec])
    out_sum = rowSums(bix.dt[,..out_vec])
    if (length(in_vec)==0){
      in_group = which(out_sum == 0)
    } else if (length(out_vec) == 0){
      in_group = which(in_sum==length(in_vec))
    } else  {
      in_group = which(in_sum == length(in_vec) & out_sum==0)
    }
    return(in_group)
  })
  
  names(group_list) = rownames(opt_df)
  for (opt_name in names(group_list)){
    i_vec = group_list[[opt_name]]
    bix.dt[i_vec, group := opt_name]
  }
  if (is.null(group_by)){
    group_count_df = data.frame(bix.dt[,table(group)])
  } else {
    group_count_df = data.frame(bix.dt[,table(group)/length(unique(get(group_by)))])
  }
  group_count_df = group_count_df[order(match(group_count_df$group, rownames(opt_df))),]
  # print(group_count_df)
  group_count_sel = group_count_df[group_count_df$Freq>=min_count,]
  group_count_sel = group_count_sel[group_count_sel$Freq<=max_count,]
  group_levels =  group_count_sel$group
  opt_sel = opt_melt[opt_melt$group %in% group_levels, ]
  
  opt_sel$group = factor(opt_sel$group, levels=group_levels)
  # print(opt_sel)
  
  indel_selection = bix.dt[group%in%group_levels, ]
  indel_selection[,group:=factor(group, levels=group_levels)]

  group_count_sel$group = factor(group_count_sel$group, levels=group_levels)
  
  count_plot = ggplot(group_count_sel, aes(x=group,y=1, label=Freq)) +
    geom_label() +
    theme_void()
  
  
  
  dots = ggplot(opt_sel, aes(x=group, y=domain, color=is_in)) +
    geom_point(size=2) +
    geom_line(data=opt_sel[opt_sel$is_in=='in',], aes(group=group)) +
    theme_bw() +
    scale_color_manual(values=c('in'='black', 'out'='grey70')) +
    guides(color=FALSE)  +
    geom_vline(xintercept=seq(1.5, length(unique(bix.dt$group))-0.5, 1),
               lwd=0.5, colour="grey90") +
    theme(axis.title=element_blank(),
          axis.text=element_blank(),
          axis.ticks=element_blank(),
          panel.grid.major = element_blank())
  
  symnum = list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1),
                symbols = c("****", "***", "**", "*", " "))
  
  if (is.null(group_by)){
    boxpl = ggplot(indel_selection, aes_string(x='group', y=y_name)) +
      geom_quasirandom(width=0.3) +
      stat_summary(fun=median, fun.min = median,
                   fun.max = median,
                   geom="crossbar", width = 0.5,
                   color='red') +
      stat_compare_means(aes(label = ..p.signif..),
                         symnum.args = symnum,
                         method = 'wilcox.test',
                         ref.group='iDomain')
  } else {
    
    boxpl = ggplot(indel_selection,
                   aes_string(x='group', y=y_name, color=group_by)) +
      guides(fill=F) +
      # scale_fill_manual(values=color) +
      scale_color_manual(values=color) +
      # geom_violin(position=position_dodge(width=1)) +
      geom_quasirandom(dodge.width = 1) +
      # stat_summary(fun.data = quantiles_95,
      #              geom="boxplot",
      #              position=position_dodge(width=1),
      #              width = 0.2) +
      stat_summary(fun=median, fun.min = median,
                   fun.max = median,
                   geom="crossbar", width = 0.6,
                   position=position_dodge(width=1))
  }
  boxpl = boxpl +
    theme_bw() +
    geom_vline(xintercept=seq(1.5, length(unique(bix.dt$group))-0.5, 1),
               lwd=0.5, colour="grey90") +
    theme(axis.title=element_blank(),
          axis.text.x=element_blank(),
          panel.grid.major.x = element_blank(),
          axis.ticks.x=element_blank()) +
    ylim(0, 1)
  
  
  count = data.frame(domain=gsub('_domain', '', domains),
                     count=colSums(bix.dt[,..domains], na.rm=T))
  
  count$domain = factor(count$domain, levels=domain_levels)
  
  
  histogram = ggplot(count, aes(x=1, y=domain, label = count)) +
    geom_label() +
    theme_bw() +
    theme(plot.background = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_blank(),
          axis.title=element_blank(),
          axis.text.x=element_blank(),
          axis.ticks=element_blank())
  
  empty = ggplot(data=data.frame(1,1),aes(x=1,y=1)) +
    geom_point() +
    theme_void()
  
  count_beeswarm = plot_grid(boxpl, count_plot, ncol=1, rel_heights=c(20,1),
                             align='v', axis='lr')
  return(plot_grid(empty, boxpl, empty, count_plot,
                   histogram, dots, nrow=3,
                   rel_widths=c(1,10), rel_heights=c(15,1,5), align='vh',
                   axis='tlbr'))
}

```

# Data import
```{r loading data, message=FALSE, warnings=FALSE}
trip_tib_mut_2000 = readRDS(paste0(in.dir, "RSTP2_Indel_Chromatin_2kb.RDS"))
trip_tib_2000 = readRDS(paste0(in.dir, "RSTP2_IndelRatios_Chromatin_2kb.RDS"))


# Import ChIP2 & ChIP3 experiments
ChIP_clone5 <- readRDS("/DATA/projects/DSBrepair/data/R/xv20200402_chipampcorrelation.rds")

# Select the final tib from xabi (contains log2 means) and add .B to the barcodes
selected.chip.tib <- ChIP_clone5$final.tib %>% ungroup() %>% mutate(barcode = paste0(barcode, ".B")) %>% group_by(barcode)
```


## Data loading

```{r preprocess experiments}
trip_tib_mut <- copy(trip_tib_2000)
trip_tib_2000 <- trip_tib_2000 %>% distinct(barcode, exp, .keep_all = TRUE)

trip_tib <- trip_tib_2000 %>%
  filter(exp_pool %in% c("mean_RSTP2_2000_LBR1_DMSO_64",
                         "mean_RSTP2_2000_LBR2_DMSO_64",
                         "mean_RSTP2_2000_LBR12_DMSO_64",
                         "mean_RSTP2_2000_LBR15_DMSO_64"))

trip_tib_mut <- trip_tib_mut_2000 %>%
  filter(exp_pool %in% c("mean_RSTP2_2000_LBR1_DMSO_64",
                         "mean_RSTP2_2000_LBR2_DMSO_64",
                         "mean_RSTP2_2000_LBR12_DMSO_64",
                         "mean_RSTP2_2000_LBR15_DMSO_64"))

mean_trip_tib_2000 <- trip_tib_2000 %>%
  gather(NHEJ, MMEJ, other_indels, key = "pathway", value = 'ratios') %>%
  distinct(barcode, exp, pathway, .keep_all = TRUE)

mean_trip_tib_2000$pathway <- factor(mean_trip_tib_2000$pathway, levels=c("other_indels", "MMEJ", "NHEJ"))
mean_trip_tib_2000 <- mean_trip_tib_2000[!is.na(mean_trip_tib_2000$ratios), ]

trip_tib$spin_state <- factor(trip_tib$spin_state, levels=c("Speckle",
                                                            "Interior_Act1",
                                                            "Interior_Act2",
                                                            "Interior_Act3",
                                                            "Interior_Repr1",
                                                            "Interior_Repr2",
                                                            "Near_Lm1",
                                                            "Near_Lm2",
                                                            "Near_Lm3",
                                                            "Lamina",
                                                            NA ))
```

# Main Figure
## Panel 3A
Schematic of the target sites of the guides


## Panel 3B
```{r Fig3B efficiency LBR1 and LBR2, fig.height=4, fig.width=5}
barcodes <- trip_tib %>% group_by(barcode) %>% dplyr::summarise(count = n()) %>% filter(count > 1) %>% pull(barcode)
length(barcodes)
trip_tib$plasmid <- factor(trip_tib$plasmid, levels = c("LBR1", "LBR12", "LBR15" ,"LBR2"))

trip_tib <- trip_tib %>% filter(barcode %in% barcodes)

trip_tib %>% group_by(plasmid) %>% dplyr::summarise(count = n())

ggplot(trip_tib, aes(efficiency, color = plasmid, alpha = 0.8)) +
  geom_density(trim = F, size = 1.5) +
  scale_color_manual(values = guide_colors) +
  theme_bw(base_size = 16) +
  theme(legend.position = c(0.2, 0.8))

```

### Figure Legend
(a) Efficiency density of two sgRNAs targeting the IPR (total of 1126 IPRs).

## Panel 3C
```{r Fig3C efficiency LBR2 vs LBR1, fig.height=5, fig.width=7}
p1 <- trip_tib %>% dplyr::select(barcode, plasmid, efficiency) %>% spread(plasmid, efficiency) %>% ggplot(., aes(LBR1, LBR2))  +  
  geom_point() +
  geom_smooth(method= "loess") +
  theme_bw(base_size = 16) +
  ylim(0, 1) +
  xlim(0, 1) +
  stat_cor(method = "spearman", label.x.npc = "left", label.y.npc = "bottom") + 
  coord_fixed()

p2 <- trip_tib %>% dplyr::select(barcode, plasmid, efficiency) %>% spread(plasmid, efficiency) %>% ggplot(., aes(LBR12, LBR2))  +  
  geom_point() +
  geom_smooth(method= "loess") +
  ylim(0, 1) +
  xlim(0, 1) +
  theme_bw(base_size = 16) +
  stat_cor(method = "spearman", label.x.npc = "left", label.y.npc = "bottom") + 
  coord_fixed()

p3 <- trip_tib %>% dplyr::select(barcode, plasmid, efficiency) %>% spread(plasmid, efficiency) %>% ggplot(., aes(LBR15, LBR2)) +  
  geom_point() +
  geom_smooth(method= "loess") +
  ylim(0, 1) +
  xlim(0, 1) +
  theme_bw(base_size = 16) +
  stat_cor(method = "spearman", label.x.npc = "left", label.y.npc = "bottom") + 
  coord_fixed()

plot_grid(p1, p2, p3, nrow = 2)
```

### Figure Legend
(b) Spearman correlation between the two sgRNAs for each IPR.

## Panel 3D
```{r Fig3D efficiency LBR2 LBR1 vs chrom, fig.height=9, fig.width=9}
chrom.mods.cols <- seq(grep("binsize", names(trip_tib))+1, grep("H3K27me3_domain", names(trip_tib))-1)

trip_tib_eff <- trip_tib %>% dplyr::select(efficiency, barcode, chrom.mods.cols, plasmid) %>% spread(plasmid, efficiency) %>% na.omit() %>% mutate(exp_pool = "Efficiency_Corr")

chrom.mods.cols <- seq(grep("barcode", names(trip_tib_eff))+1, grep("LBR1$", names(trip_tib_eff))-1)

eff.hist.corr <- CorrelationDFMean(trip_tib_eff, "Efficiency_Corr", c("LBR2"), targets = chrom.mods.cols) %>%
  arrange(corr) %>% pull(rowname) %>% factor(.,levels = .)

lims = 1

colores <- guide_colors

CorrelationPlotMean(trip_tib_eff, "Efficiency_Corr",  c("LBR1", "LBR2", "LBR12",  "LBR15") , targets = eff.hist.corr)

```

### Figure Legend
(c) Pearsons correlation of efficiency for both LBR2 (black) and LBR1 (gray) sgRNAs with different chromatin features.

## Panel 3E
```{r Fig3E heterochrom domains efficiency, fig.height=7, fig.width=12}
tib.dt = data.table(trip_tib)

domain_levels = c('late_replicating', 'LAD', 'H3K9me2', 'H3K27me3')

domains = paste0(domain_levels, '_domain')

indel.dt = unique(tib.dt[exp_pool %in% c("mean_RSTP2_2000_LBR1_DMSO_64",
                                         "mean_RSTP2_2000_LBR2_DMSO_64",
                                         "mean_RSTP2_2000_LBR12_DMSO_64",
                                         "mean_RSTP2_2000_LBR15_DMSO_64"),
                         c('exp_pool', 'barcode', 'efficiency',
                           domains), with=F])

indel.dt[indel.dt=='iDomain'] <- 0
indel.dt[indel.dt=='Domain'] <- 1

indel.dt[is.na(LAD_domain), LAD_domain:=0]

indel.dt[, (domains) := lapply(.SD, as.numeric), .SDcols = domains]

colnames(indel.dt) = gsub('_domain', '', colnames(indel.dt))

opt_df = get_combinations(domain_levels)
#
plot_comb_grid_beeswarm(indel.dt[exp_pool=='mean_RSTP2_2000_LBR2_DMSO_64'], opt_df,
                        domain_levels, 'efficiency', min_count=20, y_plabel=c(0.15, 0.25))

# plot_comb_grid(indel.dt[exp_pool=='mean_RSTP2_2000_LBR1_DMSO_64'], opt_df,
#                    domain_levels, 'efficiency', min_count=20)
#
# plot_comb_grid(indel.dt[exp_pool=='mean_RSTP2_2000_LBR12_DMSO_64'], opt_df,
#                    domain_levels, 'efficiency', min_count=20)
#
# plot_comb_grid(indel.dt[exp_pool=='mean_RSTP2_2000_LBR15_DMSO_64'], opt_df,
#                    domain_levels, 'efficiency', min_count=20)
```


### Figure Legend
(d) Efficiency per IPR, split into different combinations of heterochromatin domains. Filtered on at least 20 IPRs per group, all groups in S3a. LBR2 in black and LBR1 in grey. Numbers represent the amount of IPRs in each respective group. t.test compared to non-heterochromatin IPRs, * p < 0.5, ** p < 0.1, *** p < 0.001, **** p < 0.0001.

## Panel 3F
```{r Fig3F heterochrom domains efficiency LBR1-12-15, fig.height=7, fig.width=12}
domain_levels = c('late_replicating', 'LAD', 'H3K9me2', 'H3K27me3')

domains = paste0(domain_levels, '_domain')

barcodes_all_plasmids <- trip_tib %>%
  group_by(barcode) %>%
  dplyr::summarise(n = n()) %>%
  filter(n == 4) %>%
  pull(barcode)

indel.dt = trip_tib %>%
  dplyr::select(barcode, plasmid, efficiency, domains) %>%
  filter(barcode %in% barcodes_all_plasmids, plasmid != 'LBR2') %>%
  spread(plasmid, efficiency) %>%
  distinct(barcode, .keep_all = TRUE) %>%
  data.table()


indel.dt[indel.dt=='iDomain'] <- 0
indel.dt[indel.dt=='Domain'] <- 1

indel.dt[is.na(LAD_domain), LAD_domain:=0]

indel.dt[, (domains) := lapply(.SD, as.numeric), .SDcols = domains]

colnames(indel.dt) = gsub('_domain', '', colnames(indel.dt))

opt_df = get_combinations(domain_levels)

### CHRIST ####
# Can you now plot the ratio MMEJ_MMEJNHEJ of these domains that are significant vs the control condition to see the fold change and significance?

eff.dt = melt(indel.dt, measure.vars=c('LBR1', 'LBR12', 'LBR15'),
              variable.name='condition', value.name='efficiency')

guide_cols <- guide_colors[1:3]

plot_comb_grid_bothplot(eff.dt, opt_df, domain_levels, 'efficiency', group_by='condition', color=guide_cols, min_count = 20)

groups <- unique(eff.dt$group)[c(1, 3:13)]

for(i in groups) {
  pvals <- eff.dt %>% 
    dplyr::select(condition, efficiency, group) %>% 
    group_by(condition) %>%
    filter(group %in% c("iDomain", i)) %>%
    do(broom::tidy(wilcox.test(efficiency ~ group, data = .))) %>%
    ungroup() %>%
    mutate(p.adj = p.adjust(p.value, method = "fdr")) %>%
    dplyr::select(condition, statistic, p.value, p.adj) %>%
    arrange(p.adj)
  print(i)
  print(pvals)
} 
       
```


### Figure Legend
(F) Same as E, but boxplots for the other three sgRNAs. The boxes represent 75% confidence interval, the horizontal line within represents the median, the error bars represent and 95% confidence intervals.


# Supplementary Figure 3
## Panel S3A
```{r FigS3A indelspectra other guides, fig.height=8, fig.width=10}
ggplot(filter(trip_tib_mut, mutation != "ssODN", nexp > 1),
       aes(mutation, freq)) +
  geom_quasirandom(size = .5) +
  theme_bw(base_size = 16) +
  ylim(0, 1) +
  facet_grid(plasmid ~ .)
```

### Figure Legend

## Panel S3B
```{r FigS3B correlation TIF all guides, fig.height=5, fig.width=5}
res = trip_tib %>% dplyr::select(barcode, plasmid, efficiency) %>% 
  spread(plasmid, efficiency) %>% 
  column_to_rownames(., var = "barcode") %>%
  as.matrix() %>%
  rcorr(., type="spearman")

library(corrplot)
corrplot(res$r, type = "upper", order = "hclust", 
         tl.col = "black", tl.srt = 45, diag = FALSE)
```

### Figure Legend

## Panel S3C
```{r FigS3C chip clone 5, fig.height=4, fig.width=12}
Chip.seq.data.melt <- trip_tib_2000 %>% filter(cell_line == "RSTP2_clone5")  %>% distinct(barcode, .keep_all = TRUE) %>% dplyr::select(barcode, H3K27me3, H3K4me1, H3K27ac) %>% gather(H3K27me3, H3K4me1, H3K27ac, key = antibody, value = value)

clone5_domains <- readRDS("/DATA/projects/DSBrepair/data/R/rs20200311_domains_clone5.RDS")


signal.amp.seq <- left_join(selected.chip.tib,Chip.seq.data.melt, by = c("barcode", "antibody")) %>% left_join(clone5_domains, by = "barcode")

#Plot with free axis scales
ggplot(signal.amp.seq, aes(value,amp.seq)) + 
  geom_point() + 
  geom_smooth(method = "lm") + 
  facet_wrap(~ antibody, scales= "free") + 
  ylab("log2(normalized barcode reads)") + 
  xlab("log2(ChIP-seq signal)") + 
  stat_cor(method = "pearson", label.x.npc = 0, label.y.npc = 1, vjust = 0) +
  theme_bw(base_size = 16)

```

### Figure Legend
*(A)* Pearson's correlation between log2 normalized IPR PCR reads on ChIP of clone5 versus log2 ChIP-seq signal over 2kb surrounding the IPR location from published data. References in table S1.

<!-- ## Supplementary Figure 4 -->

<!-- ## Panel S3A -->
<!-- ```{r FigS3A heterochrom domains efficiency supp, fig.height=8, fig.width=16} -->
<!-- indel.dt = trip_tib %>% -->
<!--   dplyr::select(barcode, plasmid, efficiency, domains) %>% -->
<!--   filter(barcode %in% barcodes_all_plasmids) %>% -->
<!--   spread(plasmid, efficiency) %>% -->
<!--   distinct(barcode, .keep_all = TRUE) %>% -->
<!--   data.table() -->


<!-- indel.dt[indel.dt=='iDomain'] <- 0 -->
<!-- indel.dt[indel.dt=='Domain'] <- 1 -->

<!-- indel.dt[is.na(LAD_domain), LAD_domain:=0] -->

<!-- indel.dt[, (domains) := lapply(.SD, as.numeric), .SDcols = domains] -->

<!-- colnames(indel.dt) = gsub('_domain', '', colnames(indel.dt)) -->

<!-- opt_df = get_combinations(domain_levels) -->

<!-- eff.dt = melt(indel.dt, measure.vars=c('LBR1', 'LBR12', 'LBR15', 'LBR2'), -->
<!--               variable.name='condition', value.name='efficiency') -->

<!-- plot_comb_grid_boxplot(eff.dt, opt_df, domain_levels, 'efficiency', group_by='condition', color=guide_colors) -->

<!-- ``` -->

<!-- ### Figure Legend -->

# Bibliography
```{r citations}
cite_packages()
```


# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```
