---
title: "Figure 3 and supplemental 2-3"
author: "Ruben Schep"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output: 
  html_document:
    theme: journal #cerulean
    highlight: monochrome
    toc: true
    toc_float: true
    code_folding: show
  editor_options:
    chunk_output_type: console
---

# knitr document van Steensel lab
```{r}
StartTime <-Sys.time()
library(knitr)

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 

## Select outdir
out.dir = paste0("figures/rs", Date, "/")
dir.create(out.dir)
opts_chunk$set(dev=c('png', 'pdf'), fig.path = file.path(out.dir))
pdf.options(useDingbats = FALSE)
```

# Introduction

## Description of Data

# Data processing
## Path, Libraries, Parameters
```{r setup, message=FALSE, warnings=FALSE}


# INPUT DIRECTORY
in.dir.date = 20200121
in.dir = paste0("/DATA/projects/DSBrepair/data/R/rs", in.dir.date, "/")


# libraries:
library(ggplot2)
library(ggpubr)
library(corrr)
library(Hmisc)
library(ggbeeswarm)
library(RColorBrewer)
library(dplyr)
library(tidyr)
library(ggpmisc)
library(UpSetR)
library(stringr)
library(gridExtra)
library(data.table)
library(cowplot)
library(grid)

color_redblue <- rev(brewer.pal(11,"RdBu"))
duocolor <- c("#EE756E", "#2BB6BE")
colore <- c("wt" = "#808184", "other" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "SSTR" = "#007A4C")
colores <- c("wt" = "#808184", "other_indels" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "SSTR" = "#007A4C")
colori  <- c("other_indels" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "ssODN" = "#007A4C")
colora <- c("0" = "#808184", "1" = "#E1251B", "-7" = "#223DA0", "-14" = "#223DA0")
coloru <- c("wt" = "#808184", "other" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "MMEJplusNHEJ" = "#EE8A2A", "all" = "black", "MMEJplusNHEJplusHDR" = "#EE8A2A", "HDR" = "#007A4C", "MMEJplusHDR" = "#EE8A2A")
colory <- c("wt" = "#808184", "other" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "MMEJplusNHEJ" = "#EE8A2A", "all" = "black", "SSTR" = "#007A4C")
```

## Custom functions
Functions used thoughout this script.
```{r, message=FALSE, warnings=FALSE}
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, substr(gsub("-","",Sys.time()),1,8), "_", filename)
  filename
}

CorrelationDFMean <- function(data, condition_name, feature_name, targets, p.value=0.001){
  hist.corr.list <- data[data$exp_pool==condition_name, ] %>% 
    dplyr::select(feature_name, targets) %>% 
    as.matrix() %>% 
    rcorr(type = "pearson")
  
  histone.corr.df <- data.frame(rowname = names(hist.corr.list$r[1, -1]), 
                                corr = hist.corr.list$r[1, -1], 
                                pvalue = hist.corr.list$P[1, -1])
  
  return(histone.corr.df=histone.corr.df)
}


CorrelationPlotMean <- function(data, condition_name, feature_name, targets, p.value=0.001){
  cor.df = data.frame()
  for (i in feature_name){
    corr.list <- data %>% filter(exp_pool==condition_name) %>% 
      dplyr::select(i, as.character(targets)) %>% 
      as.matrix() %>% 
      rcorr(type = "pearson")
  
    df <- data.frame(rowname = names(corr.list$r[1, -1]), 
                                corr = corr.list$r[1, -1], 
                                pvalue = corr.list$P[1, -1],
                                feature = i)
    cor.df <- rbind(cor.df, df)
  }
  p <- cor.df %>% 
    transform(rowname = factor(rowname, levels = levels(targets)), 
              corr = corr, 
              pvalue = pvalue) %>%
    ggplot(aes(x = rowname, y = corr, fill = feature, alpha = pvalue < p.value)) +
    geom_bar(stat = "identity", position = "dodge") +
    ylab(paste("Pearson correlation with", feature_name)) +
    xlab("Chromatin Feature") + 
    ggtitle(paste("Repair pathways correlation with chromatin features in", condition_name, "exps")) +
    theme_bw(base_size = 16) +
    scale_fill_manual("legend", values = colores) +
    scale_y_continuous(breaks=seq(-1,1, .1)) + 
    coord_flip()
  return(list(dataframe=cor.df, plot = p))
}

get_combinations <- function(domain_vec){
    n_domain = length(domain_vec)
    get_ivec <-function (last_x, n_domain){
    	if (last_x < n_domain){
    		i_vec = c(0, c((last_x+1):n_domain))
    	} else{
    		i_vec = c(0)
    	}
    }

    combine <- function(x_new, x_vec, n_domain){
        new_vec = c(x_vec, x_new)
        if (length(new_vec)==n_domain){
            return(new_vec)
        } else {
            i_vec = get_ivec(max(new_vec), n_domain)
            x_list = lapply(i_vec, combine, x_vec=new_vec,
                            n_domain=n_domain)
            return(do.call(rbind, x_list))
        }
    }

    comb_list = lapply(c(0:n_domain), combine, x_vec=c(), n_domain=n_domain)

    comb = do.call(rbind, comb_list)
    comb_df = unique(t(apply(comb, 1, function(x){x[order(x)]})))

    name_list = apply(comb_df, 1, function(x){
    	name = paste(domain_vec[x], collapse='-')
    	if (name == ''){
    		name = 'iDomain'
    	}
    	return(name)
    })

    opt_list = lapply(domain_vec, function(d){
    	 apply(comb_df, 1, function(x){
    		opt = domain_vec[x]
    		return(d%in%opt)
    	})
    })

    opt_df = do.call(cbind, opt_list)
    colnames(opt_df) = domain_vec
    rownames(opt_df) = name_list

    return(opt_df)
}

plot_comb_grid <- function(indel.dt, opt_df, domains, y_name, min_count=0, max_count=Inf){

    opt_melt = reshape2::melt(opt_df)
    colnames(opt_melt) = c('group', 'domain', 'is_in')

    opt_melt$is_in = ifelse(opt_melt$is_in, 'in', 'out')
    opt_melt$domain = factor(opt_melt$domain, levels=domains)


    group_list = lapply(rownames(opt_df), function(opt_name){
    	opt = opt_df[opt_name, ]
    	in_vec = domains[opt]
    	out_vec = domains[!opt]

    	in_sum = rowSums(indel.dt[,..in_vec])
    	out_sum = rowSums(indel.dt[,..out_vec])
    	if (length(in_vec)==0){
    		in_group = which(out_sum == 0)
    	} else if (length(out_vec) == 0){
    		in_group = which(in_sum==length(in_vec))
    	} else  {
    		in_group = which(in_sum == length(in_vec) & out_sum==0)
    	}
    	return(in_group)
    })

    names(group_list) = rownames(opt_df)
    for (opt_name in names(group_list)){
    	i_vec = group_list[[opt_name]]
    	indel.dt[i_vec, group := opt_name]
    }
    group_count_df = data.frame(indel.dt[,table(group)])
    group_count_df = group_count_df[order(match(group_count_df$group, rownames(opt_df))),]
    group_count_sel = group_count_df[group_count_df$Freq >= min_count,]
    group_count_sel = group_count_sel[group_count_sel$Freq <= max_count,]
    group_levels =  group_count_sel$group
    opt_sel = opt_melt[opt_melt$group %in% group_levels, ]

    opt_sel$group = factor(opt_sel$group, levels=group_levels)

    indel_selection = indel.dt[group%in%group_levels, ]
    indel_selection[,group:=factor(group, levels=group_levels)]

    # wilcox_list = lapply(group_levels[group_levels!='iDomain'],
    #                      function(g){
    #                          x = indel_selection[group==g, y_name, with=F]
    #                          y = indel_selection[group=='iDomain', y_name,
    #                                              with=F]
    #                          wilcox.test(unlist(x),unlist(y))
    #                      })

    group_count_sel$group = factor(group_count_sel$group, levels=group_levels)

    count_plot = ggplot(group_count_sel, aes(x=group,y=1, label=Freq)) +
                            geom_label() +
                            theme_void()



    dots = ggplot(opt_sel, aes(x=group, y=domain, color=is_in)) +
    	geom_point(size=2) +
    	geom_line(data=opt_sel[opt_sel$is_in=='in',], aes(group=group)) +
    	theme_bw() +
    	scale_color_manual(values=c('in'='black', 'out'='grey70')) +
    	guides(color=FALSE)  +
    	geom_vline(xintercept=seq(1.5, length(unique(indel.dt$group))-0.5, 1),
    			   lwd=0.5, colour="grey90") +
    	theme(axis.title=element_blank(),
    		  axis.text=element_blank(),
    		  axis.ticks=element_blank(),
    		  panel.grid.major = element_blank())

    symnum = list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1),
                  symbols = c("****", "***", "**", "*", ""))

    beeswarm = ggplot(indel_selection, aes_string(x='group', y=y_name)) +
    			  geom_quasirandom(width=0.3) +
    			  theme_bw() +
                  ylim(0,1) +
    			  geom_vline(xintercept=seq(1.5, length(unique(indel.dt$group))-0.5, 1),
    			             lwd=0.5, colour="grey90") +
                  stat_summary(fun.y=median, fun.ymin = median,
                               fun.ymax = median, geom="crossbar", width = 0.5,
                               color='red') +
                  stat_compare_means(aes(label = ..p.signif..),
                                     symnum.args = symnum,
                                     ref.group='iDomain') +
    			  theme(axis.title=element_blank(),
    			        axis.text.x=element_blank(),
      				    panel.grid.major.x = element_blank(),
    				    axis.ticks.x=element_blank())

    count = data.frame(domain=gsub('_domain', '', domains),
    	               count=colSums(indel.dt[,..domains], na.rm=T))

    count$domain = factor(count$domain, levels=domain_levels)


    histogram = ggplot(count, aes(x=1, y=domain, label = count)) +
    				geom_label() +
    				theme_bw() +
    				theme(plot.background = element_blank(),
        				  panel.grid.major = element_blank(),
        				  panel.grid.minor = element_blank(),
        				  panel.border = element_blank(),
    					  axis.title=element_blank(),
    	  			      axis.text.x=element_blank(),
    	  				  axis.ticks=element_blank())

    empty = ggplot(data=data.frame(1,1),aes(x=1,y=1)) +
                geom_point() +
                theme_void()

    count_beeswarm = plot_grid(beeswarm, count_plot, ncol=1, rel_heights=c(20,1),
                               align='v', axis='lr')
    return(plot_grid(empty, beeswarm, empty, count_plot,
                     histogram, dots, nrow=3,
                     rel_widths=c(1,10), rel_heights=c(15,1,5), align='vh',
                     axis='tlbr'))
}

```

# Data import
```{r, message=FALSE, warnings=FALSE}
load(paste0(in.dir, "RSTP2_Indel_Chromatin_2kb.RData"))
load(paste0(in.dir, "RSTP2_IndelRatios_Chromatin_2kb.RData"))


# Import ChIP2 & ChIP3 experiments
load("/DATA/projects/DSBrepair/data/R/xv20190515_ChIP3Timepoints_correctedv3.RData")
ChIP3 <- chip.analysis.tib
load("/DATA/projects/DSBrepair/data/R/xv20190514_ChIP2Chromatinmarks.RData")
ChIP2 <- chip.analysis.tib

# Select samples and columns we are interested in inpbcreads basically
ChIP3.selected <- ChIP3 %>% filter(antibody %in% c("H3K27me3") & sgRNA == "noguide") %>% dplyr::select(barcode, antibody, inpindelreads)
ChIP2.selected <- ChIP2 %>% filter(antibody %in% c("H3K4me1","H3K27ac","H3K9me3") & sgRNA == "noguide") %>% dplyr::select(barcode, antibody, inpindelreads)

#Merge both tibs in a single tib
selected.chip.tib <- plyr::rbind.fill(ChIP2.selected,ChIP3.selected)

#log2 transform inpindelreads
log2.sel.amp.tib <- selected.chip.tib %>% 
  mutate(amp.indelpcr.mean = log2(inpindelreads),
         barcode = paste0(barcode, ".B")) %>%
  dplyr::select(barcode, antibody, amp.indelpcr.mean)
```


## Data loading

```{r preprocess experiments}
trip_tib_mut <- copy(trip_tib_2000)
trip_tib_2000 <- trip_tib_2000 %>% distinct(barcode, exp, .keep_all = TRUE)

trip_tib <- trip_tib_2000 %>% 
  filter(exp_pool %in% c("indel_LBR2Rep5_LBR1_-_-_-_64", "indel_mean_LBR2_0_64h"))

trip_tib_mut <- trip_tib_mut_2000 %>%
  filter(exp_pool %in% c("indel_LBR2Rep5_LBR1_-_-_-_64", "indel_mean_LBR2_0_64h"))

mean_trip_tib_2000 <- trip_tib_2000 %>% 
  gather(NHEJ, MMEJ, other_indels, key = "pathway", value = 'ratios') %>% 
  distinct(barcode, exp, pathway, .keep_all = TRUE)

mean_trip_tib_2000$pathway <- factor(mean_trip_tib_2000$pathway, levels=c("other_indels", "MMEJ", "NHEJ"))
mean_trip_tib_2000 <- mean_trip_tib_2000[!is.na(mean_trip_tib_2000$ratios), ]

trip_tib$spin_state <- factor(trip_tib$spin_state, levels=c("Speckle",
                                                                      "Interior_Act1",
                                                                      "Interior_Act2",
                                                                      "Interior_Act3",
                                                                      "Interior_Repr1",
                                                                      "Interior_Repr2",
                                                                      "Near_Lm1",
                                                                      "Near_Lm2",
                                                                      "Near_Lm3",
                                                                      "Lamina",
                                                                      NA ))
```

# Main Figure
## Panel 3A
```{r Fig3A efficiency LBR1 and LBR2, fig.height=4, fig.width=5}
barcodes <- trip_tib %>% group_by(barcode) %>% dplyr::summarise(count = n()) %>% filter(count > 1) %>% pull(barcode)
trip_tib <- trip_tib %>% filter(barcode %in% barcodes)
ggplot(trip_tib, aes(efficiency, fill = plasmid, alpha = 0.8)) + 
  geom_density(trim = F) + 
  scale_fill_manual(values = c("LBR1" = "gray", "LBR2" = "black")) + 
  theme_bw(base_size = 16)
  
```

### Figure Legend 

## Panel 3B
```{r Fig3B efficiency LBR2 vs LBR1, fig.height=5, fig.width=5}
trip_tib %>% dplyr::select(barcode, plasmid, efficiency) %>% spread(plasmid, efficiency) %>% ggplot(., aes(LBR2, LBR1))  +  
  geom_point() +
  geom_smooth(method= "loess") + 
  theme_bw(base_size = 16) + 
  stat_cor(method = "spearman", label.x = .25, label.y = .7) 
```

### Figure Legend 

## Panel 3C
```{r Fig3C efficiency LBR2 LBR1 vs chrom, fig.height=9, fig.width=9}
trip_tib <- trip_tib_2000 %>% 
  filter(exp_pool %in% c("indel_LBR2Rep5_LBR1_-_-_-_64", "indel_mean_LBR2_0_64h"))

chrom.mods.cols <- seq(grep("binsize", names(trip_tib))+1, grep("H3K27me3_domain", names(trip_tib))-1)

trip_tib_eff <- trip_tib %>% dplyr::select(efficiency, barcode, chrom.mods.cols, plasmid) %>% spread(plasmid, efficiency) %>% na.omit() %>% mutate(exp_pool = "Efficiency_Corr")

chrom.mods.cols <- seq(grep("barcode", names(trip_tib_eff))+1, grep("LBR1", names(trip_tib_eff))-1)

eff.hist.corr <- CorrelationDFMean(trip_tib_eff, "Efficiency_Corr", c("LBR2"), targets = chrom.mods.cols) %>% 
  arrange(corr) %>% pull(rowname) %>% factor(.,levels = .)


lims = 1

colores <- c("LBR1" = "gray", "LBR2" = "black")

CorrelationPlotMean(trip_tib_eff, "Efficiency_Corr",  c("LBR1", "LBR2") , targets = eff.hist.corr)

```

### Figure Legend 

## Panel 3D
```{r Fig3D heterochrom domains efficiency, fig.height=7, fig.width=12}
tib.dt = data.table(trip_tib)

domain_levels = c('late_replicating', 'LAD', 'H3K9me2', 'H3K9me3', 'H3K27me3')

domains = paste0(domain_levels, '_domain')

indel.dt = unique(tib.dt[exp_pool %in% c("indel_LBR2Rep5_LBR1_-_-_-_64",
                                         "indel_mean_LBR2_0_64h"),
                         c('exp_pool', 'barcode', 'efficiency',
                           domains), with=F])

indel.dt[indel.dt=='iDomain'] <- 0
indel.dt[indel.dt=='Domain'] <- 1

indel.dt[is.na(LAD_domain), LAD_domain:=0]

indel.dt[, (domains) := lapply(.SD, as.numeric), .SDcols = domains]

colnames(indel.dt) = gsub('_domain', '', colnames(indel.dt))

opt_df = get_combinations(domain_levels)
# 
plot_comb_grid(indel.dt[exp_pool=='indel_mean_LBR2_0_64h'], opt_df,
                   domain_levels, 'efficiency', min_count=20)

plot_comb_grid(indel.dt[exp_pool=='indel_LBR2Rep5_LBR1_-_-_-_64'], opt_df,
                   domain_levels, 'efficiency', min_count=20)
```


### Figure Legend 

# Supplementary Figure 3

## Panel S3A
```{r FigS3A heterochrom domains efficiency supp, fig.height=8, fig.width=16}
tib.dt = data.table(trip_tib)

domain_levels = c('late_replicating', 'LAD', 'H3K9me2', 'H3K9me3', 'H3K27me3')

domains = paste0(domain_levels, '_domain')

indel.dt = unique(tib.dt[exp_pool %in% c("indel_LBR2Rep5_LBR1_-_-_-_64",
                                         "indel_mean_LBR2_0_64h"),
                         c('exp_pool', 'barcode', 'efficiency',
                           domains), with=F])

indel.dt[indel.dt=='iDomain'] <- 0
indel.dt[indel.dt=='Domain'] <- 1

indel.dt[is.na(LAD_domain), LAD_domain:=0]

indel.dt[, (domains) := lapply(.SD, as.numeric), .SDcols = domains]

colnames(indel.dt) = gsub('_domain', '', colnames(indel.dt))

opt_df = get_combinations(domain_levels)
# 
plot_comb_grid(indel.dt[exp_pool=='indel_mean_LBR2_0_64h'], opt_df,
                   domain_levels, 'efficiency')

plot_comb_grid(indel.dt[exp_pool=='indel_LBR2Rep5_LBR1_-_-_-_64'], opt_df,
                   domain_levels, 'efficiency')
```

### Figure Legend 

## Panel S3B
```{r FigS3B chip clone 5, fig.height=6, fig.width=6}
Chip.seq.data.melt <- trip_tib_2000 %>% filter(cell_line == "clone5")  %>% distinct(barcode, .keep_all = TRUE) %>% select(barcode, H3K27me3, H3K4me1, H3K9me3,H3K27ac) %>% gather(H3K27me3, H3K4me1, H3K9me3,H3K27ac, key = antibody, value = value)

signal.amp.seq <- left_join(log2.sel.amp.tib,Chip.seq.data.melt, by = c("barcode", "antibody"))

#Plot with free axis scales
ggplot(signal.amp.seq, aes(value,amp.indelpcr.mean)) + geom_point() + geom_smooth(method = "lm") + facet_wrap(~ antibody, scales= "free") + ylab("log2(Input normalized indel PCR reads)") + xlab("log2(ChIPseq mean 2kb bin)") + stat_cor(method = "pearson", label.x.npc = 0, label.y.npc = 1, vjust = 0) + ggtitle("ChIP seq vs. amplicon seq signal in barcode") + theme_bw(base_size = 16)
```

### Figure Legend 

<!-- ### Panel A -->
<!-- ```{r chip values vs domains, fig.height=10, fig.width=7} -->
<!-- domains.dt <- trip_tib %>% distinct(barcode, .keep_all = TRUE) %>% -->
<!--   dplyr::select(barcode,  -->
<!--                 H3K27me3_domain,  -->
<!--                 H3K9me2_domain,  -->
<!--                 H3K9me3_domain,  -->
<!--                 late_replicating_domain,  -->
<!--                 LAD) %>%  -->
<!--   mutate(LAD = ifelse(grepl("iAD", LAD), "iDomain",  -->
<!--                       "Domain")) %>% -->
<!--   gather(H3K27me3_domain,  -->
<!--          H3K9me2_domain,  -->
<!--          H3K9me3_domain,  -->
<!--          late_replicating_domain,  -->
<!--          LAD,  -->
<!--          key = target,  -->
<!--          value = domain) %>%  -->
<!--   mutate(target = gsub("(.*)_domain", "\\1", target)) %>% -->
<!--   mutate(target = gsub("LAD", "LMNB1", target)) %>%  -->
<!--   as.data.table() -->

<!-- chip.dt <- trip_tib %>%  -->
<!--   distinct(barcode, .keep_all = TRUE) %>% -->
<!--   dplyr::select(barcode,  -->
<!--                 H3K27me3,  -->
<!--                 H3K9me2,  -->
<!--                 H3K9me3,   -->
<!--                 LMNB1,  -->
<!--                 late_replicating) %>%  -->
<!--   as.data.table()  -->

<!-- setkey(domains.dt, 'target') -->

<!-- plot_list = lapply(domains.dt[,unique(target)], function(target){ -->
<!-- 	data = cbind(domains.dt[target, 'domain'], -->
<!-- 	             chip.dt[, ..target]) -->
<!-- 	ggplot(data, aes_string(x='domain', y=target, color='domain')) + -->
<!-- 		geom_quasirandom() + -->
<!-- 		theme_bw(base_size = 16) + -->
<!-- 		ggtitle(target) -->
<!-- }) -->

<!-- do.call(grid.arrange, plot_list) -->
<!-- ``` -->

<!-- ### Panel C -->
<!-- ```{r spinstatesefficiceny, fig.height=7, fig.width=12} -->
<!-- spin_color <- trip_tib %>% select(spin_state, spin_color) %>% distinct(spin_state, .keep_all = T) %>% filter(!is.na(spin_state)) -->
<!-- spin_colors <- spin_color$spin_color -->
<!-- names(spin_colors) <- spin_color$spin_state -->
<!-- ggplot(filter(trip_tib, !is.na(spin_state)), aes(spin_state, efficiency, color = spin_state)) +  -->
<!--   geom_quasirandom(size = 3) +  -->
<!--   scale_color_manual(values = spin_colors) +  -->
<!--   theme_bw(base_size = 16) + -->
<!--   facet_wrap(. ~ plasmid) -->
<!-- ``` -->

# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

