---
title: "Figure 5 and supplemental 5"
author: "Ruben Schep"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output: 
  html_document:
    theme: journal #cerulean
    highlight: monochrome
    toc: true
    toc_float: true
    code_folding: show
  editor_options:
    chunk_output_type: console
---

# knitr document van Steensel lab
```{r}
StartTime <-Sys.time()
library(knitr)

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 

## Select outdir
out.dir = paste0("figures/rs", Date, "/")
dir.create(out.dir)
opts_chunk$set(dev=c('png', 'pdf'), fig.path = file.path(out.dir))
pdf.options(useDingbats = FALSE)
```

# Introduction

## Description of Data

# Data processing
## Path, Libraries, Parameters and Useful Functions
```{r setup, message=FALSE, warnings=FALSE}
# INPUT DIRECTORY
in.dir.date = 20200121
in.dir = paste0("/DATA/projects/DSBrepair/data/R/rs", in.dir.date, "/")

# libraries:
library(ggplot2)
library(ggpubr)
# library(reshape2)
# library(tibble)
# library(GenomicRanges)
# library(rtracklayer)
library(corrr)
library(Hmisc)
library(ggbeeswarm)
library(RColorBrewer)
library(data.table)
library(dplyr)
# library(plyr)
library(tidyr)
# library(stringr)
# library(plotly)
library(ggpmisc)
library(glmnet)
library(cowplot)
library(mda)
library(earth)
library(yaml)
library(vip)
library(caret)
library(scales)
library(glmnet)
library(gridExtra)
library(ggcorrplot)
library(bnlearn)
library(pheatmap)
library(ppcor)
library(parallel)

color_redblue <- rev(brewer.pal(11,"RdBu"))
duocolor <- c("#EE756E", "#2BB6BE")
colore <- c("wt" = "#808184", "other" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "SSTR" = "#007A4C")
colores <- c("wt" = "#808184", "other_indels" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "SSTR" = "#007A4C")
colori  <- c("other_indels" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "ssODN" = "#007A4C")
colora <- c("0" = "#808184", "1" = "#E1251B", "-7" = "#223DA0", "-14" = "#223DA0")
coloru <- c("wt" = "#808184", "other" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "MMEJplusNHEJ" = "#EE8A2A", "all" = "black", "MMEJplusNHEJplusHDR" = "#EE8A2A", "HDR" = "#007A4C", "MMEJplusHDR" = "#EE8A2A")
colory <- c("wt" = "#808184", "other" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "MMEJplusNHEJ" = "#EE8A2A", "all" = "black", "SSTR" = "#007A4C")

measure_color = c(NHEJ='#e1251b', MMEJ='#26419a', SSTR='#007A4C', None='grey')

method_color = c(MARS='#00bfc4', LASSO='#f8766d')
lad_color <- c(LAD="#802cdc", iLAD="#dfaf2f")
sign_color <- c(down="#AA6375", up="#7B9A36")

feature_fill = c(insulator="#a6b5a3", repressive="#304441", euchromatin="#cda4cc",
                 transcribing="#ED9114", HDAC="#aac9e0", accessibility='#cebc85',
                 methylation='#7dc98f')

feature_color = c(insulator="black", repressive="white", euchromatin="black",
                  transcribing="black", HDAC="black", accessibility='black',
                  methylation='black')

clustering = data.frame(row.names=c('LMNB1', 'late_replicating', 'H3K9me2',
                                    'H3K9me3', 'H3K27me3', 'EZH2', 'CTCF', 'SMC3',
							        'HDAC3', 'HDAC2', 'HDAC1', 'H3K4me1',
								    'H3K4me2', 'H3K4me3', 'H3K27ac',
								    'H4K5acK8ac', 'H2AFZ', 'DNAse', 'Dam', 'm5C',
								    'H3K79me2', 'TTseq', 'H3K36me3', 'POL2AS2',
								    'POL2'),
					   group=factor(c(rep('repressive', 6), rep('insulator', 2),
				                      rep('HDAC', 3), rep('euchromatin', 6),
                                      rep('accessibility', 2), 'methylation',
						              rep('transcribing', 5)),
									levels=c('transcribing', 'accessibility',
                                             'methylation', 'euchromatin',
								             'HDAC', 'insulator', 'repressive')))

```

## Custom functions
Functions used thoughout this script.
```{r, message=FALSE, warnings=FALSE}
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, substr(gsub("-","",Sys.time()),1,8), "_", filename)
  filename
}

CorrelationDFMean <- function(data, condition_name, feature_name, targets, p.value=0.001){
  hist.corr.list <- data[data$exp_pool==condition_name, ] %>% dplyr::select(feature_name, targets) %>% as.matrix() %>% rcorr(type = "pearson")
  
  histone.corr.df <- data.frame(rowname = names(hist.corr.list$r[1, -1]), 
                                corr = hist.corr.list$r[1, -1], 
                                pvalue = hist.corr.list$P[1, -1])
  
  return(histone.corr.df=histone.corr.df)
}


CorrelationPlotMean <- function(data, condition_name, feature_name, targets, p.value=0.001){
  cor.df = data.frame()
  for (i in feature_name){
    corr.list <- data %>% filter(exp_pool==condition_name) %>% dplyr::select(i, as.character(targets)) %>% as.matrix() %>% rcorr(type = "pearson")
  
    df <- data.frame(rowname = names(corr.list$r[1, -1]), 
                                corr = corr.list$r[1, -1], 
                                pvalue = corr.list$P[1, -1],
                                feature = i)
    cor.df <- rbind(cor.df, df)
  }
  p <- cor.df %>% 
    transform(rowname = factor(rowname, levels = levels(targets)), 
              corr = corr, 
              pvalue = pvalue) %>%
    ggplot(aes(x = rowname, y = corr, fill = feature, alpha = pvalue < p.value)) +
    geom_bar(stat = "identity", position = "dodge") +
    ylab(paste("Pearson correlation with", feature_name)) +
    xlab("Chromatin Feature") + 
    ggtitle(paste("Repair pathways correlation with chromatin features in", condition_name, "exps")) +
    theme_bw(base_size = 16) +
    scale_fill_manual("legend", values = colores) +
    scale_y_continuous(breaks=seq(-1,1, .1)) + 
    coord_flip()
  return(list(dataframe=cor.df, plot = p))
}

bayesian_boot_func <- function(Xy, y_name, blacklist){
	n = nrow(Xy)
	boot.idx <- sample(1 : n, size = n, replace = TRUE)
	Xy.boot <- Xy[boot.idx, ]
	hc.boot = hc(Xy.boot, blacklist=blacklist)
	arc_matrix = arcs(hc.boot)
	selection = arc_matrix[arc_matrix[,2]==y_name,1]
	pcor_list = lapply(selection, function(sel){
		other = selection[selection!=sel]
		pcor.test(Xy.boot[,..sel], Xy.boot[,..y_name],
			      Xy.boot[,..other])$estimate
	})

	result_vec = rep(0,ncol(Xy)-1)
	names(result_vec) = colnames(Xy)[colnames(Xy)!=y_name]
	result_vec[selection] = unlist(pcor_list)
	return(result_vec)
}


bootstrap_bayes <- function(Xy, y_name, blacklist, nboot=1000, ncores=20){
	baye_boot_list = mclapply(1:nboot, function(i){
		bayesian_boot_func(Xy, y_name, blacklist)
	}, mc.cores=ncores)
	baye_boot.df = do.call(cbind, baye_boot_list)
	sel_baye = rowSums(baye_boot.df!=0) / nboot

	z_vec = apply(baye_boot.df,1, function(x){mean(x)/sd(x)})

	med_vec = apply(baye_boot.df, 1, function(x){
		median(x[x!=0])
	})
	pos_ratio_vec = apply(baye_boot.df, 1, function(x){
		length(which(x>0))/length(which(x!=0))
	})

	baye.dt = data.table(var = names(sel_baye), importance=sel_baye, z=z_vec,
                         median=med_vec, pos_ratio=pos_ratio_vec)

	return(baye.dt)
}



bootstrap_pcorsign_func <- function(X, y, all){
	n = nrow(X)
	boot.idx <- sample(1 : n, size = n, replace = TRUE)
	X.boot <- X[boot.idx, ]
	y.boot <- y[boot.idx]
	fit <- earth(x = X.boot, y = y.boot)
	nterms = length(fit$selected.terms)

	mm = data.frame(y = y.boot,
		            model.matrix(fit, X.boot, which.terms=1:nrow(fit$dirs))[,-1])

	coef_matrix = apply(fit$prune.terms[2:nterms, ], 1, function(x){
		xnn = x[x>0]
		dirs = fit$dirs[xnn, ,drop=F]
        count = colSums(dirs>0)
        selection = names(count)[count>0]
        if (length(selection) > 1){
            pcor_list = lapply(selection, function(sel){
        		other = selection[selection!=sel]
        		pcor.test(X.boot[,..sel], y.boot,
        			      X.boot[,..other])$estimate
        	})
        } else if (length(selection) == 1){
            pcor_list = list(cor(unlist(X.boot[,..selection]), y.boot))
        }
		result = rep(0, ncol(X))
		names(result) = colnames(X)
        if (length(selection) > 0){
            result[selection] = unlist(pcor_list)
        }
		return(result)
	})
	if (all){
		imp = evimp(fit)[,'nsubsets']
		names(imp) = gsub('-unused', '', names(imp))
		imp_vec = rep(0,ncol(X))
		names(imp_vec) = colnames(X)
		imp_vec[names(imp)] = imp / nterms
	} else {
		imp_vec = rowSums(coef_matrix!=0) / nterms
	}

	# result[i_vec] = result[i_vec]/2
	return(list(importance=imp_vec, coef=coef_matrix))
}



bootstrap_mars_pcor <-function(X, y, nboot=1000, all=T, ncores=20){
	mars_boot_list = mclapply(1:nboot, function(i){
		b_lasso = bootstrap_pcorsign_func(X, y, all)
	}, mc.cores=ncores)

	mars_boot.df = do.call(cbind, lapply(mars_boot_list,
		                                 function(x){x[['importance']]}))
	sel_mars = rowSums(mars_boot.df) / nboot

	coef_matrix = do.call(cbind, lapply(mars_boot_list,
		                                function(x){x[['coef']]}))

	z_vec = apply(coef_matrix, 1, function(x){
		mean(x) / sd(x)
	})
	med_vec = apply(coef_matrix, 1, function(x){
		median(x[x!=0])
	})
	pos_ratio_vec = apply(coef_matrix, 1, function(x){
		length(which(x>0))/length(which(x!=0))
	})
	mars.dt = data.table(var = names(sel_mars), importance=sel_mars,
	                     z=z_vec[names(sel_mars)], median=med_vec,
						 pos_ratio=pos_ratio_vec)

	return(mars.dt)
}




plot_MARS_bayes <- function(left_set, right_set, none_label='ambiguous',
                            top=NULL, left_title='Bayesian Network',
                            right_title='MARS'){

	color_labels = left_set[,levels(color)]
	color_labels[color_labels == 'None'] = none_label

	g_legend<-function(a.gplot){
	  tmp <- ggplot_gtable(ggplot_build(a.gplot))
	  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
	  legend <- tmp$grobs[[leg]]
	  return(legend)
	}
	break_vec = seq(0,1,0.2)
	minor_break_vec = seq(0.1,0.9,0.2)

	g_left <- ggplot(data = left_set, aes(x = target, y = importance, fill=color)) +
	  geom_bar(stat = "identity") +
	  theme_bw() +
	  ggtitle(left_title) +
	  theme(axis.title.x = element_blank(),
	        axis.title.y = element_blank(),
	        axis.text.y = element_blank(),
	        axis.ticks.y = element_blank(),
	        plot.margin = unit(c(1,0,1,0), "mm"),
		    plot.title = element_text(hjust = 1)) +
	  scale_fill_manual(values=measure_color,
	                    labels=color_labels) +
	  scale_y_reverse(limits=c(1.1,0), expand=c(0,0),
	                  breaks=break_vec,
				      minor_breaks=minor_break_vec) +
	  coord_flip()


	g_right <- ggplot(right_set, aes(x = target, y = importance, fill=color)) +
	  geom_bar(stat = "identity") +
	  theme_bw() +
	  ggtitle(right_title) +
	  scale_fill_manual(values=c(measure_color, None='gray'),
	        		    labels=color_labels) +
	  theme(axis.title.x = element_blank(),
	        axis.title.y = element_blank(),
	        axis.text.y = element_blank(),
			axis.ticks.y = element_blank(),
	        plot.margin = unit(c(1,0,1,-1), "mm")) +
	  scale_y_continuous(limits=c(0,1.1),
		                 expand=c(0,0),
	                     breaks=break_vec,
				         minor_breaks=minor_break_vec) +
	  coord_flip() +
	  guides(fill=F)

	g_mid<-ggplot(left_set, aes(x=1,y=target, label=target, fill=feature_group,
		          color=feature_group, yend=target)) +
	  geom_label(show.legend = F) +
	  geom_segment(aes(x=0.94,xend=0.945), color='black')+
	  geom_segment(aes(x=1.056,xend=1.061), color='black')+
	  ggtitle("")+
	  ylab(NULL)+
	  scale_color_manual(values=feature_color) +
	  scale_fill_manual(values=feature_fill) +
	  scale_x_continuous(expand=c(0,0),limits=c(0.94,1.065)) +
	  theme(axis.title=element_blank(),
	        panel.grid=element_blank(),
	        axis.text.y=element_blank(),
	        axis.ticks.y=element_blank(),
	        panel.background=element_blank(),
	        axis.text.x=element_text(color=NA),
	        axis.ticks.x=element_line(color=NA),
	        plot.margin = unit(c(1,-1,1,-1), "mm"))

	gg_left <- ggplot_gtable(ggplot_build(g_left + guides(fill=F)))
	gg_right <- ggplot_gtable(ggplot_build(g_right))
	gg_mid <- ggplot_gtable(ggplot_build(g_mid + guides(fill=F, color=F)))

	measure_legend <- g_legend(g_left + guides(fill=guide_legend(title="preferred\npathway")))
	feature_legend <- g_legend(g_mid +
		                       geom_bar(stat='identity') +
							   guides(fill=guide_legend(title="feature group"),
	                                  color=F))

	layout = rbind(c(rep(1,3),2,rep(3,3),4),
	               c(rep(1,3),2,rep(3,3),5))

	return(grid.arrange(gg_left,gg_mid, gg_right, measure_legend, feature_legend,
		                layout_matrix=layout, top=top))
}

cor_test <- function(y, X, left, right){
    cor_list = lapply(X, function(x){
        cor.test(x,y)
    })
    p_vec = unlist(lapply(cor_list, function(x){x$p.value}))
    cor_vec = unlist(lapply(cor_list, function(x){x$estimate}))
    p_adj = p.adjust(p_vec)
    color = ifelse(p_adj < 0.05, ifelse(cor_vec > 0, left, right),
                   'None')
    importance = abs(cor_vec)
    target=factor(colnames(X), levels=rownames(clustering))
    feature_group = clustering[colnames(X), ]
    return(list(correlation=cor_vec,importance = importance, p=p_vec,
                p_adj=p_adj, target=target, color=color,
                feature_group=feature_group))
}


get_combinations <- function(domain_vec){
    n_domain = length(domain_vec)
    get_ivec <-function (last_x, n_domain){
    	if (last_x < n_domain){
    		i_vec = c(0, c((last_x+1):n_domain))
    	} else{
    		i_vec = c(0)
    	}
    }

    combine <- function(x_new, x_vec, n_domain){
        new_vec = c(x_vec, x_new)
        if (length(new_vec)==n_domain){
            return(new_vec)
        } else {
            i_vec = get_ivec(max(new_vec), n_domain)
            x_list = lapply(i_vec, combine, x_vec=new_vec,
                            n_domain=n_domain)
            return(do.call(rbind, x_list))
        }
    }

    comb_list = lapply(c(0:n_domain), combine, x_vec=c(), n_domain=n_domain)

    comb = do.call(rbind, comb_list)
    comb_df = unique(t(apply(comb, 1, function(x){x[order(x)]})))

    name_list = apply(comb_df, 1, function(x){
    	name = paste(domain_vec[x], collapse='-')
    	if (name == ''){
    		name = 'iDomain'
    	}
    	return(name)
    })

    opt_list = lapply(domain_vec, function(d){
    	 apply(comb_df, 1, function(x){
    		opt = domain_vec[x]
    		return(d%in%opt)
    	})
    })

    opt_df = do.call(cbind, opt_list)
    colnames(opt_df) = domain_vec
    rownames(opt_df) = name_list

    return(opt_df)
}

plot_comb_grid <- function(indel.dt, opt_df, domains, y_name, min_count=0, max_count=Inf){

    opt_melt = reshape2::melt(opt_df)
    colnames(opt_melt) = c('group', 'domain', 'is_in')

    opt_melt$is_in = ifelse(opt_melt$is_in, 'in', 'out')
    opt_melt$domain = factor(opt_melt$domain, levels=domains)


    group_list = lapply(rownames(opt_df), function(opt_name){
    	opt = opt_df[opt_name, ]
    	in_vec = domains[opt]
    	out_vec = domains[!opt]

    	in_sum = rowSums(indel.dt[,..in_vec])
    	out_sum = rowSums(indel.dt[,..out_vec])
    	if (length(in_vec)==0){
    		in_group = which(out_sum == 0)
    	} else if (length(out_vec) == 0){
    		in_group = which(in_sum==length(in_vec))
    	} else  {
    		in_group = which(in_sum == length(in_vec) & out_sum==0)
    	}
    	return(in_group)
    })

    names(group_list) = rownames(opt_df)
    for (opt_name in names(group_list)){
    	i_vec = group_list[[opt_name]]
    	indel.dt[i_vec, group := opt_name]
    }
    group_count_df = data.frame(indel.dt[,table(group)])
    group_count_df = group_count_df[order(match(group_count_df$group, rownames(opt_df))),]
    group_count_sel = group_count_df[group_count_df$Freq>=min_count,]
    group_count_sel = group_count_sel[group_count_sel$Freq<=max_count,]
    group_levels =  group_count_sel$group
    opt_sel = opt_melt[opt_melt$group %in% group_levels, ]

    opt_sel$group = factor(opt_sel$group, levels=group_levels)

    indel_selection = indel.dt[group%in%group_levels, ]
    indel_selection[,group:=factor(group, levels=group_levels)]

    # wilcox_list = lapply(group_levels[group_levels!='iDomain'],
    #                      function(g){
    #                          x = indel_selection[group==g, y_name, with=F]
    #                          y = indel_selection[group=='iDomain', y_name,
    #                                              with=F]
    #                          wilcox.test(unlist(x),unlist(y))
    #                      })

    group_count_sel$group = factor(group_count_sel$group, levels=group_levels)

    count_plot = ggplot(group_count_sel, aes(x=group,y=1, label=Freq)) +
                            geom_label() +
                            theme_void()



    dots = ggplot(opt_sel, aes(x=group, y=domain, color=is_in)) +
    	geom_point(size=2) +
    	geom_line(data=opt_sel[opt_sel$is_in=='in',], aes(group=group)) +
    	theme_bw() +
    	scale_color_manual(values=c('in'='black', 'out'='grey70')) +
    	guides(color=FALSE)  +
    	geom_vline(xintercept=seq(1.5, length(unique(indel.dt$group))-0.5, 1),
    			   lwd=0.5, colour="grey90") +
    	theme(axis.title=element_blank(),
    		  axis.text=element_blank(),
    		  axis.ticks=element_blank(),
    		  panel.grid.major = element_blank())

    symnum = list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1),
                  symbols = c("****", "***", "**", "*", ""))

    beeswarm = ggplot(indel_selection, aes_string(x='group', y=y_name)) +
    			  geom_quasirandom(width=0.3) +
    			  theme_bw() +
    			  geom_vline(xintercept=seq(1.5, length(unique(indel.dt$group))-0.5, 1),
    			             lwd=0.5, colour="grey90") +
                  stat_summary(fun.y=median, fun.ymin = median,
                               fun.ymax = median, geom="crossbar", width = 0.5,
                               color='red') +
                  stat_compare_means(aes(label = ..p.signif..),
                                     symnum.args = symnum,
                                     ref.group='iDomain') +
    			  theme(axis.title=element_blank(),
    			        axis.text.x=element_blank(),
      				    panel.grid.major.x = element_blank(),
    				    axis.ticks.x=element_blank())

    count = data.frame(domain=gsub('_domain', '', domains),
    	               count=colSums(indel.dt[,..domains], na.rm=T))

    count$domain = factor(count$domain, levels=domain_levels)


    histogram = ggplot(count, aes(x=1, y=domain, label = count)) +
    				geom_label() +
    				theme_bw() +
    				theme(plot.background = element_blank(),
        				  panel.grid.major = element_blank(),
        				  panel.grid.minor = element_blank(),
        				  panel.border = element_blank(),
    					  axis.title=element_blank(),
    	  			      axis.text.x=element_blank(),
    	  				  axis.ticks=element_blank())

    empty = ggplot(data=data.frame(1,1),aes(x=1,y=1)) +
                geom_point() +
                theme_void()

    count_beeswarm = plot_grid(beeswarm, count_plot, ncol=1, rel_heights=c(20,1),
                               align='v', axis='lr')
    return(plot_grid(empty, beeswarm, empty, count_plot,
                     histogram, dots, nrow=3,
                     rel_widths=c(1,10), rel_heights=c(15,1,5), align='vh',
                     axis='tlbr'))
}


```

# Data import
```{r, message=FALSE, warnings=FALSE}
load(paste0(in.dir, "RSTP2_Indel_Chromatin_2kb.RData"))
load(paste0(in.dir, "RSTP2_IndelRatios_Chromatin_2kb.RData"))
```


## Data loading

```{r preprocess experiments}
trip_tib_mut <- copy(trip_tib_2000)
trip_tib_2000 <- trip_tib_2000 %>% distinct(barcode, exp, .keep_all = TRUE)

trip_tib <- trip_tib_2000 %>% 
  filter(exp_pool  %in% c("indel_mean_LBR2_DMSO_64h", "indel_mean_LBR2_NU7441_64h", "indel_mean_LBR2_ssODN_DMSO_64h", "indel_mean_LBR2_ssODN_NU7441_64h"))

trip_tib_mut <- trip_tib_mut_2000 %>%
  filter(exp_pool %in% c("indel_mean_LBR2_DMSO_64h", "indel_mean_LBR2_NU7441_64h", "indel_mean_LBR2_ssODN_DMSO_64h", "indel_mean_LBR2_ssODN_NU7441_64h"))

mean_trip_tib_2000 <- trip_tib_2000 %>% 
  gather(NHEJ, MMEJ, other_indels, key = "pathway", value = 'ratios') %>% 
  distinct(barcode, exp, pathway, .keep_all = TRUE)

mean_trip_tib_2000$pathway <- factor(mean_trip_tib_2000$pathway, levels=c("other_indels", "MMEJ", "NHEJ"))
mean_trip_tib_2000 <- mean_trip_tib_2000[!is.na(mean_trip_tib_2000$ratios), ]
```

# Main Figure 5
## Panel 5A

Panel A is the scheme of the ssODN. That has been made. 
Add loading here I guess. 

## Panel 5B
```{r Fig5B indel distribution ssODN, fig.height=4, fig.width=7}
ggplot(filter(trip_tib_mut, drug == "DMSO", ssODN == "ssODN"), 
       aes(mutation, freq, color = color)) + 
  geom_quasirandom() + 
  theme_bw(base_size = 16) + 
  theme(legend.position = c(0.2, 0.6)) +
  scale_color_manual("legend", values = colory)

trip_tib_mut %>% filter(drug == "DMSO", ssODN == "ssODN") %>% group_by(mutation) %>% dplyr::summarise(median = median(freq)) %>% arrange(desc(median))
```
###  Figure legend

### Panel 5C
```{r Fig5C pathways vs contribution ssODN, fig.height=6, fig.width=5}
trip_tib_pathways <- trip_tib %>% 
  gather(NHEJ, MMEJ, other_indels, SSTR, key = "pathway", value = 'ratios') %>% 
  distinct(barcode, exp, pathway, .keep_all = TRUE)

exp_order <- c("indel_mean_LBR2_DMSO_64h", 
               "indel_mean_LBR2_ssODN_DMSO_64h",
                "indel_mean_LBR2_NU7441_64h", 
               "indel_mean_LBR2_ssODN_NU7441_64h")

trip_tib_pathways$pathway <- factor(trip_tib_pathways$pathway, levels=c("other_indels", "MMEJ", "SSTR", "NHEJ"))
trip_tib_pathways$exp_pool <- factor(trip_tib_pathways$exp_pool, levels=exp_order)
trip_tib_pathways <- trip_tib_pathways[!is.na(trip_tib_pathways$ratios), ]

pathway_contribution <- trip_tib_pathways %>% group_by(exp_pool, pathway) %>% dplyr::summarise(pathway_contr = mean(ratios))

colores <- c("other_indels" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "SSTR" = "#007A4C")

p <- ggplot(pathway_contribution, aes(exp_pool, pathway_contr, fill = pathway, order = pathway)) + 
  geom_bar(stat = "identity", position = "fill")  + 
  theme_bw(base_size = 16)
p + scale_fill_manual("legend", values = colores) + 
  scale_x_discrete(labels=c("indel_mean_LBR2_DMSO_64h" = "DMSO",
                            "indel_mean_LBR2_ssODN_DMSO_64h" = "ssODN + DMSO",
                            "indel_mean_LBR2_NU7441_64h" = "NU7441",
                            "indel_mean_LBR2_ssODN_NU7441_64h" = "ssODN + NU7441")) +
  labs(x = element_blank(),
       y = "abundance")


```

### Figure legend 

<!-- ### Panel C2 -->
<!-- ```{r} -->
<!-- ###### OR ###### -->

<!-- p1 <- trip_tib %>%  -->
<!--    filter(ssODN == "-") %>%  -->
<!--   gather(other_indels, NHEJ, MMEJ, -->
<!--                     key = "pathway",  -->
<!--                     value = 'ratios') %>%  -->
<!--   ggplot(., aes(ratios, fill = pathway, alpha = 0.8)) +  -->
<!--   geom_density() +  -->
<!--   theme_bw(base_size = 16) + -->
<!--   scale_fill_manual("legend", values = colores) -->

<!-- p2 <- trip_tib %>%  -->
<!--   filter(ssODN == "ssODN") %>%  -->
<!--   gather(other_indels, NHEJ, MMEJ, SSTR, -->
<!--                     key = "pathway",  -->
<!--                     value = 'ratios') %>%  -->
<!--   ggplot(., aes(ratios, fill = pathway, alpha = 0.8)) +  -->
<!--   geom_density() +  -->
<!--   theme_bw(base_size = 16) + -->
<!--   scale_fill_manual("legend", values = colores) -->

<!-- p3 <- trip_tib %>%  -->
<!--   filter(drug == "DMSO") %>%  -->
<!--   gather(other_indels, NHEJ, MMEJ, SSTR, -->
<!--                     key = "pathway",  -->
<!--                     value = 'ratios') %>%  -->
<!--   ggplot(., aes(ratios, fill = pathway, alpha = 0.8)) +  -->
<!--   geom_density() +  -->
<!--   theme_bw(base_size = 16) + -->
<!--   scale_fill_manual("legend", values = colores) -->

<!-- p1 + facet_wrap(. ~ drug) -->
<!-- p2 + facet_wrap(. ~ drug) -->
<!-- ``` -->

## Panel 5D
```{r Fig5D scatters MMEJ ssODN vs MMEJ -, fig.height=5, fig.width=5}
formula <- y ~ x

trip_lbr2_ssODN_MMEJ <- trip_tib %>% filter(drug == "DMSO") %>% dplyr::select(barcode, ssODN, MMEJ, late_replicating) %>% spread(ssODN, MMEJ) 

ggplot(trip_lbr2_ssODN_MMEJ, aes(`-`, ssODN)) + 
  geom_point() + 
  theme_bw(base_size = 16) + 
  geom_abline(slope = 1)  + 
  geom_smooth(method = glm) +
  xlim(0,.45) + ylim(0,.45)  + 
  labs(title = "MMEJ - vs MMEJ + ssODN in DMSO condition", 
       x = "MMEJ -", 
       y = "MMEJ ssODN") +
  stat_poly_eq(aes(label = paste(..rr.label..)), 
       label.x.npc = "right", label.y.npc = 0.15,
       formula = formula, parse = TRUE, size = 5)+ 
  stat_fit_glance(method = 'lm',
        method.args = list(formula = formula),
        aes(label = paste("P-value = ", signif(..p.value.., digits = 4), sep = "")),
        size = 5) +
  stat_cor(method = "pearson", label.x = 0, label.y = .3)
```
###  Figure legend

## Panel 5E
```{r Fig5E MMEJ vs SSTR, fig.height=5, fig.width=5}
trip_tib %>%
  filter(drug == "DMSO", ssODN == "ssODN") %>%
  ggplot(., aes(MMEJ, SSTR)) + 
  geom_point() + 
  theme_bw(base_size = 16) + 
  geom_smooth(method = glm) +
  xlim(0,.35) + ylim(0,.35)  + 
  labs(title = "MMEJ vs SSTR in DMSO condition", 
       x = "MMEJ", 
       y = "SSTR") +
  stat_poly_eq(aes(label = paste(..rr.label..)), 
       label.x.npc = "right", label.y.npc = 0.15,
       formula = formula, parse = TRUE, size = 5)+ 
  stat_fit_glance(method = 'lm',
        method.args = list(formula = formula),
        aes(label = paste("P-value = ", signif(..p.value.., digits = 4), sep = "")),
        size = 5) +
  stat_cor(method = "pearson")
```
### Figure legend

## Panel 5F
```{r Fig5F pathways vs chrom ssODN DMSO, fig.height=8, fig.width=9}
chrom.mods.cols <- seq(grep("binsize", names(trip_tib))+1, grep("H3K27me3_domain", names(trip_tib))-1)

SSTR.hist.corr <- CorrelationDFMean(trip_tib, "indel_mean_LBR2_ssODN_DMSO_64h", "NHEJ", targets = chrom.mods.cols) %>% 
  arrange(corr) %>% pull(rowname) %>% factor(.,levels = .)
 
lims = 1
CorrelationPlotMean(trip_tib, "indel_mean_LBR2_ssODN_DMSO_64h", c("NHEJ", "MMEJ", "SSTR", "other_indels") , targets = SSTR.hist.corr)
```

### Figure legend

## Panel 5G
```{r Fig5G SSTR ratio per heterochomatin domain, fig.height=6, fig.width=10}
tib.dt = data.table(trip_tib)

domain_levels = c('late_replicating', 'LAD', 'H3K9me2', 'H3K9me3', 'H3K27me3')

domains = paste0(domain_levels, '_domain')

indel.dt = unique(tib.dt[exp_pool == "indel_mean_LBR2_ssODN_DMSO_64h",  
                         c('exp_pool', 'barcode', 'SSTR', 'SSTR_MMEJSSTR',
                           domains), with=F])

indel.dt[indel.dt=='iDomain'] <- 0
indel.dt[indel.dt=='Domain'] <- 1

indel.dt[is.na(LAD_domain), LAD_domain:=0]

indel.dt[, (domains) := lapply(.SD, as.numeric), .SDcols = domains]

colnames(indel.dt) = gsub('_domain', '', colnames(indel.dt))

opt_df = get_combinations(domain_levels)


# Ratio of SSTR vs all indels
plot_comb_grid(indel.dt[exp_pool=='indel_mean_LBR2_ssODN_DMSO_64h'], opt_df,
                   domain_levels, 'SSTR', min_count = 20)

```

### Figure legend

## Panel 5H
```{r Fig5H SSTR_MMEJ ratio per heterochomatin domain, fig.height=6, fig.width=10}

# Ratio of SSTR vs MMEJ
plot_comb_grid(indel.dt[exp_pool=='indel_mean_LBR2_ssODN_DMSO_64h'], opt_df,
                   domain_levels, 'SSTR_MMEJSSTR', min_count = 20)
```


# Supplementary Figure 5

## Panel S5A
```{r FigS5A scatter NHEJ with and without ssODN, fig.height=5, fig.width=5}
trip_lbr2_ssODN_NHEJ <- trip_tib %>% filter(drug == "DMSO") %>% dplyr::select(barcode, ssODN, NHEJ, late_replicating) %>% spread(ssODN, NHEJ) 

ggplot(trip_lbr2_ssODN_NHEJ, aes(`-`, ssODN)) + 
  geom_point() + 
  theme_bw(base_size = 16) + 
  geom_abline(slope = 1)  + 
  geom_smooth(method = glm) +
  xlim(.3, .82) + ylim(.3, .82)  + 
  labs(title = "NHEJ - vs NHEJ + ssODN in DMSO condition", 
       x = "NHEJ -", 
       y = "NHEJ ssODN") +
  stat_poly_eq(aes(label = paste(..rr.label..)), 
       label.x.npc = "right", label.y.npc = 0.15,
       formula = formula, parse = TRUE, size = 5)+ 
  stat_fit_glance(method = 'lm',
        method.args = list(formula = formula),
        aes(label = paste("P-value = ", signif(..p.value.., digits = 4), sep = "")),
        size = 5) +
  stat_cor(method = "pearson")
```

### Figure legend

<!-- ### Panel B -->
<!-- ```{r seperate replicates late replicatung, fig.height=12, fig.width=12} -->
<!-- trip_tib_2000 %>%  -->
<!--   filter(drug == "DMSO", ssODN == "ssODN", plasmid == "LBR2", time == 64, siRNA == "-") %>%  -->
<!--   dplyr::select(barcode, exp_pool, MMEJ, NHEJ, other_indels, SSTR, late_replicating, H3K4me1) %>%  -->
<!--   group_by(barcode, exp_pool) %>%  -->
<!--   gather(MMEJ, NHEJ, other_indels, SSTR, key = "pathway", value = "ratio") %>% -->
<!--   gather(late_replicating, H3K4me1, key = "feature", value = "value") %>%  -->
<!--     filter(feature == "late_replicating") %>% -->
<!--   ggplot(., aes(ratio, value))  +   -->
<!--   geom_point() + -->
<!--   geom_smooth(method= "lm") +  -->
<!--   theme_bw(base_size = 16) +  -->
<!--   stat_cor(method = "pearson", label.x = .25, label.y = .2) + -->
<!--   facet_wrap(pathway ~ exp_pool, ncol = 4, scales = "free_x") -->
<!-- ``` -->

## Panel S5B
```{r FigS5B scatter other with and without ssODN, fig.height=5, fig.width=5}
trip_lbr2_ssODN_other <- trip_tib %>% filter(drug == "DMSO") %>% dplyr::select(barcode, ssODN, other_indels, late_replicating) %>% spread(ssODN, other_indels) 

ggplot(trip_lbr2_ssODN_other, aes(`-`, ssODN)) + 
  geom_point() + 
  theme_bw(base_size = 16) + 
  geom_abline(slope = 1)  + 
  geom_smooth(method = glm) +
  xlim(0.1, .4) + ylim(0.1, .4)  + 
  labs(title = "other - vs other + ssODN in DMSO condition", 
       x = "other ssODN -", 
       y = "other ssODN +") +
  stat_poly_eq(aes(label = paste(..rr.label..)), 
       label.x.npc = "right", label.y.npc = 0.15,
       formula = formula, parse = TRUE, size = 5)+ 
  stat_fit_glance(method = 'lm',
        method.args = list(formula = formula),
        aes(label = paste("P-value = ", signif(..p.value.., digits = 4), sep = "")),
        size = 5) +
  stat_cor(method = "pearson")
```


## Panel S5C
```{r FigS5C scatterplot pathways vs lmnb1 and K4me1 ssODN, fig.height=12, fig.width=6}
trip_tib %>% 
  filter(drug == "DMSO", ssODN == "ssODN") %>% 
  dplyr::select(barcode, MMEJ, NHEJ, other_indels, SSTR, late_replicating, H3K4me1) %>% 
  group_by(barcode) %>% 
  gather(MMEJ, NHEJ, other_indels, SSTR, key = "pathway", value = "ratio") %>%
  gather(late_replicating, H3K4me1, key = "feature", value = "value") %>%
  ggplot(., aes(ratio, value))  +  
  geom_point() +
  geom_smooth(method= "lm") + 
  theme_bw(base_size = 16) + 
  stat_cor(method = "pearson") +
  facet_wrap(pathway ~ feature, scales = "free", ncol = 2)
```
### Figure legend

## Panel S5D
```{r FigS5D mars bayes sstr, fig.height=12, fig.width=12}
chip.dt = readRDS('/DATA/projects/DSBrepair/data/R/cl20191217_ChIP_mean_selection.RDS')
indel.dt = data.table(trip_tib_2000)

indel_selection_64.dt = unique(indel.dt[exp=='indel_mean_LBR2_ssODN_DMSO_64h',
			                         	c('barcode', 'SSTR_MMEJSSTR')])

indel_chip.dt = merge(indel_selection_64.dt, chip.dt,
                      by.x='barcode', by.y='ID')
xy2000_64 = indel_chip.dt[, -c('barcode', 'binsize', 'pool')]

cor_vec = xy2000_64[,cor(SSTR_MMEJSSTR, .SD[,-c('SSTR_MMEJSSTR')])][1,]

cor_set_64 = xy2000_64[,cor_test(SSTR_MMEJSSTR, .SD[,-c('SSTR_MMEJSSTR')],
                                 'SSTR', 'MMEJ')]


blacklist = data.frame(from='SSTR_MMEJSSTR',
                       to=grep('SSTR_MMEJSSTR', colnames(xy2000_64),invert=T, value=T))

boot = boot.strength(xy2000_64, algorithm='hc', R=200,
                     algorithm.args=list(blacklist=blacklist))

bayes_set_64 = bootstrap_bayes(xy2000_64, 'SSTR_MMEJSSTR', blacklist, 1000)
bayes_set_64[,color:=ifelse(pos_ratio>0.7, 'SSTR',
                        ifelse(pos_ratio < 0.3, 'MMEJ', 'None'))]
bayes_set_64[,feature_group := clustering[var,]]
bayes_set_64[,target:=factor(var, levels=rownames(clustering))]
bayes_set_64[,color:=factor(color, levels=c('SSTR', 'MMEJ', 'None'))]

x2000_64 = xy2000_64[, -c('SSTR_MMEJSSTR')]
y2000_64 = xy2000_64[, c('SSTR_MMEJSSTR')]


# full_set = bootstrap_mars(x2000_64, y2000_64$SSTR_MMEJSSTR, nboot=200)
# full_set[,sign:=sign_vec[var]]
# full_set[,color:=ifelse(sign>0, 'up', 'down')]
# full_set[,signed_importance:=importance * sign]
# full_set[,target:=factor(var, levels=var[order(importance, decreasing=F)])]
MARS_set_64 = bootstrap_mars_pcor(x2000_64, y2000_64$SSTR_MMEJSSTR, nboot=1000)
MARS_set_64[,color:=ifelse(pos_ratio>0.7, 'SSTR',
                        ifelse(pos_ratio < 0.3, 'MMEJ', 'None'))]
MARS_set_64[,feature_group := clustering[var,]]
MARS_set_64[,target:=factor(var, levels=rownames(clustering))]
MARS_set_64[,color:=factor(color, levels=c('SSTR', 'MMEJ', 'None'))]



bayes_set_64[,method:='bayes']
MARS_set_64[,method:='MARS']

full_set_64 = rbind(bayes_set_64, MARS_set_64)

full_set_64[,signed_importance:=ifelse(method=='bayes', -importance, importance)]

plot_MARS_bayes(bayes_set_64, MARS_set_64, none_label = 'ambiguous',
	            top='SSTR / (MMEJ + SSTR) 64h feature importance')
```

### Figure legend

## Panel S5E
```{r FigS5E SSTR heterochrom domains, fig.height=10, fig.width=12}
tib.dt = data.table(trip_tib)

domain_levels = c('late_replicating', 'LAD', 'H3K9me2', 'H3K9me3', 'H3K27me3')

domains = paste0(domain_levels, '_domain')

indel.dt = unique(tib.dt[exp_pool == "indel_mean_LBR2_ssODN_DMSO_64h",  
                         c('exp_pool', 'barcode', 'SSTR', 'SSTR_MMEJSSTR',
                           domains), with=F])

indel.dt[indel.dt=='iDomain'] <- 0
indel.dt[indel.dt=='Domain'] <- 1

indel.dt[is.na(LAD_domain), LAD_domain:=0]

indel.dt[, (domains) := lapply(.SD, as.numeric), .SDcols = domains]

colnames(indel.dt) = gsub('_domain', '', colnames(indel.dt))

opt_df = get_combinations(domain_levels)

# Ratio of SSTR vs all indels
plot_comb_grid(indel.dt[exp_pool=='indel_mean_LBR2_ssODN_DMSO_64h'], opt_df,
                   domain_levels, 'SSTR')

```

### Figure legend

## Panel S5F
```{r FigS5F SSTR_MMEJ heterochrom domains, fig.height=10, fig.width=12}
# Ratio of SSTR vs MMEJ
plot_comb_grid(indel.dt[exp_pool=='indel_mean_LBR2_ssODN_DMSO_64h'], opt_df,
                   domain_levels, 'SSTR_MMEJSSTR')
```

### Figure legend


# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

