---
title: "Figure 2 and supplemental 1"
author: "Ruben Schep"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    theme: journal #cerulean
    highlight: monochrome
    toc: true
    toc_float: true
    code_folding: show
  editor_options:
    chunk_output_type: console
---

# knitr document van Steensel lab
```{r}
StartTime <-Sys.time()
library(knitr)
opts_chunk$set(dev=c('png', 'pdf'), fig.path = file.path("figures/"))
pdf.options(useDingbats = FALSE)
```

# Introduction

## Description of Data

# Data processing

## Path, Libraries, Parameters and Useful Functions

```{r setup, message=FALSE, warnings=FALSE}
# 6-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),3,8)

# libraries:
library(ggplot2)
library(ggpubr)
library(reshape)
library(reshape2)
# library(tibble)
# library(GenomicRanges)
# library(rtracklayer)
# library(corrr)
# library(Hmisc)
library(ggbeeswarm)
library(RColorBrewer)
library(data.table)
library(dplyr)
# library(plyr)
library(tidyr)
# library(stringr)
# library(plotly)
library(ggpmisc)
library(pheatmap)



color_redblue <- rev(brewer.pal(11,"RdBu"))
duocolor <- c("#EE756E", "#2BB6BE")
colore <- c("wt" = "#808184", "other" = "black", "NHEJ" = "#e1251b", "MMEJ" = "#26419a", "SSTR" = "#007A4C")
colores <- c("wt" = "#808184", "other_indels" = "black", "NHEJ" = "#e1251b", "MMEJ" = "#26419a", "SSTR" = "#007A4C")
colori  <- c("other_indels" = "black", "NHEJ" = "#e1251b", "MMEJ" = "#26419a", "SSTR" = "#007A4C")
colora <- c("0" = "#808184", "1" = "#e1251b", "-7" = "#26419a", "-14" = "#26419a")
coloru <- c("wt" = "#808184", "other" = "black", "NHEJ" = "#e1251b", "MMEJ" = "#26419a", "MMEJplusNHEJ" = "#EE8A2A", "all" = "black", "MMEJplusNHEJplusHDR" = "#EE8A2A", "HDR" = "#007A4C", "MMEJplusHDR" = "#EE8A2A")
colory <- c("wt" = "#808184", "other" = "black", "NHEJ" = "#e1251b", "MMEJ" = "#26419a", "MMEJplusNHEJ" = "#EE8A2A", "all" = "black", "SSTR" = "#007A4C")
colorys <- c("wt" = "#808184", "other" = "black", "NHEJ" = "#e1251b", "MMEJ" = "#26419a", "MMEJplusNHEJ" = "#EE8A2A", "all" = "black")
```

### Custom functions
Functions used thoughout this script.
```{r, message=FALSE, warnings=FALSE}
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, substr(gsub("-","",Sys.time()),1,8), "_", filename)
  filename
}

changeLevels <- function(object, column = "chr_r", levels = c(paste0("chr", 1:22), "chrX")) {
  object[, column] <- as.character(object[, column])
  object[, column] <- factor(object[, column], levels)
  object
}
```

## Data import
```{r, message=FALSE, warnings=FALSE}
load("/DATA/projects/DSBrepair/data/R/RSTP2_Indel_Chromatin_2kb.RData")
load("/DATA/projects/DSBrepair/data/R/RSTP2_Indel_Chromatin_2kb_all.RData")
load("/DATA/projects/DSBrepair/data/R/RSTP2_IndelRatios_Chromatin_2kb.RData")

chromsizes<-read.table("/home/r.schep/mydata/data/genomes/GRCh38/hg38_chromsizes2.txt",header=F, stringsAsFactors = F)
names(chromsizes)<-c("chr_r","pos_r")

load("/DATA/projects/DSBrepair/data/R/rs20191126_Analyis_Mapping_RSTP2_2000.RData")

clone5barcodes <- c("AGGGCGTAAAATATTT.B",
                    "TATGGCTGTCGGGTAG.B",
                    "TGTCCCTTAGTACTTT.B",
                    "AGAAAATAATATGACG.B",
                    "CGGCCTGAAGGTCAGG.B",
                    "TTGAACGCGGGCTCGG.B",
                    "GCTAACATCACGAATC.B",
                    "GCGCACCCTTTAATTG.B",
                    "ACTGTCGAGTTGTCCG.B",
                    "CCGGGGACGTATGCAC.B",
                    "TCTTTTGAGGAGCTGA.B",
                    "ATATCGTTGCTGGAGA.B",
                    "CATCCACCACACTTCA.B",
                    "ACCCCTAAAGGCGCTG.B",
                    "ATACTATATTTAACGG.B",
                    "CATTTCTGATCAATAA.B",
                    "GAGCGCGTCACCGGGT.B",
                    "GTACCTCTCGATAGTG.B",
                    "TGGCCAATATTTGTCT.B")
```


## Data processing
```{r preprocess experiments}
trip_tib_mut <- copy(trip_tib_2000)
trip_tib_2000 <- trip_tib_2000 %>% distinct(barcode, exp, .keep_all = TRUE)
trip_lbr2_DMSO_tib <- trip_tib_2000 %>%
  filter(exp_pool=="indel_mean_LBR2_DMSO_64h")

trip_tib <- trip_tib_2000 %>%
  filter(exp_pool %in% c("indel_mean_LBR2_DMSO_64h", "indel_mean_LBR2_0_64h"))
trip_tib_mut <- trip_tib_mut_2000 %>%
  filter(exp_pool %in% c("indel_mean_LBR2_DMSO_64h", "indel_mean_LBR2_0_64h", "indel_mean_LBR2_NU7441_64h"))

trip_tib_mut_all <- trip_tib_mut_all_2000 %>%
  filter(exp_pool %in% c("indel_mean_LBR2_0_64h"))

mean_trip_tib_2000 <- trip_tib_2000 %>%
  gather(NHEJ, MMEJ, other_indels, key = "pathway", value = 'ratios') %>%
  distinct(barcode, exp, pathway, .keep_all = TRUE)

mean_trip_tib_2000$pathway <- factor(mean_trip_tib_2000$pathway, levels=c("other_indels", "MMEJ", "NHEJ"))
mean_trip_tib_2000 <- mean_trip_tib_2000[!is.na(mean_trip_tib_2000$ratios), ]



mutation_barcodes <- trip_tib_2000[trip_tib_2000$exp=="indel_mean_LBR2_0_64h" & trip_tib_2000$nexp > 1, ] %>% distinct(barcode, exp, .keep_all = TRUE)


mutationintegrations <-na.omit(analysis.mapped.integrations.df[c("barcode", "seqname","start")])
colnames(mutationintegrations) <- c("barcode", "chr_r","pos_r")
dim(mutationintegrations)
mutationintegrations$mapping <- "mapped"
mutationintegrations[mutationintegrations$barcode %in% mutation_barcodes$barcode, ]$mapping <- "filtered"

mutationintegrations <- changeLevels(mutationintegrations)
chromsizes <- changeLevels(chromsizes)
chromsizes <- chromsizes[!is.na(chromsizes$chr_r), ]
mutationintegrations <- mutationintegrations[!is.na(mutationintegrations$chr_r), ]
chromsizes$pos_r <- as.numeric(chromsizes$pos_r)


chrom.mods.cols <- seq(grep("binsize", names(trip_tib))+1, grep("LAD", names(trip_tib))-1)
```
```{r}
mutationintegrations <- mutationintegrations %>% mutate(pool = ifelse(barcode %in% clone5barcodes, "clone5", ifelse(grepl("[.]B", barcode), "B", "A")))
```


# Figures
## Main Figure
### Panel A
```{r chromosome ideogram, fig.height=6, fig.width=8}
ggplot() +
  geom_segment(data = chromsizes,
               aes(x = chr_r, xend = chr_r, y = 0, yend = pos_r),
               lineend = "round", color = "lightgrey", size = 7) +
  geom_segment(data = mutationintegrations,
               aes(x = as.integer(chr_r) - 0.4, xend = as.integer(chr_r) + 0.4,
                   y = pos_r, yend = pos_r, color = pool),
               size = .8) +
  theme_classic(base_size = 16) +
  scale_color_manual(values = c("A" = "purple", "B" = "black", "clone5" = "yellow")) +
  xlab("Chromosome")

table(mutationintegrations$pool)
```

#### Figure legend


### Panel B
```{r indel spectra, fig.height=6, fig.width=12}
# pdf(filename, width = 11, height = 5, useDingbats = F)
ggplot(filter(trip_tib_mut, drug == "-", mutation != "ssODN"),
       aes(mutation, freq, color = color)) +
  geom_quasirandom() +
  theme_bw(base_size = 16) +
  ylim(0, 1) +
  theme(legend.position = c(0.1, 0.8)) +
  scale_color_manual("legend", values = colory)
```
#### Figure legend ####


### Panel C
```{r pathway contribution, fig.height=6, fig.width=10}
pathway_distribution <- trip_tib %>% filter(exp_pool == "indel_mean_LBR2_0_64h") %>%
  gather(NHEJ, MMEJ, SSTR, other_indels, key = "pathway", value = 'ratios') %>% arrange(desc(ratios))
barcode_order <- pathway_distribution %>% filter(pathway == "NHEJ")  %>% arrange(desc(ratios)) %>% select(barcode)

pathway_distribution$barcode <- factor(pathway_distribution$barcode, levels=barcode_order$barcode)
pathway_distribution$pathway <- factor(pathway_distribution$pathway, levels=c("SSTR", "MMEJ", "other_indels",  "NHEJ"))

ggplot(pathway_distribution,
       aes(barcode, ratios, width=1)) +
  geom_bar(stat = "identity", aes(fill = pathway)) +
  theme_bw(base_size = 16) +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  ylab("Pathway Contribution") +
  scale_fill_manual("legend", values = colores) +
  ggtitle("Pathway contribution per barcode")
```
#### Figure legend ####



## Supplementary 1
### Panel A
```{r chromideograms supp, fig.height=6, fig.width=10}
ggplot() +
  geom_segment(data = chromsizes,
               aes(x = chr_r, xend = chr_r, y = 0, yend = pos_r),
               lineend = "round", color = "lightgrey", size = 7) +
  geom_segment(data = mutationintegrations,
               aes(x = as.integer(chr_r) - 0.3, xend = as.integer(chr_r) + 0.3,
                   y = pos_r, yend = pos_r, color = mapping),
               size = .8) +
  theme_classic(base_size = 16) +
  scale_color_manual(values = c("filtered" = "red", "mapped" = "black")) +
  xlab("Chromosome")

table(mutationintegrations$mapping)
```
#### Figure Legend

### Panel C
```{r clone5 heatmap, fig.height=6, fig.width=8}


# Keeping the scale in the pheatmap function
myBreaks1 <- seq(0,1,0.01)

# pheatmap function: logratio heatmap of each individual drug vs. barcode - 10uM drugs added

chom_matrix <- trip_tib %>% filter(drug == "-", barcode %in% clone5barcodes) %>% select(chrom.mods.cols) %>% as.matrix() %>%  apply(., 2, rescaler.default, type = "range")

pheatmap(chom_matrix,
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         breaks = myBreaks1)



```

<!-- start Christ -->

```{r clone5 heatmap Christ, fig.height=6, fig.width=8}

chip.dt = readRDS('../data/R/cl20191219_ChIP_zscore_selection_chipmentation.RDS')

chip_clone5 = chip.dt[pool=='clone5',]

clustering = data.frame(row.names=c('LMNB1', 'late_replicating', 'H3K9me2',
                                    'H3K9me3', 'H3K27me3', 'EZH2', 'CTCF', 'SMC3',
							        'HDAC3', 'HDAC2', 'HDAC1', 'H3K4me1',
								    'H3K4me2', 'H3K4me3', 'H3K27ac',
								    'H4K5acK8ac', 'H2AFZ', 'DNAse', 'Dam', 'm5C',
								    'H3K79me2', 'TTseq', 'H3K36me3', 'POL2AS2',
								    'POL2'),
					   group=factor(c(rep('repressive', 6), rep('insulator', 2),
				                      rep('HDAC', 3), rep('euchromatin', 6),
                                      rep('accessibility', 2), 'methylation',
						              rep('transcribing', 5)),
									levels=c('transcribing', 'accessibility',
                                             'methylation', 'euchromatin',
								             'HDAC', 'insulator', 'repressive')))
feature_fill = c(insulator="#a6b5a3", repressive="#304441", euchromatin="#cda4cc",
                 transcribing="#ED9114", HDAC="#aac9e0", accessibility='#cebc85',
                 methylation='#7dc98f')

chom_df = t(chip_clone5[,-c('pool', 'binsize', 'ID')])
colnames(chom_df) = chip_clone5$ID

order_vec=rownames(clustering)[nrow(clustering):1]

pheatmap(chom_df[order_vec,], cluster_rows=F, annotation_row=clustering,
		 annotation_colors=list(group=feature_fill),
         color = colorRampPalette(rev(brewer.pal(n = 11, name="RdBu")))(100))

```

<!-- end Christ -->

#### Figure Legend


### Panel D
```{r all indels, fig.height=6, fig.width=12}
ggplot(filter(trip_tib_mut_all, mutation != "ssODN"),
       aes(mutation, freq, color = color)) +
  geom_quasirandom() +
  theme_bw(base_size = 16) +
  ylim(0, 1) +
  theme(legend.position = c(0.1, 0.6)) +
  scale_x_discrete(breaks = seq(-115, 15, by = 5)) +
  scale_color_manual("legend", values = colory)

trip_tib_mut_all %>% filter(mutation != "ssODN") %>% group_by(mutation) %>% summarise(median = log10(median(freq))) %>%
  ggplot(., aes(mutation, fill = median, width=1)) + geom_bar(stat = "count") + theme_bw(base_size = 16)

trip_tib_mut_all %>% filter(mutation != "ssODN") %>% group_by(mutation) %>% summarise(mean_count = log10(mean(count))) %>%
  ggplot(., aes(mutation, fill = mean_count, width=1)) + geom_bar(stat = "count", ) + theme_bw(base_size = 16)

```

#### Figure legend ####

### Panel E
```{r difference per indel with NU7441 and DMSO, fig.height=6, fig.width=7}
barcodes_DMSO <- trip_tib_mut_all_2000 %>% filter(exp_pool == "indel_mean_LBR2_DMSO_64h") %>% pull(barcode) %>% unique
barcodes_NU7441 <- trip_tib_mut_all_2000 %>% filter(exp_pool == "indel_mean_LBR2_NU7441_64h") %>% pull(barcode) %>% unique


barcode_NU7441 <- Reduce(intersect, list(barcodes_DMSO,
                       barcodes_NU7441))

trip_tib_mut_filt <- trip_tib_mut_all_2000 %>% filter(barcode %in% barcode_NU7441) %>% filter(exp_pool %in% c("indel_mean_LBR2_DMSO_64h", "indel_mean_LBR2_NU7441_64h"))

trip_mut_lbr2_NU7441_diff <- trip_tib_mut_filt  %>%
  select("barcode", "mutation", "freq", "exp_pool", "color") %>%
  filter(barcode %in% barcode_NU7441) %>%  
  spread(exp_pool, freq) %>%
  filter(!is.na(indel_mean_LBR2_NU7441_64h), !is.na(indel_mean_LBR2_DMSO_64h)) %>%
  gather(indel_mean_LBR2_NU7441_64h, indel_mean_LBR2_DMSO_64h, key = "exp_pool", value = "freq") %>%
  mutate(mutation = as.character(mutation)) %>% select(-barcode, -color) %>% group_by(mutation)

mutations <- trip_mut_lbr2_NU7441_diff %>% group_by(mutation) %>% dplyr::summarise(count = n()) %>% filter(count > 10) %>% pull(mutation)

trip_mut_lbr2_NU7441_diff %>% filter(mutation %in% mutations) %>%
  do(broom::tidy(t.test(freq ~ exp_pool, data = .))) %>%
  ungroup() %>%
  mutate(p.adj = p.adjust(p.value, method = "fdr")) %>%
  select(mutation, p.value, p.adj) %>%
  arrange(p.adj)

trip_tib_mut_filt <- trip_tib_mut_all_2000 %>%
  filter(mutation %in% c("-22", "-14", "-7", "1") &
           exp_pool %in% c("indel_mean_LBR2_DMSO_64h",
                           "indel_mean_LBR2_NU7441_64h")) %>%
  group_by(mutation, drug)

ggplot(trip_tib_mut_filt, aes(mutation, freq, color = drug, alpha = 0.5)) +
  geom_quasirandom(dodge.width=1) +
  theme_bw(base_size = 16) +
  scale_color_manual(values = c("DMSO" = "black", "NU7441" = "gray"))
```
#### Figure legend ####


### Panel F
```{r NU7441 in the pool, fig.height=6, fig.width=5}
# DMSO samples
LBR2.DMSO.64h.samples <- trip_tib_mut_all_2000 %>% filter(plasmid == "LBR2",
                                                          drug == "DMSO",
                                                          ssODN == "-",
                                                          siRNA == "-",
                                                          replicate != "mean",
                                                          time == 64) %>%
  pull(exp_pool) %>% unique()

# NU7441 LBR2 sample
LBR2.NU7441.64h.samples <- trip_tib_mut_all_2000 %>% filter(plasmid == "LBR2",
                                                            drug == "NU7441",
                                                            ssODN == "-",
                                                            siRNA == "-",
                                                            replicate != "mean",
                                                            time == 64) %>%
  pull(exp_pool) %>% unique()

trip_tib_mut_all_2000


barcodes_DMSO <- trip_tib_mut_all_2000 %>% filter(exp_pool %in% LBR2.DMSO.64h.samples) %>% pull(barcode) %>% unique
barcodes_NU7441 <- trip_tib_mut_all_2000 %>% filter(exp_pool %in% LBR2.NU7441.64h.samples) %>% pull(barcode) %>% unique


barcode_NU7441 <- Reduce(intersect, list(barcodes_DMSO,
                       barcodes_NU7441))

trip_tib_mut_filt <- trip_tib_mut_all_2000 %>% filter(barcode %in% barcode_NU7441) %>% filter(exp_pool %in% c(LBR2.DMSO.64h.samples, LBR2.NU7441.64h.samples))

trip_mut_lbr2_NU7441_diff <- trip_tib_mut_filt  %>%
  select("barcode", "mutation", "freq", "exp_pool", "drug") %>%
  mutate(mutation = as.character(mutation)) %>% group_by(mutation, barcode)

p.values.NU7441.mutations <- trip_mut_lbr2_NU7441_diff  %>%
  select(-exp_pool) %>%  # Remove exp_pool so it doesn't interfere.
  group_by(mutation, barcode, drug) %>% # I want to know how many times each mutation was picked up per barcode per condition
  dplyr::summarise(count = n()) %>% # Summary to count this
  filter(drug == "NU7441" & count > 3 | drug == "DMSO" & count > 5) %>% # Select the barcodes that are present in all replicates.
  group_by(mutation, barcode) %>%
  dplyr::summarise(count = n()) %>%
  filter(count > 1) %>%
  select(-count) %>%
  left_join(trip_mut_lbr2_NU7441_diff, by = c("barcode", "mutation")) %>%
  group_by(mutation, barcode) %>%
  select(-exp_pool) %>%
  do(broom::tidy(t.test(freq ~ drug, data = .))) %>%
  ungroup() %>%
  mutate(p.adj = p.adjust(p.value, method = "fdr")) %>% # Calculate adjusted P.Val.
  arrange(p.adj)

p.values.NU7441.mutations %>%
  group_by(mutation) %>%
  dplyr::summarise(p.adj = median(p.adj),
            p.val = median(p.value)) %>%
  arrange(p.val)

barcodes <- p.values.NU7441.mutations %>% pull(barcode) %>% unique()

trip_tib_mut_filt <- trip_tib_mut_all_2000 %>%
  filter(exp_pool %in% c(LBR2.DMSO.64h.samples, LBR2.NU7441.64h.samples) &
           barcode %in% barcodes &
           mutation %in% c("-22", "-14", "-7", "1")) %>%
  group_by(mutation, exp_pool, drug) %>%
  dplyr::summarise(freq = median(freq))

ggplot(trip_tib_mut_filt, aes(mutation, freq, color = drug)) +
  geom_quasirandom(dodge.width=1) +
  theme_bw(base_size = 16) +
  scale_color_manual(values = c("DMSO" = "black", "NU7441" = "gray"))
```
#### Figure legend ####


### Panel G
```{r pathway contribution with siRNA, }
siRNAexps <- trip_tib_2000 %>% filter(replicate %in% c('mean'), siRNA != "-") %>% pull(exp_pool) %>% unique()

trip_lbr2_siRNA <- trip_tib_2000 %>%
  filter(exp_pool %in% c(siRNAexps))

trip_mut_lbr2_siRNA <- trip_tib_mut_2000 %>%
  filter(exp_pool %in% c(siRNAexps))

trip_mut_lbr2_all_siRNA <- trip_tib_mut_all_2000 %>%
  filter(exp_pool %in% c(siRNAexps))

trip_lbr2_siRNA_pathways <- trip_lbr2_siRNA %>%
  gather(NHEJ, MMEJ, SSTR, other_indels, key = "pathway", value = 'ratios') %>%
  distinct(barcode, siRNA, pathway, drug, .keep_all = TRUE)

exp_order <- c("siNT",
               "siLigIV",
               "siPolQ",
               "siCtIP",
               "siRad51")

trip_lbr2_siRNA_pathways$pathway <- factor(trip_lbr2_siRNA_pathways$pathway, levels=c("other_indels", "SSTR", "MMEJ", "NHEJ"))
trip_lbr2_siRNA_pathways$siRNA <- factor(trip_lbr2_siRNA_pathways$siRNA, levels=exp_order)
trip_lbr2_siRNA_pathways <- trip_lbr2_siRNA_pathways[!is.na(trip_lbr2_siRNA_pathways$ratios), ]

trip_mut_lbr2_siRNA_diff <- trip_mut_lbr2_all_siRNA  %>%
  select("barcode", "mutation", "freq", "color", "siRNA", "drug") %>%
  spread(siRNA, freq) %>%
  na.omit() %>%
  gather(siCtIP, siLigIV, siNT, siPolQ, siRad51, key = "siRNA", value = "freq") %>%
  mutate(mutation = as.character(mutation)) %>% select(-barcode, -color) %>% group_by(mutation)

exp_order <- c("siNT",
               "siLigIV",
               "siPolQ",
               "siCtIP",
               "siRad51")

trip_mut_lbr2_siRNA_diff$siRNA <- factor(trip_mut_lbr2_siRNA_diff$siRNA, levels=exp_order)

pathway_contribution <- trip_lbr2_siRNA_pathways %>% group_by(siRNA, pathway, drug) %>% dplyr::summarise(pathway_contr = mean(ratios))

p <- ggplot(pathway_contribution, aes(siRNA, pathway_contr, fill = pathway, order = pathway)) +
  geom_bar(stat = "identity", position = "fill")  +
  theme_bw(base_size = 16)
p + scale_fill_manual("legend", values = colori) +
  scale_x_discrete(labels=c("siNT" = "LBR2 siNT",
                            "siLigIV" = "LBR2 siLigIV",
                            "siCtIP" = "LBR2 siCtIP",
                            "siPolQ" = "LBR2 siPolQ",
                            "siRad51" = "LBR2 siRad51")) +
  labs(x = element_blank(),
       y = "abundance") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_grid(~ drug)
```
#### Figure legend ####

### Panel H
```{r pathway contribution with siRNA per barcode, }
ggplot(trip_lbr2_siRNA_pathways,
       aes(siRNA, ratios, color = pathway)) +
  geom_quasirandom() +
  theme_bw(base_size = 16) +
  ylim(0, 1) +
  scale_color_manual("legend", values = colori) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))  +
  facet_grid(drug ~ pathway)
```

#### Figure legend ####

### Panel I
```{r pathways with siRNA, }
mutations <- trip_mut_lbr2_siRNA_diff %>% group_by(mutation) %>% dplyr::summarise(count = n()) %>% filter(count > 10) %>% pull(mutation)

for(i in c("siLigIV",
               "siPolQ",
               "siCtIP",
               "siRad51")) {
  pvals <- trip_mut_lbr2_siRNA_diff %>%
    filter(mutation %in% mutations,
           siRNA %in% c("siNT", i),
           drug == "DMSO") %>%
    do(broom::tidy(t.test(freq ~ siRNA, data = .))) %>%
    ungroup() %>%
    mutate(p.adj = p.adjust(p.value, method = "fdr")) %>%
    select(mutation,conf.low, conf.high, p.value, p.adj) %>%
    arrange(p.adj)
  print(i)
  print(pvals)
}

trip_tib_mut_filt <- trip_mut_lbr2_all_siRNA %>%
  filter(mutation %in% c("-22", "-14", "-7", "1")) %>%
  group_by(mutation, drug, siRNA)

ggplot(trip_tib_mut_filt, aes(mutation, freq, color = siRNA, alpha = 0.5)) +
  geom_quasirandom(dodge.width=1) +
  theme_bw(base_size = 16) +
  facet_wrap(~ drug)
```
#### Figure legend ####

# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```
