---
title: "Gene expression processing"
author: "Tom van Schaik"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output: 
  html_document:
    theme: journal #cerulean
    highlight: monochrome
    toc: true
    toc_float: true
    code_folding: show
  editor_options:
    chunk_output_type: console
---


### Introduction

This document contains a basic differential analysis for the K562 cells from
Ruben and lamin knockouts (LBR and Lamin A).


### Method

DESeq2 on the gene counts for differential analysis and additional analyses. 


### Set-up

Load the libraries and set the parameters.

```{r set-up}

# Load dependencies
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(GenomicRanges))
suppressPackageStartupMessages(library(rtracklayer))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(reshape2))
suppressPackageStartupMessages(library(ggbeeswarm))
suppressPackageStartupMessages(library(DESeq2))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(GGally))
suppressPackageStartupMessages(library(UpSetR))

# # Prepare output 
output_dir <- "/DATA/usr/r.schep/projects/CRISPR_TRIP/20200514_RNAseq_clone5_LBRKO"
dir.create(output_dir, showWarnings = FALSE)

```

```{r knits setup}
library(knitr)
opts_chunk$set(fig.width = 10, fig.height = 4, 
               dev=c('png', 'pdf'), fig.path = file.path(output_dir, "figures/")) 
pdf.options(useDingbats = FALSE)
```

```{r functions}

ExonLength <- function(genes) {
  # Get the combined exon length for the genes
  tib <- as_tibble(genes) %>%
    filter(type == "exon") %>%
    mutate(gene_id = factor(gene_id, levels = unique(gene_id))) %>%
    group_by(gene_id) %>%
    dplyr::summarise(exon_number = n(),
              exon_length = sum(width))
  
  tib
}

# From Fede:
# ggpairs custom functions
corColor <- function(data, mapping, color = I("black"), sizeRange = c(1, 3), ...) {

  x   <- eval_data_col(data, mapping$x)
  y   <- eval_data_col(data, mapping$y)
  r   <- cor(x, y)
  rt  <- format(r, digits = 3)
  tt  <- as.character(rt)
  cex <- max(sizeRange)

  # helper function to calculate a useable size
  percent_of_range <- function(percent, range) {
    percent * diff(range) + min(range, na.rm = TRUE)
  }

  # plot correlation coefficient
  p <- ggally_text(label = tt, mapping = aes(), xP = 0.5, yP = 0.5,
                   size = I(percent_of_range(cex * abs(r), sizeRange)), color = color, ...) +
    theme(panel.grid.minor=element_blank(),
          panel.grid.major=element_blank())

  corColors <- RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")[2:6]

  if (r <= boundaries[1]) {
    corCol <- corColors[1]
  } else if (r <= boundaries[3]) {
    corCol <- corColors[2]
  } else if (r < boundaries[3]) {
    corCol <- corColors[3]
  } else if (r < boundaries[4]) {
    corCol <- corColors[4]
  } else {
    corCol <- corColors[5]
  }

  p <- p +
    theme(panel.background = element_rect(fill = corCol))

  return(p)
}

plotMAWithLabels <- function(results, main) {
  # Plot MA plot with lamin genes highlighted 
  
  plotMA(results, ylim = c(-8, 8), main = main,
         ylab = "log2 fold change", alpha = 0.05)
  points(results[c("ENSG00000143815", "ENSG00000160789"), "baseMean"],
         results[c("ENSG00000143815", "ENSG00000160789"), "log2FoldChange"],
         col = c("blue", "green"), cex = 1.5, pch = 19)
  legend("topright", 
         legend = c(paste0("non-sign (n=", sum(results$padj >= 0.5, na.rm = T), ")"), 
                    paste0("sign (n=", sum(results$padj < 0.5, na.rm = T), ")"), 
                    "LBR", "LMNA"), 
         col = c("black", "red", "blue", "green"), pch = 19)
}

```


### 1. Prepare data and DESeq2

Read data files and initialize the DESeq2 object.


```{r load rna-seq}

#######################################
## Prepare metadata

# Prepare metadata sheet
metadata <- read_tsv("metadata_concat_SE.tsv") %>%
  dplyr::select(-fastq.file) %>%
  mutate(sample.id.short = str_remove(sample.id, "_R1_001")) %>%
  separate(sample.name, c("cell", "sample", "rep"), sep = "_", remove = F) %>%
  filter(! str_detect(sample, "RSTP2")) %>%
  mutate(class = str_remove(sample, "-[0-9]+")) %>%
  mutate(sample = factor(sample, levels = unique(sample)),
         class = factor(class, levels = unique(class)))


#######################################
## Prepare gene annotation

# Load genes
genes <- import("/home/t.v.schaik/mydata/data/gene_builds/GRCh38/gencode.v24.primary_assembly.annotation.gtf")

# Get gene lengths (total exon length)
gene_length <- ExonLength(genes)

# dplyr::select genes only
genes <- genes[genes$type == "transcript"]

# Add seqinfo
# chrom_sizes <- read.table("/DATA/usr/t.v.schaik/data/genomes/GRCh38/hg38.chrom.sizes", sep = "\t")
# row.names(chrom_sizes) <- chrom_sizes[, 1]
# seqlengths(genes) <- chrom_sizes[seqlevels(genes), 2]


#######################################
## Prepare count matrix

# Read count matrix corresponding to the metadata
rnaseq.matrix <- read_tsv("results_se/Star/count_table.tsv") %>%
  dplyr::select(-contains("RSTP")) %>%
  rename_at(vars(names(.)), ~ c("ensembl.id", metadata$sample.name))

# Filter for strange chromosomes
idx.chromosome <- which(seqnames(genes) %in% c(paste0("chr", 1:22),
                                               "chrX"))

genes <- genes[idx.chromosome]
rnaseq.matrix <- rnaseq.matrix[idx.chromosome, ]

# Filter for protein coding genes (for now)
filter <- c("protein_coding")
idx.filter <- which(genes$gene_type %in% filter)

genes <- genes[idx.filter]
rnaseq.matrix <- rnaseq.matrix[idx.filter, ]

# Filter for genes with no counts
# tib.nocounts <- rnaseq.matrix %>%
#   mutate(ensembl.id = factor(ensembl.id, levels = ensembl.id)) %>%
#   gather(key, value, -ensembl.id) %>%
#   group_by(ensembl.id) %>%
#   dplyr::summarise(count = sum(value > 2) > 2)
# idx.nocounts <- which(tib.nocounts$count)
# 
# genes <- genes[idx.nocounts]
# rnaseq.matrix <- rnaseq.matrix[idx.nocounts, ]

```

```{r initialize deseq2, fig.width = 6, fig.height = 3.5}


#######################################
## Prepare count matrix

# Convert into data.frame with row.names for deseq2
rnaseq.counts <- rnaseq.matrix %>%
  dplyr::select(-ensembl.id)
rnaseq.counts <- data.frame(rnaseq.counts, row.names = rnaseq.matrix$ensembl.id)


#######################################
## Initialize deseq2

# Prepare metadata ready for deseq
metadata.df <- data.frame(metadata)
metadata.df[] <- lapply(metadata.df, function(x) gsub("-", "_", x))

# Initialize
rnaseq.dds <- DESeqDataSetFromMatrix(countData = rnaseq.counts,
                                     colData = metadata.df,
                                     design= ~ class)

# Execute deseq2
rnaseq.dds <- DESeq(rnaseq.dds)


#######################################
## PCA analysis - quality control

# Get the "normalized" values and create PCA plot
# Note that "normalized" is simply log2 + 0.01 transformed normalized counts!
rnaseq.dds.norm <- normTransform(rnaseq.dds, pc = 0.01)
plt <- plotPCA(rnaseq.dds.norm, intgroup = c("sample"))

plt + 
  theme_bw() +
  theme(aspect.ratio = 1)

```

In the PCA plot above you can observe that the replicates cluster together 
pretty well. Very roughly, you can argue that PC1 separates LBR knockout and
PC2 LMNA knockout (up to various degree). 

Without anything to really back this up, the low percentages and the fact that 
the LBR knockouts are almost "most deviating" suggest that the samples are all
quite similar.


### 2. Correlation plot replicates

A quick initial test: how well do the replicates correlate with each other and
between the different experiments?

```{r correlation plots, fig.width = 12, fig.height = 12, message = F, warning = F}

#######################################
## Make a correlation plot of all samples - quality control

# Get the "normalized" counts for each experiment - this is a different
# normalized from above!
rnaseq.norm <- as_tibble(assay(rnaseq.dds.norm)) %>%
  rename_at(vars(names(.)), ~ metadata$sample.name) %>%
  add_column(ensembl.id = rnaseq.matrix$ensembl.id)

# rnaseq.norm <- rnaseq.norm[, order(samples.df$clone)]

# Also, combine the normalized counts
rnaseq.norm.combined <- do.call(cbind,
                                tapply(as.character(metadata$sample.name),
                                       metadata$class,
                                       function(i) rowMeans(rnaseq.norm[, i])))

tib.norm.combined <- as_tibble(rnaseq.norm.combined, 
                               .name_repair = ~c("wt", "LBR_KO", 
                                                 "LMNA_KO", "LMNA_LBR_KO")) %>%
  add_column(gene_id = rnaseq.norm$ensembl.id)

tib.norm.combined <- as_tibble(genes) %>%
  dplyr::select(seqnames, start, end, strand, gene_id, gene_name) %>%
  inner_join(tib.norm.combined, by = "gene_id")

# Get a sample for plotting dplyr::select points
n <- sample(1:nrow(rnaseq.norm), 1000)

# Use GGally to make correlation plots
boundaries <- seq(from = 0.9, by = 0.01, length.out = 4)
plt <- ggpairs(rnaseq.norm %>% dplyr::select(-ensembl.id),
               upper = list(continuous = corColor),
               lower = list(continuous = function(data, mapping, ...) {
                   ggally_points(data = data[n, ], mapping = mapping, alpha = 0.1, size = 0.5) +
                   geom_abline(slope = 1, lty = "dashed", col = "red") +
                   theme_bw()}),
               diag = list(continuous = function(data, mapping, ...) {
                   ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red") +
                   theme_bw()})) +
  ggtitle("Correlation gene expression") +
  xlab("Gene expression") +
  ylab("Gene expression")
  # theme_bw()

print(plt)

```

The main point of the figure: there is no single replicate misbehaving.


### 3. Differential analysis

I will perform differential tests:

  * LBR-KO vs wt
  * LMNA-KO vs wt
  * LBR-LMNA-KO vs wt

Where "wt" is the parental clone.

```{r differential analysis, fig.width = 5.5, fig.height = 5}

#######################################
## Differential analysis - test for lfc > 0.5

# Differential analysis
results.LBR <- results(rnaseq.dds, lfcThreshold = 0.5,
                       contrast = c("class", "LBR_KO", "wt"))
summary(results.LBR)

results.LMNA <- results(rnaseq.dds, lfcThreshold = 0.5,
                        contrast = c("class", "LMNA_KO", "wt"))
summary(results.LMNA)

results.LBR_LMNA <- results(rnaseq.dds, lfcThreshold = 0.5,
                            contrast = c("class", "LMNA_LBR_KO", "wt"))
summary(results.LBR_LMNA)


#######################################
## MA-plots, with LBR and LMNA highlighted (blue and green)

# MA plots with significant hits
plotMAWithLabels(results.LBR, main = "LBR KO")

plotMAWithLabels(results.LMNA, main = "LMNA KO")

plotMAWithLabels(results.LBR_LMNA, main = "LBR and LMNA KO")

```

Single knockout of LBR or Lamin A seems to have a modest effect. Double KO 
results in more differentially regulated genes. 


### 4. Overlap with LADs

Finally, let's ask the important question: are the differentially regulated
genes found within LADs? This would suggest that LAD disruption has an effect 
on gene expression.

```{r load LADs and determine gene overlap}

#######################################
## Load LADs

# LADs: LAD definition in "wt" cells (clone 17)
LADs <- import("/DATA/usr/t.v.schaik/proj/lamin_KO/ts190306_Ruben_LBR-KO/results/HMM/bin-10kb/K562_17_LMNB1-10kb-combined_AD.bed.gz")


#######################################
## Determine overlap genes with LADs

genes$overlaps_LAD <- overlapsAny(genes, LADs, type = "any")

```

Plot the enrichment of differential genes within LADs.

```{r plot enrichment differential genes in LADs, fig.width = 5, fig.height = 3.5}

#######################################
## Create summarizing tibble - only using active genes

tib <- bind_cols(as_tibble(results.LBR), 
                 as_tibble(results.LMNA), 
                 as_tibble(results.LBR_LMNA)) %>%
  dplyr::select(baseMean, log2FoldChange, log2FoldChange1, log2FoldChange2, 
                padj, padj1, padj2) %>%
  dplyr::rename(log2FoldChange_LBR_KO = "log2FoldChange",
                log2FoldChange_LMNA_KO = "log2FoldChange1",
                log2FoldChange_LMNA_LBR_KO = "log2FoldChange2",
                pvalue_LBR_KO = "padj",
                pvalue_LMNA_KO = "padj1",
                pvalue_LMNA_LBR_KO = "padj2") %>%
  add_column(emsembl.id = genes$gene_id,
             overlaps.LAD = genes$overlaps_LAD) #%>%
  #filter(baseMean > 10)

tib.gather <- tib %>%
  dplyr::select(-starts_with("log")) %>%
  gather(key, value, -baseMean, -emsembl.id, -overlaps.LAD) %>%
  mutate(key = str_remove(key, "pvalue_")) %>%
  mutate(key = factor(key, levels = c("LBR_KO", "LMNA_KO", "LMNA_LBR_KO")),
         value = replace_na(value, 1)) %>%
  mutate(sign = value < 0.05)

# Group by LAD overlap or significance
tib.diff <- tib.gather %>%
  group_by(key, overlaps.LAD) %>%
  dplyr::summarise(fraction = mean(sign),
            total = sum(sign),
            background = length(sign))

tib.lad <- tib.gather %>%
  group_by(key, sign) %>%
  dplyr::summarise(fraction = mean(overlaps.LAD),
                   total = sum(overlaps.LAD),
                   background = length(overlaps.LAD))

#######################################
## Plot results

tib.diff %>%
  ggplot(aes(x = key, y = fraction, fill = overlaps.LAD)) +
    geom_bar(stat = "identity", position = "dodge", col = "black") +
    xlab("Knockout cells") +
    ylab("Fraction differentially expressed") +
    scale_fill_manual(values = c("grey", "black"), name = "Overlaps LAD") +
    theme_bw() +
    theme(aspect.ratio = 1)

tib.diff %>%
  ggplot(aes(x = key, y = total, fill = overlaps.LAD)) +
    geom_bar(stat = "identity", position = "dodge", col = "black") +
    xlab("Knockout cells") +
    ylab("Total differentially expressed") +
    scale_fill_manual(values = c("grey", "black"), name = "Overlaps LAD") +
    theme_bw() +
    theme(aspect.ratio = 1)


tib.lad %>%
  ggplot(aes(x = key, y = fraction, fill = sign)) +
    geom_bar(stat = "identity", position = "dodge", col = "black") +
    xlab("Knockout cells") +
    ylab("Fraction overlapping LAD") +
    scale_fill_manual(values = c("grey", "black"), name = "Significant") +
    theme_bw() +
    theme(aspect.ratio = 1)

tib.lad %>%
  ggplot(aes(x = key, y = total, fill = sign)) +
    geom_bar(stat = "identity", position = "dodge", col = "black") +
    xlab("Knockout cells") +
    ylab("Total overlapping LAD") +
    scale_fill_manual(values = c("grey", "black"), name = "Significant") +
    theme_bw() +
    theme(aspect.ratio = 1)

```

The following plot does suggest that LAD genes are specifically misregulated 
upon knockout of factors involved in lamina tethering.


### 5) LAD promoter classes

Finally, I will look at different LAD promoter classes from the Leemans, 2019 
paper. Promoters are classified based on endogenous expression and plasmid 
expression and grouped into inactive, repressed (by environment) and escaper 
(active).

```{r load promoter classes}

#######################################
## Load classes from Leemans, 2019

promoter.classes <- read_tsv("external_data/Leemans_2019_Dataset_S1_Promoter_SuRE_Classification.txt") %>%
  arrange(desc(log10_GROcap)) %>%
  distinct(gene_id, .keep_all = TRUE) 


#######################################
## Add promoter classes to genes

promoter.classes <- as_tibble(mcols(genes)) %>%
  left_join(promoter.classes, by = "gene_id") %>%
  dplyr::select(gene_id, overlaps_LAD, class)

```

Important note before plotting: every promoter has its own classification. I now
dplyr::selected the highest expressing promoter. Repeat the previous plot, subdivided in
promoter classes.

```{r plot enrichment differential genes in promoter classes, fig.width = 5, fig.height = 3.5}

#######################################
## Create summarizing tibble

tib.promoters <- bind_cols(as_tibble(results.LBR), 
                 as_tibble(results.LMNA), 
                 as_tibble(results.LBR_LMNA)) %>%
  dplyr::select(baseMean, log2FoldChange, padj, padj1, padj2) %>%
  dplyr::rename(LBR = "padj",
                LMNA = "padj1",
                LBR_LMNA = "padj2") %>%
  add_column(emsembl.id = promoter.classes$gene_id,
             promoter.class = promoter.classes$class) %>%
  gather(key, value, -baseMean, -log2FoldChange, -emsembl.id, -promoter.class) %>%
  mutate(key = factor(key, levels = c("LBR", "LMNA", "LBR_LMNA")),
         promoter.class = factor(promoter.class, 
                                 levels = c("iLAD", "boundary", "escaper", "repressed", "inactive")),
         value = replace_na(value, 1)) %>%
  mutate(sign = value < 0.05) %>%
  drop_na() %>%
  group_by(key, promoter.class) %>%
  dplyr::summarise(count = n(),
            fraction = mean(sign))


#######################################
## Plot results

tib.promoters %>%
  ggplot(aes(x = key, y = fraction, fill = promoter.class)) +
    geom_bar(stat = "identity", position = "dodge", col = "black") +
    xlab("Knockout cells") +
    ylab("Fraction differentially expressed") +
    scale_fill_brewer(palette = "Set1", name = "Promoter class") +
    theme_bw() +
    theme(aspect.ratio = 1)

```


### 6) Overlap differential results

Do we see the same genes changing upon LBR and LMNA knockouts? Let's test this.

```{r overlap diff results, fig.width = 9, fig.height = 5}

# Prepare upsetR 
tib.upset <- tib %>%
  dplyr::select(#starts_with("log"),
                starts_with("pvalue")) %>%
  mutate_at(vars(starts_with("pvalue")), function(x) (x < 0.05) + 0)

# Plot
upset(as.data.frame(tib.upset), keep.order = T)

```

So, yes, the majority of differential genes in LBR and LMNA separately, are also
differential in the DKO. However, there is little overlap in LBR and LMNA 
differential genes.




### Write normalized expression values

Finally, I will write several objects for further analyses comparing the
expression with DamID differences.

```{r write normalized expression values}

# Save RDS
saveRDS(genes,
        file.path(output_dir,
                  "genes.rds"))
saveRDS(tib,
        file.path(output_dir,
                  "tib_diff_results.rds"))
saveRDS(tib.norm.combined,
        file.path(output_dir,
                  "tib_norm_scores.rds"))

# Create BED files of the TSS of down / upregulated genes for visualization
chrom.info <- read.table("/DATA/usr/t.v.schaik/data/genomes/GRCh38/hg38.chrom.sizes", 
                         sep = "\t")
names(chrom.info) <- c("seqnames", "length")
row.names(chrom.info) <- chrom.info$seqnames

bigwig.dir <- file.path(output_dir, "bigwig")
dir.create(bigwig.dir, showWarnings = F)

for (diff_class in levels(metadata$class)) {
  
  # Select data per class
  diff_class <- gsub("-", "_", diff_class)
  
  tib.tmp <- tib.norm.combined %>%
    inner_join(tib, by = c("gene_id" = "emsembl.id")) %>%
    dplyr::select(seqnames, start, end, strand, 
                  starts_with(diff_class),
                  starts_with(paste0("pvalue_", diff_class)),
                  starts_with(paste0("log2FoldChange_", diff_class)))
  
  # Export bigwig with score per gene
  gr.tmp <- as(tib.tmp, "GRanges")
  start(gr.tmp) <- end(gr.tmp) <- ifelse(strand(gr.tmp) == "+",
                                         start(gr.tmp), end(gr.tmp))
  seqlengths(gr.tmp) <- chrom.info[seqlevels(gr.tmp), "length"]
  
  gr.tmp.bw <- gr.tmp
  gr.tmp.bw <- gr.tmp.bw[! duplicated(paste(seqnames(gr.tmp.bw),
                                            start(gr.tmp.bw)))]
  mcols(gr.tmp.bw) <- data.frame(score = mcols(gr.tmp.bw)[, diff_class])
  export.bw(gr.tmp.bw, file.path(bigwig.dir, paste0(diff_class, ".bw")))
  
  # bed file with diff calls
  if (diff_class == "wt") next
  
  names(mcols(gr.tmp)) <- c("mean", "pvalue", "logfoldchange")
  export.bed(gr.tmp[which(gr.tmp$pvalue < 0.05 & gr.tmp$logfoldchange > 0)], 
             file.path(bigwig.dir, paste0(diff_class, "_up.bed")))
  export.bed(gr.tmp[which(gr.tmp$pvalue < 0.05 & gr.tmp$logfoldchange < 0)], 
             file.path(bigwig.dir, paste0(diff_class, "_down.bed")))
  
}

```


### Conclusion

The change in gene expression upon knockout of lamina factors is limited, but 
does seem enriched within LADs. 

Looking at the bed files, there is a suspecious enrichment of downregulated
genes in the left arm of chromosome 2 and right arm of chromosome 9 for the 
double knockout. 

Looking at the unnormalized data tracks for Dam counts, there seems to be an
chromosome loss going on. I should leave out these regions for future analyses.
Even better, I should find simply create a log2-ratio of Dam-only tracks and 
find chromosomal changes like this.


### SessionInfo

```{r sessioninfo}

sessionInfo()

```
