---
title: "Figure 7 and supplementary 7-8"
author: "Ruben Schep"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    theme: journal #cerulean
    highlight: monochrome
    toc: true
    toc_float: true
    code_folding: show
  editor_options:
    chunk_output_type: console
---

# knitr document van Steensel lab
```{r}
StartTime <-Sys.time()
library(knitr)

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8)

## Select outdir
out.dir = paste0("figures/rs", Date, "/")
dir.create(out.dir)
opts_chunk$set(dev=c('png', 'pdf'), fig.path = file.path(out.dir))
pdf.options(useDingbats = FALSE)
```

# Introduction

## Description of Data

# Data processing
## Path, Libraries, Parameters and Useful Functions
```{r setup, message=FALSE, warnings=FALSE}
# INPUT DIRECTORY
in.dir.date = 20201029
in.dir = paste0("/DATA/projects/DSBrepair/data/R/rs", in.dir.date, "/")

# libraries:
library(ggplot2)
library(ggpubr)
# library(reshape2)
library(tibble)
# library(GenomicRanges)
# library(rtracklayer)
library(corrr)
library(Hmisc)
library(ggbeeswarm)
library(RColorBrewer)
library(data.table)
library(dplyr)
# library(plyr)
library(tidyr)
# library(stringr)
# library(plotly)
library(ggpmisc)
library(glmnet)
library(cowplot)
library(mda)
library(earth)
library(yaml)
library(vip)
library(caret)
library(scales)
library(glmnet)
library(gridExtra)
library(ggcorrplot)
library(bnlearn)
library(pheatmap)
library(ppcor)
library(parallel)

color_redblue <- rev(brewer.pal(11,"RdBu"))
duocolor <- c("#EE756E", "#2BB6BE")
colore <- c("wt" = "#808184", "other" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "SSTR" = "#007A4C")
colores <- c("wt" = "#808184", "other_indels" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "SSTR" = "#007A4C")
colori  <- c("other_indels" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "ssODN" = "#007A4C")
colora <- c("0" = "#808184", "1" = "#E1251B", "-7" = "#223DA0", "-14" = "#223DA0")
coloru <- c("wt" = "#808184", "other" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "MMEJplusNHEJ" = "#EE8A2A", "all" = "black", "MMEJplusNHEJplusHDR" = "#EE8A2A", "HDR" = "#007A4C", "MMEJplusHDR" = "#EE8A2A")
colory <- c("wt" = "#808184", "other" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "MMEJplusNHEJ" = "#EE8A2A", "all" = "black", "SSTR" = "#007A4C")

measure_color = c(NHEJ='#e1251b', MMEJ='#26419a', SSTR='#007A4C', None='grey')

method_color = c(MARS='#00bfc4', LASSO='#f8766d')
lad_color <- c(LAD="#802cdc", iLAD="#dfaf2f")
sign_color <- c(down="#AA6375", up="#7B9A36")

feature_fill = c(insulator="#a6b5a3", repressive="#304441", euchromatin="#cda4cc",
                 transcribing="#ED9114", HDAC="#aac9e0", accessibility='#cebc85',
                 methylation='#7dc98f')

feature_color = c(insulator="black", repressive="white", euchromatin="black",
                  transcribing="black", HDAC="black", accessibility='black',
                  methylation='black')

clustering = data.frame(row.names=c('LMNB1', 'late_replicating', 'H3K9me2','H3K27me3', 'EZH2', 'CTCF', 'SMC3',
							        'HDAC3', 'HDAC2', 'HDAC1', 'H3K4me1',
								    'H3K4me2', 'H3K4me3', 'H3K27ac',
								    'H4K5acK8ac', 'H2AFZ', 'DNAse', 'Dam', 'm5C',
								    'H3K79me2', 'TTseq', 'H3K36me3', 'POL2AS2',
								    'POL2'),
					   group=factor(c(rep('repressive', 5), rep('insulator', 2),
				                      rep('HDAC', 3), rep('euchromatin', 6),
                                      rep('accessibility', 2), 'methylation',
						              rep('transcribing', 5)),
									levels=c('transcribing', 'accessibility',
                                             'methylation', 'euchromatin',
								             'HDAC', 'insulator', 'repressive')))

```

## Custom functions
Functions used thoughout this script.
```{r, message=FALSE, warnings=FALSE}
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, substr(gsub("-","",Sys.time()),1,8), "_", filename)
  filename
}

CorrelationDFMean <- function(data, condition_name, feature_name, targets, p.value=0.001){
  hist.corr.list <- data[data$exp_pool==condition_name, ] %>% dplyr::select(all_of(feature_name), targets) %>% as.matrix() %>% rcorr(type = "pearson")

  histone.corr.df <- data.frame(rowname = names(hist.corr.list$r[1, -1]),
                                corr = hist.corr.list$r[1, -1],
                                pvalue = hist.corr.list$P[1, -1])

  return(histone.corr.df=histone.corr.df)
}


CorrelationPlotMean <- function(data, condition_name, feature_name, targets, p.value=0.001){
  cor.df = data.frame()
  for (i in feature_name){
    corr.list <- data %>% filter(exp_pool==condition_name) %>% dplyr::select(all_of(i), as.character(targets)) %>% as.matrix() %>% rcorr(type = "pearson")

    df <- data.frame(rowname = names(corr.list$r[1, -1]),
                                corr = corr.list$r[1, -1],
                                pvalue = corr.list$P[1, -1],
                                feature = i)
    cor.df <- rbind(cor.df, df)
  }
  p <- cor.df %>%
    transform(rowname = factor(rowname, levels = levels(targets)),
              corr = corr,
              pvalue = pvalue) %>%
    ggplot(aes(x = rowname, y = corr, fill = feature, alpha = pvalue < p.value)) +
    geom_bar(stat = "identity", position = "dodge") +
    ylab("Pathway correlation with chromatin feature") +
    xlab("chromatin feature") +
    theme_bw(base_size = 16) +
    scale_fill_manual("legend", values = colores) +
    scale_y_continuous(breaks=seq(-1,1, .1)) +
    coord_flip()
  return(list(dataframe=cor.df, plot = p))
}

cor_test <- function(y, X, left, right){
    cor_list = lapply(X, function(x){
        cor.test(x,y)
    })
    p_vec = unlist(lapply(cor_list, function(x){x$p.value}))
    cor_vec = unlist(lapply(cor_list, function(x){x$estimate}))
    p_adj = p.adjust(p_vec)
    color = ifelse(p_adj < 0.05, ifelse(cor_vec > 0, left, right),
                   'None')
    importance = abs(cor_vec)
    target=factor(colnames(X), levels=rownames(clustering))
    feature_group = clustering[colnames(X), ]
    return(list(correlation=cor_vec,importance = importance, p=p_vec,
                p_adj=p_adj, target=target, color=color,
                feature_group=feature_group))
}

get_combinations <- function(domain_vec){
    n_domain = length(domain_vec)
    get_ivec <-function (last_x, n_domain){
    	if (last_x < n_domain){
    		i_vec = c(0, c((last_x+1):n_domain))
    	} else{
    		i_vec = c(0)
    	}
    }

    combine <- function(x_new, x_vec, n_domain){
        new_vec = c(x_vec, x_new)
        if (length(new_vec)==n_domain){
            return(new_vec)
        } else {
            i_vec = get_ivec(max(new_vec), n_domain)
            x_list = lapply(i_vec, combine, x_vec=new_vec,
                            n_domain=n_domain)
            return(do.call(rbind, x_list))
        }
    }

    comb_list = lapply(c(0:n_domain), combine, x_vec=c(), n_domain=n_domain)

    comb = do.call(rbind, comb_list)
    comb_df = unique(t(apply(comb, 1, function(x){x[order(x)]})))

    name_list = apply(comb_df, 1, function(x){
    	name = paste(domain_vec[x], collapse='-')
    	if (name == ''){
    		name = 'iDomain'
    	}
    	return(name)
    })

    opt_list = lapply(domain_vec, function(d){
    	 apply(comb_df, 1, function(x){
    		opt = domain_vec[x]
    		return(d%in%opt)
    	})
    })

    opt_df = do.call(cbind, opt_list)
    colnames(opt_df) = domain_vec
    rownames(opt_df) = name_list

    return(opt_df)
}

test_wilcox <- function(dt, group_by='condition', y_name='MMEJ_MMEJNHEJ'){
    bc_vec = dt[,unique(barcode)]
    name1 = unlist(unique(dt[,group_by, with=F])[1])
    dt_c = dt[get(group_by)==eval(name1), ]
    dt_s = dt[get(group_by)!=eval(name1), ]
    setkey(dt_c, 'barcode')
    setkey(dt_s, 'barcode')

    w = wilcox.test(unlist(dt_c[bc_vec,..y_name]),
                    unlist(dt_s[bc_vec,..y_name]),
                    paired=T)
    return(list(p_value = w$p.value))
}



plot_comb_grid <- function(bix.dt, opt_df, domains, y_name, min_count=0,
                           max_count=Inf, group_by=NULL, color=NULL,
                           y_plabel=c(-3, 3)){

    opt_melt = reshape2::melt(opt_df)
    colnames(opt_melt) = c('group', 'domain', 'is_in')

    opt_melt$is_in = ifelse(opt_melt$is_in, 'in', 'out')
    opt_melt$domain = factor(opt_melt$domain, levels=domains)


    group_list = lapply(rownames(opt_df), function(opt_name){
    	opt = opt_df[opt_name, ]
    	in_vec = domains[opt]
    	out_vec = domains[!opt]

    	in_sum = rowSums(bix.dt[,..in_vec])
    	out_sum = rowSums(bix.dt[,..out_vec])
    	if (length(in_vec)==0){
    		in_group = which(out_sum == 0)
    	} else if (length(out_vec) == 0){
    		in_group = which(in_sum==length(in_vec))
    	} else  {
    		in_group = which(in_sum == length(in_vec) & out_sum==0)
    	}
    	return(in_group)
    })

    names(group_list) = rownames(opt_df)
    for (opt_name in names(group_list)){
    	i_vec = group_list[[opt_name]]
    	bix.dt[i_vec, group := opt_name]
    }
    if (is.null(group_by)){
        group_count_df = data.frame(bix.dt[,table(group)])
    } else {
        group_count_df = data.frame(bix.dt[,table(group)/length(unique(get(group_by)))])
    }
    group_count_df = group_count_df[order(match(group_count_df$group, rownames(opt_df))),]
    # print(group_count_df)
    group_count_sel = group_count_df[group_count_df$Freq>=min_count,]
    group_count_sel = group_count_sel[group_count_sel$Freq<=max_count,]
    group_levels =  group_count_sel$group
    opt_sel = opt_melt[opt_melt$group %in% group_levels, ]

    opt_sel$group = factor(opt_sel$group, levels=group_levels)
    # print(opt_sel)

    indel_selection = bix.dt[group%in%group_levels, ]
    indel_selection[,group:=factor(group, levels=group_levels)]


    group_count_sel$group = factor(group_count_sel$group, levels=group_levels)

    count_plot = ggplot(group_count_sel, aes(x=group,y=1, label=Freq)) +
                            geom_label() +
                            theme_void()



    dots = ggplot(opt_sel, aes(x=group, y=domain, color=is_in)) +
    	geom_point(size=2) +
    	geom_line(data=opt_sel[opt_sel$is_in=='in',], aes(group=group)) +
    	theme_bw() +
    	scale_color_manual(values=c('in'='black', 'out'='grey70')) +
    	guides(color=FALSE)  +
    	geom_vline(xintercept=seq(1.5, length(unique(bix.dt$group))-0.5, 1),
    			   lwd=0.5, colour="grey90") +
    	theme(axis.title=element_blank(),
    		  axis.text=element_blank(),
    		  axis.ticks=element_blank(),
    		  panel.grid.major = element_blank())

    symnum = list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1),
                  symbols = c("****", "***", "**", "*", " "))


    if (is.null(group_by)){
      level_vec = indel_selection[, levels(group)]
      wilcox_list = lapply(level_vec[level_vec!='iDomain'],
                           function(g){
                               x = indel_selection[group==g, y_name, with=F]
                               y = indel_selection[group=='iDomain', y_name,
                                                 with=F]
                               w = wilcox.test(unlist(x),unlist(y))
                               return(data.table('group1'='iDomain', 'group2'=g,
                                                 'p'=w$p.value))
                           })
      wilcox_dt = do.call(rbind, wilcox_list)
      wilcox_dt[,p_adj := p.adjust(p)]
      wilcox_dt[,p_label := sprintf('%.3g', p_adj)]
      wilcox_dt$y.position = seq(y_plabel[1], y_plabel[2],
                                 length.out=nrow(wilcox_dt))

      comparisons = lapply(level_vec[level_vec!='iDomain'],
                           function(x){return(c('iDomain', x))})
      beeswarm = ggplot(indel_selection, aes_string(x='group', y=y_name)) +
        geom_quasirandom(width=0.3) +
        stat_summary(fun=median, fun.min = median,
                     fun.max = median,
                     geom="crossbar", width = 0.5,
                     color='red') +
        stat_pvalue_manual(wilcox_dt, label = "p_label")
    } else {
        wilcox_dt = indel_selection[,test_wilcox(.SD, group_by, y_name),
                                    by='group']
        wilcox_dt[,p_adj:=p.adjust(p_value)]
        wilcox_dt[,p_signif:=symnum(p_adj, cutpoints=symnum$cutpoints,
                                    symbols=symnum$symbols)]
        beeswarm = ggplot(indel_selection,
                          aes_string(x='group', y=y_name, color=group_by)) +
        			   geom_quasirandom(dodge.width=1) +
                       scale_color_manual(values=color) +
                       stat_summary(aes_string(group=group_by), fun=median,
                                    fun.min = median, fun.max = median,
                                    geom="crossbar", width = 0.5,
                                    color='red',
                                    position= position_dodge(width =1)) +
                       geom_text(data=wilcox_dt, aes(x=group, y=0.8, label=p_signif),
                                 inherit.aes=F)
    }
    beeswarm = beeswarm +
    			  theme_bw() +
    			  geom_vline(xintercept=seq(1.5, length(unique(bix.dt$group))-0.5, 1),
    			             lwd=0.5, colour="grey90") +
    			  theme(axis.title=element_blank(),
    			        axis.text.x=element_blank(),
      				    panel.grid.major.x = element_blank(),
    				    axis.ticks.x=element_blank())


    count = data.frame(domain=gsub('_domain', '', domains),
    	               count=colSums(bix.dt[,..domains], na.rm=T))

    count$domain = factor(count$domain, levels=domain_levels)


    histogram = ggplot(count, aes(x=1, y=domain, label = count)) +
    				geom_label() +
    				theme_bw() +
    				theme(plot.background = element_blank(),
        				  panel.grid.major = element_blank(),
        				  panel.grid.minor = element_blank(),
        				  panel.border = element_blank(),
    					  axis.title=element_blank(),
    	  			      axis.text.x=element_blank(),
    	  				  axis.ticks=element_blank())

    empty = ggplot(data=data.frame(1,1),aes(x=1,y=1)) +
                geom_point() +
                theme_void()

    count_beeswarm = plot_grid(beeswarm, count_plot, ncol=1, rel_heights=c(20,1),
                               align='v', axis='lr')
    return(plot_grid(empty, beeswarm, empty, count_plot,
                     histogram, dots, nrow=3,
                     rel_widths=c(1,10), rel_heights=c(15,1,5), align='vh',
                     axis='tlbr'))
}

```

# Data import
```{r data loading, message=FALSE, warnings=FALSE}
trip_tib_mut_2000 = readRDS(paste0(in.dir, "RSTP2_Indel_Chromatin_2kb.RDS"))
trip_tib_2000 = readRDS(paste0(in.dir, "RSTP2_IndelRatios_Chromatin_2kb.RDS"))
```


## Data loading

```{r preprocess experiments}
trip_tib_mut <- copy(trip_tib_2000)
trip_tib_2000 <- trip_tib_2000 %>% distinct(barcode, exp, .keep_all = TRUE)

trip_tib <- trip_tib_2000 %>%
  filter(exp_pool  %in% c("mean_RSTP2_2000_LBR2_DMSO_64", "mean_RSTP2_2000_LBR2_NU7441_64", "mean_RSTP2_2000_LBR2_DMSO_ssODN_64", "mean_RSTP2_2000_LBR2_NU7441_ssODN_64"))

trip_tib_mut <- trip_tib_mut_2000 %>%
  filter(exp_pool %in% c("mean_RSTP2_2000_LBR2_DMSO_64", "mean_RSTP2_2000_LBR2_NU7441_64", "mean_RSTP2_2000_LBR2_DMSO_ssODN_64", "mean_RSTP2_2000_LBR2_NU7441_ssODN_64"))

mean_trip_tib_2000 <- trip_tib_2000 %>%
  gather(NHEJ, MMEJ, other_indels, key = "pathway", value = 'ratios') %>%
  distinct(barcode, exp, pathway, .keep_all = TRUE)

mean_trip_tib_2000$pathway <- factor(mean_trip_tib_2000$pathway, levels=c("other_indels", "MMEJ", "NHEJ"))
mean_trip_tib_2000 <- mean_trip_tib_2000[!is.na(mean_trip_tib_2000$ratios), ]
```

# Main Figure 7

## Panel 7A

```{r Fig7A pathways vs contribution ssODN, fig.height=6, fig.width=5}
trip_tib_pathways <- trip_tib %>%
  gather(NHEJ, MMEJ, other_indels, SSTR, key = "pathway", value = 'ratios') %>%
  distinct(barcode, exp, pathway, .keep_all = TRUE)

exp_order <- c("mean_RSTP2_2000_LBR2_DMSO_64",
              "mean_RSTP2_2000_LBR2_NU7441_64",
              "mean_RSTP2_2000_LBR2_DMSO_ssODN_64",
              "mean_RSTP2_2000_LBR2_NU7441_ssODN_64")

trip_tib_pathways$pathway <- factor(trip_tib_pathways$pathway, levels=c("other_indels", "MMEJ", "SSTR", "NHEJ"))
trip_tib_pathways$exp_pool <- factor(trip_tib_pathways$exp_pool, levels=exp_order)
trip_tib_pathways <- trip_tib_pathways[!is.na(trip_tib_pathways$ratios), ]

pathway_contribution <- trip_tib_pathways %>% group_by(exp_pool, pathway) %>% dplyr::summarise(pathway_contr = mean(ratios))

colores <- c("other_indels" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "SSTR" = "#007A4C")

p <- ggplot(pathway_contribution, aes(exp_pool, pathway_contr, fill = pathway, order = pathway)) +
  geom_bar(stat = "identity", position = "fill")  +
  theme_bw(base_size = 16)
p + scale_fill_manual("legend", values = colores) +
  scale_x_discrete(labels=c("mean_RSTP2_2000_LBR2_DMSO_64" = "DMSO",
                            "mean_RSTP2_2000_LBR2_NU7441_64" = "ssODN + DMSO",
                            "mean_RSTP2_2000_LBR2_DMSO_ssODN_64" = "NU7441",
                            "mean_RSTP2_2000_LBR2_NU7441_ssODN_64" = "ssODN + NU7441")) +
  labs(x = element_blank(),
       y = "abundance")
```

### Figure legend
(A) Average pathway contribution across all IPRs in the cell pools, in the absence or presence of a ssODN donor, and with or without NU7441 treatment (n = 2-8). Red: +1 insertion (NHEJ); blue: -7 deletion (MMEJ); green: +2 insertion due to SSTR; black: other indels.

## Panel 7B

```{r Fig7B pathways vs chrom ssODN NU7441, fig.height=8, fig.width=9}
chrom.mods.cols <- seq(grep("binsize", names(trip_tib))+1, grep("H3K27me3_domain", names(trip_tib))-1)

SSTR.hist.corr <- CorrelationDFMean(trip_tib, "mean_RSTP2_2000_LBR2_NU7441_ssODN_64", "NHEJ", targets = chrom.mods.cols) %>%
  arrange(corr) %>% pull(rowname) %>% factor(.,levels = .)

colores <- c("other_indels" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "SSTR" = "#007A4C")

lims = 1
# Over all indels (excl. 0)
CorrelationPlotMean(trip_tib, "mean_RSTP2_2000_LBR2_NU7441_ssODN_64", c("NHEJ", "MMEJ", "SSTR", "other_indels") , targets = SSTR.hist.corr)

# # Over all reads (incl. 0)
# colores <- c("other_indels" = "black", "NHEJ_freq" = "#E1251B", "MMEJ_freq" = "#223DA0", "SSTR_freq" = "#007A4C")
# CorrelationPlotMean(trip_tib, "mean_RSTP2_2000_LBR2_DMSO_ssODN_64", c("NHEJ_freq", "MMEJ_freq", "SSTR_freq", "other_indels") , targets = SSTR.hist.corr)


# # Same with NU7441 to check
# SSTR.hist.corr <- CorrelationDFMean(trip_tib, "mean_RSTP2_2000_LBR2_NU7441_ssODN_64", "NHEJ", targets = chrom.mods.cols) %>%
#   arrange(corr) %>% pull(rowname) %>% factor(.,levels = .)
# # Over all indels (excl. 0)
# colores <- c("other_indels" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "SSTR" = "#007A4C")
# CorrelationPlotMean(trip_tib, "mean_RSTP2_2000_LBR2_NU7441_ssODN_64", c("NHEJ", "MMEJ", "SSTR", "other_indels") , targets = SSTR.hist.corr)
# # Over all reads (incl. 0)
# colores <- c("other_indels" = "black", "NHEJ_freq" = "#E1251B", "MMEJ_freq" = "#223DA0", "SSTR_freq" = "#007A4C")
# CorrelationPlotMean(trip_tib, "mean_RSTP2_2000_LBR2_NU7441_ssODN_64", c("NHEJ_freq", "MMEJ_freq", "SSTR_freq", "other_indels") , targets = SSTR.hist.corr)
# 
# # Make balance check.
# trip_tib_bal = trip_tib %>% mutate(SSTR_MMEJNHEJSSTR = SSTR/(MMEJ+NHEJ+SSTR),
#                                    MMEJ_MMEJNHEJSSTR = MMEJ/(MMEJ+NHEJ+SSTR),
#                                    NHEJ_MMEJNHEJSSTR = NHEJ/(MMEJ+NHEJ+SSTR)) 
# 
# chrom.mods.cols <- seq(grep("binsize", names(trip_tib_bal))+1, grep("H3K27me3_domain", names(trip_tib_bal))-1)
# 
# SSTR.hist.corr <- CorrelationDFMean(trip_tib_bal, "mean_RSTP2_2000_LBR2_DMSO_ssODN_64", "SSTR_MMEJNHEJSSTR", targets = chrom.mods.cols) %>%
#   arrange(corr) %>% pull(rowname) %>% factor(.,levels = .)
# 
# colores <- c("NHEJ_MMEJNHEJSSTR" = "#E1251B", "MMEJ_MMEJNHEJSSTR" = "#223DA0", "SSTR_MMEJNHEJSSTR" = "#007A4C", "SSTR_MMEJSSTR" = "gray20")
# 
# 
# lims = 1
# CorrelationPlotMean(trip_tib_bal, "mean_RSTP2_2000_LBR2_DMSO_ssODN_64", c("SSTR_MMEJNHEJSSTR", "MMEJ_MMEJNHEJSSTR", "NHEJ_MMEJNHEJSSTR") , targets = SSTR.hist.corr)
```

### Figure legend
(B) Pearson’s correlations of the local intensities of 24 chromatin features versus the relative activity of each pathway (n = 2-3). Correlations with p > 0.001 not shown. Colors as in A.


## Panel 7C

```{r Fig7C SSTR NU7441 ratio per heterochomatin domain, fig.height=6, fig.width=10}
tib.dt = data.table(trip_tib)

domain_levels = c('late_replicating', 'LAD', 'H3K9me2', 'H3K27me3')

domains = paste0(domain_levels, '_domain')

indel.dt = unique(tib.dt[exp_pool == "mean_RSTP2_2000_LBR2_DMSO_ssODN_64",  
                         c('exp_pool', 'barcode', 'SSTR', 'SSTR_MMEJSSTR', 'NHEJ', 'MMEJ', 'efficiency', 'SSTR_freq', 'MMEJ_freq','NHEJ_freq',
                           domains), with=F])

indel.dt = indel.dt %>% mutate(NHEJ_eff = NHEJ/efficiency, MMEJ_eff = MMEJ/efficiency, SSTR_eff = SSTR/efficiency)

indel.dt[indel.dt=='iDomain'] <- 0
indel.dt[indel.dt=='Domain'] <- 1

indel.dt[is.na(LAD_domain), LAD_domain:=0]

indel.dt[, (domains) := lapply(.SD, as.numeric), .SDcols = domains]

colnames(indel.dt) = gsub('_domain', '', colnames(indel.dt))

opt_df = get_combinations(domain_levels)


# Ratio of SSTR vs all indels

# plot_comb_grid(indel.dt[exp_pool=='mean_RSTP2_2000_LBR2_DMSO_ssODN_64'], opt_df,
#                    domain_levels, 'efficiency', min_count = 25, y_plabel=c(0, 1.1))

plot_comb_grid(indel.dt[exp_pool=='mean_RSTP2_2000_LBR2_DMSO_ssODN_64'], opt_df,
                   domain_levels, 'SSTR', min_count = 25, y_plabel=c(0.15, 0.26))
# 
# plot_comb_grid(indel.dt[exp_pool=='mean_RSTP2_2000_LBR2_DMSO_ssODN_64'], opt_df,
#                    domain_levels, 'SSTR_freq', min_count = 25, y_plabel=c(0, 0.5))
# 
# plot_comb_grid(indel.dt[exp_pool=='mean_RSTP2_2000_LBR2_DMSO_ssODN_64'], opt_df,
#                    domain_levels, 'SSTR_eff', min_count = 25, y_plabel=c(0, 1))
# 
# plot_comb_grid(indel.dt[exp_pool=='mean_RSTP2_2000_LBR2_DMSO_ssODN_64'], opt_df,
#                    domain_levels, 'MMEJ', min_count = 25, y_plabel=c(0, 0.6))
# 
# plot_comb_grid(indel.dt[exp_pool=='mean_RSTP2_2000_LBR2_DMSO_ssODN_64'], opt_df,
#                    domain_levels, 'MMEJ_freq', min_count = 25, y_plabel=c(0, 0.6))
# 
# plot_comb_grid(indel.dt[exp_pool=='mean_RSTP2_2000_LBR2_DMSO_ssODN_64'], opt_df,
#                    domain_levels, 'MMEJ_eff', min_count = 25, y_plabel=c(0, 1))
# 
# plot_comb_grid(indel.dt[exp_pool=='mean_RSTP2_2000_LBR2_DMSO_ssODN_64'], opt_df,
#                    domain_levels, 'NHEJ', min_count = 25, y_plabel=c(0, 1))
# 
# plot_comb_grid(indel.dt[exp_pool=='mean_RSTP2_2000_LBR2_DMSO_ssODN_64'], opt_df,
#                    domain_levels, 'NHEJ_freq', min_count = 25, y_plabel=c(0, 1))
# 
# plot_comb_grid(indel.dt[exp_pool=='mean_RSTP2_2000_LBR2_DMSO_ssODN_64'], opt_df,
                   # domain_levels, 'NHEJ_eff', min_count = 25, y_plabel=c(0, 1))

```



### Figure legend
(C) Proportion of DSBs repaired by SSTR in the presence of the ssODN donor, for each IPR, split according to the heterochromatin features present as indicated by black dots in the scheme below the graph. Boxed numbers indicate the number of IPRs in each group; only groups with >20 IPRs are shown; see Figure S8I for all groups. Asterisks mark p-values according to the Wilcoxon test, compared to euchromatin IPRs (most left column): * p < 0.05, ** p < 0.01, *** p < 0.001, **** p < 0.0001. 


## Panel 7D

```{r Fig7D SSTR_MMEJ ratio per heterochomatin domain, fig.height=6, fig.width=10}
indel.dt[, MMEJ_SSTR := 1-SSTR_MMEJSSTR, ]

plot_comb_grid(indel.dt[exp_pool=='mean_RSTP2_2000_LBR2_DMSO_ssODN_64'], opt_df,
                   domain_levels, 'MMEJ_SSTR', min_count = 25, y_plabel=c(0.85, 1))
```

### Figure legend
(D) Same as in C, but showing the MMEJ:SSTR balance. See also Figure S8J.


# Supplementary Figure 7
## Panel S7A
Schematic of the ssODN
### Figure legend
(A) Schematic of the strategy to probe SSTR simultaneously with NHEJ and MMEJ. Prior to Cas9 activation, the ssODN is co-transfected with a plasmid that encodes the LBR2 sgRNA. The ssODN (black bar) covers the reporter sequence but not the barcode, and contains a 2 bp insertion (green) at the PAM. 


## Panel S7B
```{r FigS7B indel distribution ssODN, fig.height=4, fig.width=7}
SSTR_indels = filter(trip_tib_mut, exp_pool %in% c("mean_RSTP2_2000_LBR2_DMSO_ssODN_64", "mean_RSTP2_2000_LBR2_NU7441_ssODN_64"))

ggplot(SSTR_indels,
       aes(mutation, freq, color = color)) +
  geom_quasirandom() +
  theme_bw(base_size = 16) +
  theme(legend.position = c(0.2, 0.6)) +
  scale_color_manual("legend", values = colory) +
  facet_grid(. ~ drug)

SSTR_indels %>%
  group_by(mutation, drug) %>%
  dplyr::summarise(median = median(freq)) %>%
  arrange(dplyr::desc(median))

SSTR_indels %>%
  group_by(drug) %>%
  distinct(barcode, drug) %>% 
  dplyr::summarise(n_barcodes = n())

trip_tib_mut %>%
  distinct(exp_pool, nexp)

```

### Figure legend
(B) Indel frequencies generated by NHEJ, MMEJ and SSTR in 965 IPRs in the two cell pools, 64 hours after Cas9 activation (average of two replicate experiments). Left panel with DMSO, right panel with 1mM NU7441. N = 2-3. 

## Panel S7C
```{r FigS7C scatter NHEJ with and without ssODN, fig.height=5, fig.width=5}
trip_lbr2_ssODN_NHEJ <- trip_tib %>% filter(drug == "DMSO") %>% dplyr::select(barcode, ssODN, NHEJ, late_replicating) %>% spread(ssODN, NHEJ)

ggplot(trip_lbr2_ssODN_NHEJ, aes(`-`, ssODN)) +
  geom_point() +
  theme_bw(base_size = 16) +
  geom_abline(slope = 1)  +
  geom_smooth(method = glm) +
  xlim(.3, .82) + ylim(.3, .82)  +
  labs(title = "NHEJ - vs NHEJ + ssODN in DMSO condition",
       x = "NHEJ -",
       y = "NHEJ ssODN") +
  stat_cor(method = "pearson")


# trip_lbr2_ssODN_NHEJ <- trip_tib %>% filter(drug == "NU7441") %>% dplyr::select(barcode, ssODN, NHEJ, late_replicating) %>% spread(ssODN, NHEJ)
# 
# ggplot(trip_lbr2_ssODN_NHEJ, aes(`-`, ssODN)) +
#   geom_point() +
#   theme_bw(base_size = 16) +
#   geom_abline(slope = 1)  +
#   geom_smooth(method = glm) +
#   xlim(0, .65) + ylim(0, .65)  +
#   labs(title = "NHEJ - vs NHEJ + ssODN in NU7441 condition",
#        x = "NHEJ -",
#        y = "NHEJ ssODN") +
#   stat_cor(method = "pearson")
```

### Figure legend
(C-D) Comparison of +1 (NHEJ; panel C) and -7 (MMEJ, panel D) indel frequencies in all IPRs in cell pools in the presence (+) or absence (-) of the ssODN. Black line: diagonal. 


## Panel S7D
```{r FigS7D scatters MMEJ ssODN vs MMEJ - DMSO, fig.height=5, fig.width=5}
formula <- y ~ x

trip_lbr2_ssODN_MMEJ <- trip_tib %>% filter(drug == "DMSO") %>% dplyr::select(barcode, ssODN, MMEJ, late_replicating) %>% spread(ssODN, MMEJ)

ggplot(trip_lbr2_ssODN_MMEJ, aes(`-`, ssODN)) +
  geom_point() +
  theme_bw(base_size = 16) +
  geom_abline(slope = 1)  +
  geom_smooth(method = glm) +
  xlim(0,.45) + ylim(0,.45)  +
  labs(title = "MMEJ - vs MMEJ + ssODN in DMSO condition",
       x = "MMEJ -",
       y = "MMEJ ssODN") +
  stat_cor(method = "pearson", label.x = 0, label.y = .3)


# formula <- y ~ x
# 
# trip_lbr2_ssODN_MMEJ <- trip_tib %>% filter(drug == "NU7441") %>% dplyr::select(barcode, ssODN, MMEJ, late_replicating) %>% spread(ssODN, MMEJ)
# 
# ggplot(trip_lbr2_ssODN_MMEJ, aes(`-`, ssODN)) +
#   geom_point() +
#   theme_bw(base_size = 16) +
#   geom_abline(slope = 1)  +
#   geom_smooth(method = glm) +
#   xlim(0,.8) + ylim(0,.8)  +
#   labs(title = "MMEJ - vs MMEJ + ssODN in NU7441 condition",
#        x = "MMEJ -",
#        y = "MMEJ ssODN") +
#   stat_cor(method = "pearson", label.x = 0, label.y = .3)
```

###  Figure legend

## Panel S7E

```{r FigS7E MMEJ vs SSTR with DMSO, fig.height=5, fig.width=5}
trip_tib %>%
  filter(drug == "DMSO", ssODN == "ssODN") %>%
  ggplot(., aes(MMEJ, SSTR)) +
  geom_point() +
  theme_bw(base_size = 16) +
  geom_smooth(method = glm) +
  xlim(0,.3) + ylim(0,.3)  +
  labs(title = "MMEJ vs SSTR in DMSO condition",
       x = "MMEJ",
       y = "SSTR") +
  stat_cor(method = "pearson",label.x.npc = "left", label.y.npc = "top")
```

## Panel S7F
(E) Comparison of MMEJ and SSTR frequencies across all IPRs in cell pools treated with the ssODN. 

```{r FigS7F MMEJ vs SSTR with NU7441, fig.height=5, fig.width=5}
trip_tib %>%
  filter(drug == "NU7441", ssODN == "ssODN") %>%
  ggplot(., aes(MMEJ, SSTR)) +
  geom_point() +
  theme_bw(base_size = 16) +
  geom_smooth(method = glm) +
  xlim(0,.5) + ylim(0,.5)   +
  labs(title = "MMEJ vs SSTR in NU7441 condition",
       x = "MMEJ",
       y = "SSTR")  +
  stat_cor(method = "pearson", label.x.npc = "left", label.y.npc = "bottom")

trip_tib %>%
  filter(drug == "NU7441", ssODN == "ssODN") %>%
  ggplot(., aes(efficiency, SSTR)) +
  geom_point() +
  theme_bw(base_size = 16) +
  geom_smooth(method = glm) +
  labs(title = "efficiency vs SSTR in NU7441 condition",
       x = "efficiency",
       y = "SSTR")  +
  stat_cor(method = "pearson", label.x.npc = "left", label.y.npc = "bottom")

trip_tib %>%
  filter(drug == "DMSO", ssODN == "ssODN") %>%
  ggplot(., aes(efficiency, SSTR)) +
  geom_point() +
  theme_bw(base_size = 16) +
  geom_smooth(method = glm) +
  labs(title = "efficiency vs SSTR in DMSO condition",
       x = "efficiency",
       y = "SSTR")  +
  stat_cor(method = "pearson", label.x.npc = "left", label.y.npc = "bottom")

```

###  Figure legend
(F) Same as (E) but in the presence of NU7441. 
```{r}
pathway_distribution <- trip_tib %>%
  filter(exp_pool == "mean_RSTP2_2000_LBR2_NU7441_ssODN_64") %>%
  gather(NHEJ, MMEJ, SSTR, other_indels,
         key = "pathway", value = 'ratios') %>%
    distinct(barcode, pathway, ratios) %>%
  group_by(barcode) %>%
  dplyr::mutate(ratios = ratios/sum(ratios))

barcode_order <- trip_tib %>%
  filter(exp_pool == "mean_RSTP2_2000_LBR2_NU7441_ssODN_64") %>%
  gather(NHEJ, MMEJ, SSTR, other_indels,efficiency,
         key = "pathway", value = 'ratios') %>%
    distinct(barcode, pathway, ratios) %>%
  group_by(barcode) %>%
  dplyr::mutate(ratios = ratios/sum(ratios)) %>%
  filter(pathway == "efficiency")  %>%
  arrange(dplyr::desc(ratios)) %>%
  dplyr::select(barcode)

pathway_distribution$barcode <- factor(pathway_distribution$barcode, levels=barcode_order$barcode)
pathway_distribution$pathway <- factor(pathway_distribution$pathway, levels=c("SSTR", "MMEJ", "other_indels",  "NHEJ"))

ggplot(pathway_distribution,
       aes(barcode, ratios, width=1)) +
  geom_bar(stat = "identity", aes(fill = pathway)) +
  theme_bw(base_size = 16) +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  ylab("Pathway Contribution") +
  xlab("IPR") +
  scale_fill_manual("legend", values = colores) +
  ggtitle("Pathway contribution per barcode") 
```

```{r FigS7E SSTR in DMSO vs NU7441, fig.height=5, fig.width=5}
trip_tib %>%
  filter(ssODN == "ssODN") %>%
  dplyr::select(barcode, SSTR, drug) %>%
  tidyr::spread(drug, SSTR) %>%
  ggplot(., aes(DMSO, NU7441)) +
  geom_point() +
  theme_bw(base_size = 16) +
  geom_smooth(method = glm) +
  labs(title = "SSTR in DMSO vs NU7441",
       x = "DMSO",
       y = "NU7441") +
  stat_cor(method = "pearson",label.x.npc = "left", label.y.npc = "top")



trip_tib %>%
  gather(key = "pathway", value = "proportion", NHEJ, MMEJ, SSTR, other_indels, efficiency) %>%
  dplyr::select(barcode, pathway, proportion, drug, ssODN) %>%
  tidyr::spread(drug, proportion) %>%
  ggplot(., aes(DMSO, NU7441)) +
  geom_point() +
  theme_bw(base_size = 16) +
  geom_smooth(method = glm) +
  labs(title = "pathways in DMSO vs NU7441",
       x = "DMSO",
       y = "NU7441") +
  stat_cor(method = "pearson",label.x.npc = "left", label.y.npc = "top") + 
  facet_grid(ssODN ~ pathway, scales = "free")



trip_tib_2000 %>% filter(ssODN == "ssODN", plasmid == "LBR2", cell_line == "RSTP2_2000", time == 64) %>% distinct(barcode, exp_pool, .keep_all = T) %>% dplyr::select(barcode, replicate, drug, SSTR) %>% spread(replicate, SSTR) %>%
    ggplot(., aes(LBR2ssODNRep3, LBR2ssODNRep1)) +
  geom_point() +
  theme_bw(base_size = 16) +
  geom_smooth(method = glm) +
  labs(title = "SSTR in rep3 vs rep1",
       x = "rep3",
       y = "rep1") +
  stat_cor(method = "pearson",label.x.npc = "left", label.y.npc = "top") + 
  facet_grid(. ~ drug, scales = "free")

trip_tib %>%
  filter(ssODN == "ssODN") %>%
  dplyr::select(barcode, efficiency, drug) %>%
  tidyr::spread(drug, efficiency) %>%
  ggplot(., aes(DMSO, NU7441)) +
  geom_point() +
  theme_bw(base_size = 16) +
  geom_smooth(method = glm) +
  labs(title = "efficiency in DMSO vs NU7441",
       x = "DMSO",
       y = "NU7441") +
  stat_cor(method = "pearson",label.x.npc = "left", label.y.npc = "top")

# exp_pool  %in% c("mean_RSTP2_2000_LBR2_DMSO_64", "mean_RSTP2_2000_LBR2_NU7441_64", "mean_RSTP2_2000_LBR2_DMSO_ssODN_64", "mean_RSTP2_2000_LBR2_NU7441_ssODN_64")

trip_tib %>%
  filter(exp_pool  %in% c("mean_RSTP2_2000_LBR2_DMSO_64", "mean_RSTP2_2000_LBR2_NU7441_ssODN_64")) %>%
  dplyr::select(barcode, efficiency, drug) %>%
  tidyr::spread(drug, efficiency) %>%
  ggplot(., aes(DMSO, NU7441)) +
  geom_point() +
  theme_bw(base_size = 16) +
  geom_smooth(method = glm) +
  labs(title = "efficiency in DMSO vs NU7441 ssODN",
       x = "DMSO",
       y = "NU7441 ssODN") +
  stat_cor(method = "pearson",label.x.npc = "left", label.y.npc = "top")

trip_tib %>%
  filter(exp_pool  %in% c("mean_RSTP2_2000_LBR2_DMSO_64", "mean_RSTP2_2000_LBR2_NU7441_ssODN_64")) %>%
  dplyr::select(barcode, NHEJ, drug) %>%
  tidyr::spread(drug, NHEJ) %>%
  ggplot(., aes(DMSO, NU7441)) +
  geom_point() +
  theme_bw(base_size = 16) +
  geom_smooth(method = glm) +
  labs(title = "NHEJ in DMSO vs NU7441 ssODN",
       x = "DMSO",
       y = "NU7441 ssODN") +
  stat_cor(method = "pearson",label.x.npc = "left", label.y.npc = "top")
```




```{r replicate correlation DMSO}
res.tib = trip_tib_2000 %>% 
  filter(ssODN == "ssODN", plasmid == "LBR2", cell_line == "RSTP2_2000", time == 64) %>%
  distinct(barcode, exp_pool, .keep_all = T)
  
res =  res.tib %>% filter(drug == "DMSO") %>%
  dplyr::select(barcode, replicate, SSTR) %>% 
  spread(replicate, SSTR) %>% 
  column_to_rownames(., var = "barcode") %>%
  as.matrix() %>%
  rcorr()

library(corrplot)
corrplot(res$r, type = "upper", order = "hclust", 
         tl.col = "black", tl.srt = 45, diag = FALSE)


res = res.tib  %>%
  filter(drug == "DMSO") %>%
  dplyr::select(barcode, replicate, MMEJ) %>%
  spread(replicate, MMEJ) %>%
  column_to_rownames(., var = "barcode") %>%
  as.matrix() %>%
  rcorr()

library(corrplot)
corrplot(res$r, type = "upper", order = "hclust",
         tl.col = "black", tl.srt = 45, diag = FALSE)


res = res.tib %>%
  filter(drug == "DMSO") %>%
  dplyr::select(barcode, replicate, NHEJ) %>%
  spread(replicate, NHEJ) %>%
  column_to_rownames(., var = "barcode") %>%
  as.matrix() %>%
  rcorr()

library(corrplot)
corrplot(res$r, type = "upper", order = "hclust",
         tl.col = "black", tl.srt = 45, diag = FALSE)

res = res.tib %>%
  filter(drug == "DMSO") %>%
  dplyr::select(barcode, replicate, efficiency) %>%
  spread(replicate, efficiency) %>%
  column_to_rownames(., var = "barcode") %>%
  as.matrix() %>%
  rcorr()

library(corrplot)
corrplot(res$r, type = "upper", order = "hclust",
         tl.col = "black", tl.srt = 45, diag = FALSE)
```

```{r replicate correlation NU7441}
res = res.tib %>%
  filter(drug == "NU7441") %>%
  dplyr::select(barcode, replicate, SSTR) %>% 
  spread(replicate, SSTR) %>% 
  column_to_rownames(., var = "barcode") %>%
  as.matrix() %>%
  rcorr()

library(corrplot)
corrplot(res$r, type = "upper", order = "hclust", 
         tl.col = "black", tl.srt = 45, diag = FALSE)


res = res.tib %>%
  filter(drug == "NU7441") %>%
  dplyr::select(barcode, replicate, MMEJ) %>%
  spread(replicate, MMEJ) %>%
  column_to_rownames(., var = "barcode") %>%
  as.matrix() %>%
  rcorr()

library(corrplot)
corrplot(res$r, type = "upper", order = "hclust",
         tl.col = "black", tl.srt = 45, diag = FALSE)


res = res.tib %>%
  filter(drug == "NU7441") %>%
  dplyr::select(barcode, replicate, NHEJ) %>%
  spread(replicate, NHEJ) %>%
  column_to_rownames(., var = "barcode") %>%
  as.matrix() %>%
  rcorr()

library(corrplot)
corrplot(res$r, type = "upper", order = "hclust",
         tl.col = "black", tl.srt = 45, diag = FALSE)

res = res.tib %>%
  filter(drug == "NU7441") %>%
  dplyr::select(barcode, replicate, efficiency) %>%
  spread(replicate, efficiency) %>%
  column_to_rownames(., var = "barcode") %>%
  as.matrix() %>%
  rcorr()

library(corrplot)
corrplot(res$r, type = "upper", order = "hclust",
         tl.col = "black", tl.srt = 45, diag = FALSE)
```


```{r}
pathway_distribution <- trip_tib %>%
  filter(exp_pool == "mean_RSTP2_2000_LBR2_DMSO_ssODN_64") %>%
  gather(NHEJ, MMEJ, SSTR, other_indels,
         key = "pathway", value = 'ratios') %>%
    distinct(barcode, pathway, ratios) %>%
  group_by(barcode) %>%
  dplyr::mutate(ratios = ratios/sum(ratios))

barcode_order <- trip_tib %>%
  filter(exp_pool == "mean_RSTP2_2000_LBR2_DMSO_ssODN_64") %>%
  gather(NHEJ, MMEJ, SSTR, other_indels,efficiency,
         key = "pathway", value = 'ratios') %>%
    distinct(barcode, pathway, ratios) %>%
  group_by(barcode) %>%
  dplyr::mutate(ratios = ratios/sum(ratios)) %>%
  filter(pathway == "SSTR")  %>%
  arrange(dplyr::desc(ratios)) %>%
  dplyr::select(barcode)

pathway_distribution$barcode <- factor(pathway_distribution$barcode, levels=barcode_order$barcode)
pathway_distribution$pathway <- factor(pathway_distribution$pathway, levels=c("SSTR", "MMEJ", "other_indels",  "NHEJ"))

ggplot(pathway_distribution,
       aes(barcode, ratios, width=1)) +
  geom_bar(stat = "identity", aes(fill = pathway)) +
  theme_bw(base_size = 16) +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  ylab("Pathway Contribution") +
  xlab("IPR") +
  scale_fill_manual("legend", values = colores) +
  ggtitle("Pathway contribution per barcode") 
```




## Panel S7FG
```{r FigS7FG scatterplot pathways vs lmnb1 and K4me1 ssODN, fig.height=6, fig.width=12}
trip_tib %>%
  filter(ssODN == "ssODN") %>%
  dplyr::select(barcode, drug, MMEJ, NHEJ, other_indels, SSTR, late_replicating, H3K4me1) %>%
  group_by(barcode) %>%
  gather(MMEJ, NHEJ, other_indels, SSTR, key = "pathway", value = "ratio") %>%
  gather(late_replicating, H3K4me1, key = "feature", value = "value") %>%
  ggplot(., aes(ratio, value, alpha = 0.5))  +  
  geom_point(size = 0.5) +
  geom_smooth(method= "lm") +
  theme_bw(base_size = 16) +
  stat_cor(method = "pearson") +
  facet_grid(feature + drug  ~ pathway, scales = "free")

```
### Figure legend
(G-H) Scatterplots of the relative indel frequencies (proportion of all indels) versus (G) log2 H3K4me1 ChIP-seq signal and (H) log2 replication timing (positive values: late, negative values: early). R in panels C-G are Pearson’s correlation coefficients; blue lines with shading are linear regression fits with 95% confidence interval.

## Panel S7H
```{r FigS7H SSTR heterochrom domains, fig.height=6, fig.width=12}
plot_comb_grid(indel.dt[exp_pool=='mean_RSTP2_2000_LBR2_DMSO_ssODN_64'], opt_df,
                   domain_levels, 'SSTR', y_plabel=c(0.2, 0.25))
```

### Figure legend
(I-J) Same as Figures 7C-D, but now including heterochromatin types with 20 or fewer IPRs.

## Panel S7I
```{r FigS7I SSTR MMEJ balance heterochrom domains, fig.height=6, fig.width=12}
# Ratio of SSTR vs MMEJ
plot_comb_grid(indel.dt[exp_pool=='mean_RSTP2_2000_LBR2_DMSO_ssODN_64'], opt_df,
                   domain_levels, 'MMEJ_SSTR', y_plabel=c(0.85, 1))
```

### Figure legend

## Panel S7J
```{r FigS7C SSTR MMEJ balance per heterochomatin domain in NU7441, fig.height=6, fig.width=10}
tib.dt = data.table(trip_tib)

domain_levels = c('late_replicating', 'LAD', 'H3K9me2', 'H3K27me3')

domains = paste0(domain_levels, '_domain')

indel.dt = unique(tib.dt[exp_pool == "mean_RSTP2_2000_LBR2_NU7441_ssODN_64",  
                         c('exp_pool', 'barcode', 'SSTR', 'SSTR_MMEJSSTR', 'NHEJ', 'MMEJ', 'efficiency', 'SSTR_freq', 'MMEJ_freq',
                           domains), with=F])

indel.dt = indel.dt %>% mutate(NHEJ_eff = NHEJ/efficiency, MMEJ_eff = MMEJ/efficiency)

indel.dt[indel.dt=='iDomain'] <- 0
indel.dt[indel.dt=='Domain'] <- 1

indel.dt[, MMEJ_SSTR := 1-SSTR_MMEJSSTR, ]

indel.dt[is.na(LAD_domain), LAD_domain:=0]

indel.dt[, (domains) := lapply(.SD, as.numeric), .SDcols = domains]

colnames(indel.dt) = gsub('_domain', '', colnames(indel.dt))

opt_df = get_combinations(domain_levels)


# Ratio of SSTR vs all indels
plot_comb_grid(indel.dt[exp_pool=='mean_RSTP2_2000_LBR2_NU7441_ssODN_64'], opt_df,
                   domain_levels, 'SSTR', y_plabel=c(0.5, 0.6))



# plot_comb_grid(indel.dt[exp_pool=='mean_RSTP2_2000_LBR2_NU7441_ssODN_64'], opt_df,
#                    domain_levels, 'SSTR_freq', min_count = 25)
# 
# plot_comb_grid(indel.dt[exp_pool=='mean_RSTP2_2000_LBR2_NU7441_ssODN_64'], opt_df,
#                    domain_levels, 'MMEJ', min_count = 25)
# 
# plot_comb_grid(indel.dt[exp_pool=='mean_RSTP2_2000_LBR2_NU7441_ssODN_64'], opt_df,
#                    domain_levels, 'MMEJ_freq', min_count = 25)
# 
# plot_comb_grid(indel.dt[exp_pool=='mean_RSTP2_2000_LBR2_NU7441_ssODN_64'], opt_df,
#                    domain_levels, 'MMEJ_SSTR', min_count = 25)


barcodes_all_plasmids <- trip_tib %>%
  group_by(barcode) %>%
  dplyr::summarise(n = n()) %>%
  filter(n == 4) %>%
  pull(barcode)

indel.dt = trip_tib %>%
  dplyr::select(barcode, plasmid, efficiency, domains) %>%
  filter(barcode %in% barcodes_all_plasmids, plasmid != 'LBR2') %>%
  spread(plasmid, efficiency) %>%
  distinct(barcode, .keep_all = TRUE) %>%
  data.table()


indel.dt[indel.dt=='iDomain'] <- 0
indel.dt[indel.dt=='Domain'] <- 1

indel.dt[is.na(LAD_domain), LAD_domain:=0]

indel.dt[, (domains) := lapply(.SD, as.numeric), .SDcols = domains]

colnames(indel.dt) = gsub('_domain', '', colnames(indel.dt))

opt_df = get_combinations(domain_levels)

### CHRIST ####
# Can you now plot the ratio MMEJ_MMEJNHEJ of these domains that are significant vs the control condition to see the fold change and significance?

eff.dt = melt(indel.dt, measure.vars=c('LBR1', 'LBR12', 'LBR15'),
              variable.name='condition', value.name='efficiency')

guide_cols <- guide_colors[1:3]

plot_comb_grid_boxplot(eff.dt, opt_df, domain_levels, 'efficiency', group_by='condition', color=guide_cols, min_count = 20)

groups <- unique(eff.dt$group)[c(1, 3:13)]

for(i in groups) {
  pvals <- eff.dt %>% 
    dplyr::select(condition, efficiency, group) %>% 
    group_by(condition) %>%
    filter(group %in% c("iDomain", i)) %>%
    do(broom::tidy(wilcox.test(efficiency ~ group, data = .))) %>%
    ungroup() %>%
    mutate(p.adj = p.adjust(p.value, method = "fdr")) %>%
    dplyr::select(condition, statistic, p.value, p.adj) %>%
    arrange(p.adj)
  print(i)
  print(pvals)
} 
```
### Figure Legend

## Panel S7K
```{r FigS7K SSTR ratio per heterochomatin domain in NU7441, fig.height=6, fig.width=10}
indel.dt[, MMEJ_SSTR := 1-SSTR_MMEJSSTR, ]

# Ratio of SSTR vs MMEJ
plot_comb_grid(indel.dt[exp_pool=='mean_RSTP2_2000_LBR2_NU7441_ssODN_64'], opt_df,
                   domain_levels, 'MMEJ_SSTR', min_count = 25, y_plabel=c(0.85, 1))
```

## Panel S7L
```{r FigS7L scatterplot of SSTR proportion and frequency vs efficiency}
formula <- y ~ x

trip_tib %>% filter(ssODN == "ssODN") %>% 
ggplot(., aes(SSTR_freq, efficiency)) +
  geom_point() +
  theme_bw(base_size = 16) +
  geom_smooth(method = glm) +
  ylim(0.45, 1) +
  labs(title = "SSTR overall frequency vs efficiency",
       x = "SSTR_freq",
       y = "efficiency") +
  stat_cor(method = "pearson", label.x = 0.15, label.y = .5) + facet_wrap(. ~ drug)

trip_tib %>% filter(ssODN == "ssODN") %>% 
ggplot(., aes(SSTR, efficiency)) +
  geom_point() +
  theme_bw(base_size = 16) +
  geom_smooth(method = glm) +
  ylim(0.45, 1) +
  labs(title = "SSTR proportion vs efficiency",
       x = "SSTR",
       y = "efficiency") +
  stat_cor(method = "pearson", label.x = 0.15, label.y = .5) + facet_wrap(. ~ drug)

```


# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```
