---
title: "Figure 5 and supplemental 5"
author: "Ruben Schep"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    theme: journal #cerulean
    highlight: monochrome
    toc: true
    toc_float: true
    code_folding: show
  editor_options:
    chunk_output_type: console
---

# knitr document van Steensel lab
```{r}
StartTime <-Sys.time()
library(knitr)

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8)

## Select outdir
out.dir = paste0("figures/rs", Date, "/")
dir.create(out.dir)
opts_chunk$set(dev=c('png', 'pdf'), fig.path = file.path(out.dir))
pdf.options(useDingbats = FALSE)
```

# Introduction

## Description of Data

# Data processing
## Path, Libraries, Parameters and Useful Functions
```{r setup, message=FALSE, warnings=FALSE}
# INPUT DIRECTORY
in.dir.date = 20200305
in.dir = paste0("/DATA/projects/DSBrepair/data/R/rs", in.dir.date, "/")

# libraries:
library(ggplot2)
library(ggpubr)
# library(reshape2)
# library(tibble)
# library(GenomicRanges)
# library(rtracklayer)
library(corrr)
library(Hmisc)
library(ggbeeswarm)
library(RColorBrewer)
library(data.table)
library(dplyr)
# library(plyr)
library(tidyr)
# library(stringr)
# library(plotly)
library(ggpmisc)
library(glmnet)
library(cowplot)
library(mda)
library(earth)
library(yaml)
library(vip)
library(caret)
library(scales)
library(glmnet)
library(gridExtra)
library(ggcorrplot)
library(bnlearn)
library(pheatmap)
library(ppcor)
library(parallel)

color_redblue <- rev(brewer.pal(11,"RdBu"))
duocolor <- c("#EE756E", "#2BB6BE")
colore <- c("wt" = "#808184", "other" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "SSTR" = "#007A4C")
colores <- c("wt" = "#808184", "other_indels" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "SSTR" = "#007A4C")
colori  <- c("other_indels" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "ssODN" = "#007A4C")
colora <- c("0" = "#808184", "1" = "#E1251B", "-7" = "#223DA0", "-14" = "#223DA0")
coloru <- c("wt" = "#808184", "other" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "MMEJplusNHEJ" = "#EE8A2A", "all" = "black", "MMEJplusNHEJplusHDR" = "#EE8A2A", "HDR" = "#007A4C", "MMEJplusHDR" = "#EE8A2A")
colory <- c("wt" = "#808184", "other" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "MMEJplusNHEJ" = "#EE8A2A", "all" = "black", "SSTR" = "#007A4C")

measure_color = c(NHEJ='#e1251b', MMEJ='#26419a', SSTR='#007A4C', None='grey')

method_color = c(MARS='#00bfc4', LASSO='#f8766d')
lad_color <- c(LAD="#802cdc", iLAD="#dfaf2f")
sign_color <- c(down="#AA6375", up="#7B9A36")

feature_fill = c(insulator="#a6b5a3", repressive="#304441", euchromatin="#cda4cc",
                 transcribing="#ED9114", HDAC="#aac9e0", accessibility='#cebc85',
                 methylation='#7dc98f')

feature_color = c(insulator="black", repressive="white", euchromatin="black",
                  transcribing="black", HDAC="black", accessibility='black',
                  methylation='black')

clustering = data.frame(row.names=c('LMNB1', 'late_replicating', 'H3K9me2',
                                    'H3K27me3', 'EZH2', 'CTCF', 'SMC3',
                                    'HDAC3', 'HDAC2', 'HDAC1', 'H3K4me1',
                                    'H3K4me2', 'H3K4me3', 'H3K27ac',
                                    'H4K5acK8ac', 'H2AFZ', 'DNAse', 'Dam', 'm5C',
                                    'H3K79me2', 'TTseq', 'H3K36me3', 'POL2AS2',
                                    'POL2'),
                        group=factor(c(rep('repressive', 5), rep('insulator', 2),
                                       rep('HDAC', 3), rep('euchromatin', 6),
                                       rep('accessibility', 2), 'methylation',
                                       rep('transcribing', 5)),
                                     levels=c('transcribing', 'accessibility',
                                              'methylation', 'euchromatin',
                                              'HDAC', 'insulator', 'repressive')))

clone5barcodes <- c("AGGGCGTAAAATATTT.B",
                    "TATGGCTGTCGGGTAG.B",
                    "TGTCCCTTAGTACTTT.B",
                    "AGAAAATAATATGACG.B",
                    "CGGCCTGAAGGTCAGG.B",
                    "TTGAACGCGGGCTCGG.B",
                    "GCTAACATCACGAATC.B",
                    "GCGCACCCTTTAATTG.B",
                    "ACTGTCGAGTTGTCCG.B",
                    "CCGGGGACGTATGCAC.B",
                    "TCTTTTGAGGAGCTGA.B",
                    "ATATCGTTGCTGGAGA.B",
                    "CATCCACCACACTTCA.B",
                    "ACCCCTAAAGGCGCTG.B",
                    "ATACTATATTTAACGG.B",
                    "CATTTCTGATCAATAA.B",
                    "GAGCGCGTCACCGGGT.B",
                    "GTACCTCTCGATAGTG.B",
                    "TGGCCAATATTTGTCT.B")

```

### Custom functions
Functions used thoughout this script.
```{r functions, message=FALSE, warnings=FALSE}
CorrelationDFMean <- function(data, condition_name, feature_name, targets, p.value=0.001){
  hist.corr.list <- data[data$exp_pool==condition_name, ] %>% select(feature_name, targets) %>% as.matrix() %>% rcorr(type = "pearson")

  histone.corr.df <- data.frame(rowname = names(hist.corr.list$r[1, -1]),
                                corr = hist.corr.list$r[1, -1],
                                pvalue = hist.corr.list$P[1, -1])

  return(histone.corr.df=histone.corr.df)
}


CorrelationPlotMean <- function(data, condition_name, feature_name, targets, p.value=0.001){
  cor.df = data.frame()
  for (i in feature_name){
    corr.list <- data %>% filter(exp_pool==condition_name) %>% select(i, as.character(targets)) %>% as.matrix() %>% rcorr(type = "pearson")

    df <- data.frame(rowname = names(corr.list$r[1, -1]),
                                corr = corr.list$r[1, -1],
                                pvalue = corr.list$P[1, -1],
                                feature = i)
    cor.df <- rbind(cor.df, df)
  }
  p <- cor.df %>%
    transform(rowname = factor(rowname, levels = levels(targets)),
              corr = corr,
              pvalue = pvalue) %>%
    ggplot(aes(x = rowname, y = corr, fill = feature, alpha = pvalue < p.value)) +
    geom_bar(stat = "identity", position = "dodge") +
    ylab(paste("Pearson correlation with", feature_name)) +
    xlab("Chromatin Feature") +
    ggtitle(paste("Repair pathways correlation with chromatin features in", condition_name, "exps")) +
    theme_bw(base_size = 16) +
    scale_fill_manual("legend", values = colores) +
    scale_y_continuous(breaks=seq(-1,1, .1)) +
    coord_flip()
  return(list(dataframe=cor.df, plot = p))
}


cor_test <- function(y, X, left, right){
    cor_list = lapply(X, function(x){
        cor.test(x,y)
    })
    p_vec = unlist(lapply(cor_list, function(x){x$p.value}))
    cor_vec = unlist(lapply(cor_list, function(x){x$estimate}))
    p_adj = p.adjust(p_vec)
    color = ifelse(p_adj < 0.05, ifelse(cor_vec > 0, left, right),
                   'None')
    importance = abs(cor_vec)
    target=factor(colnames(X), levels=rownames(clustering))
    feature_group = clustering[colnames(X), ]
    return(list(correlation=cor_vec,importance = importance, p=p_vec,
                p_adj=p_adj, target=target, color=color,
                feature_group=feature_group))
}


get_combinations <- function(domain_vec){
    n_domain = length(domain_vec)
    get_ivec <-function (last_x, n_domain){
    	if (last_x < n_domain){
    		i_vec = c(0, c((last_x+1):n_domain))
    	} else{
    		i_vec = c(0)
    	}
    }

    combine <- function(x_new, x_vec, n_domain){
        new_vec = c(x_vec, x_new)
        if (length(new_vec)==n_domain){
            return(new_vec)
        } else {
            i_vec = get_ivec(max(new_vec), n_domain)
            x_list = lapply(i_vec, combine, x_vec=new_vec,
                            n_domain=n_domain)
            return(do.call(rbind, x_list))
        }
    }

    comb_list = lapply(c(0:n_domain), combine, x_vec=c(), n_domain=n_domain)

    comb = do.call(rbind, comb_list)
    comb_df = unique(t(apply(comb, 1, function(x){x[order(x)]})))

    name_list = apply(comb_df, 1, function(x){
    	name = paste(domain_vec[x], collapse='-')
    	if (name == ''){
    		name = 'iDomain'
    	}
    	return(name)
    })

    opt_list = lapply(domain_vec, function(d){
    	 apply(comb_df, 1, function(x){
    		opt = domain_vec[x]
    		return(d%in%opt)
    	})
    })

    opt_df = do.call(cbind, opt_list)
    colnames(opt_df) = domain_vec
    rownames(opt_df) = name_list

    return(opt_df)
}

test_wilcox <- function(dt, group_by='condition', y_name='MMEJ_MMEJNHEJ'){
    bc_vec = dt[,unique(barcode)]
    name1 = unlist(unique(dt[,group_by, with=F])[1])
    dt_c = dt[get(group_by)==eval(name1), ]
    dt_s = dt[get(group_by)!=eval(name1), ]
    setkey(dt_c, 'barcode')
    setkey(dt_s, 'barcode')

    w = wilcox.test(unlist(dt_c[bc_vec,..y_name]),
                    unlist(dt_s[bc_vec,..y_name]),
                    paired=T)
    return(list(p_value = w$p.value))
}



plot_comb_grid_beeswarm <- function(bix.dt, opt_df, domains, y_name, min_count=0,
                           max_count=Inf, group_by=NULL, color=NULL){

    opt_melt = reshape2::melt(opt_df)
    colnames(opt_melt) = c('group', 'domain', 'is_in')

    opt_melt$is_in = ifelse(opt_melt$is_in, 'in', 'out')
    opt_melt$domain = factor(opt_melt$domain, levels=domains)


    group_list = lapply(rownames(opt_df), function(opt_name){
    	opt = opt_df[opt_name, ]
    	in_vec = domains[opt]
    	out_vec = domains[!opt]

    	in_sum = rowSums(bix.dt[,..in_vec])
    	out_sum = rowSums(bix.dt[,..out_vec])
    	if (length(in_vec)==0){
    		in_group = which(out_sum == 0)
    	} else if (length(out_vec) == 0){
    		in_group = which(in_sum==length(in_vec))
    	} else  {
    		in_group = which(in_sum == length(in_vec) & out_sum==0)
    	}
    	return(in_group)
    })

    names(group_list) = rownames(opt_df)
    for (opt_name in names(group_list)){
    	i_vec = group_list[[opt_name]]
    	bix.dt[i_vec, group := opt_name]
    }
    if (is.null(group_by)){
        group_count_df = data.frame(bix.dt[,table(group)])
    } else {
        group_count_df = data.frame(bix.dt[,table(group)/length(unique(get(group_by)))])
    }
    group_count_df = group_count_df[order(match(group_count_df$group, rownames(opt_df))),]
    # print(group_count_df)
    group_count_sel = group_count_df[group_count_df$Freq>=min_count,]
    group_count_sel = group_count_sel[group_count_sel$Freq<=max_count,]
    group_levels =  group_count_sel$group
    opt_sel = opt_melt[opt_melt$group %in% group_levels, ]

    opt_sel$group = factor(opt_sel$group, levels=group_levels)
    # print(opt_sel)

    indel_selection = bix.dt[group%in%group_levels, ]
    indel_selection[,group:=factor(group, levels=group_levels)]

    # wilcox_list = lapply(group_levels[group_levels!='iDomain'],
    #                      function(g){
    #                          x = indel_selection[group==g, y_name, with=F]
    #                          y = indel_selection[group=='iDomain', y_name,
    #                                              with=F]
    #                          wilcox.test(unlist(x),unlist(y))
    #                      })

    group_count_sel$group = factor(group_count_sel$group, levels=group_levels)

    count_plot = ggplot(group_count_sel, aes(x=group,y=1, label=Freq)) +
                            geom_label() +
                            theme_void()



    dots = ggplot(opt_sel, aes(x=group, y=domain, color=is_in)) +
    	geom_point(size=2) +
    	geom_line(data=opt_sel[opt_sel$is_in=='in',], aes(group=group)) +
    	theme_bw() +
    	scale_color_manual(values=c('in'='black', 'out'='grey70')) +
    	guides(color=FALSE)  +
    	geom_vline(xintercept=seq(1.5, length(unique(bix.dt$group))-0.5, 1),
    			   lwd=0.5, colour="grey90") +
    	theme(axis.title=element_blank(),
    		  axis.text=element_blank(),
    		  axis.ticks=element_blank(),
    		  panel.grid.major = element_blank())

    symnum = list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1),
                  symbols = c("****", "***", "**", "*", " "))


    if (is.null(group_by)){
      beeswarm = ggplot(indel_selection, aes_string(x='group', y=y_name)) +
        geom_quasirandom(width=0.3) +
        stat_summary(fun.y=median, fun.ymin = median,
                     fun.ymax = median,
                     geom="crossbar", width = 0.5,
                     color='red') +
        stat_compare_means(aes(label = ..p.signif..),
                           symnum.args = symnum,
                           method = 'wilcox.test',
                           ref.group='iDomain')
    } else {
        wilcox_dt = indel_selection[,test_wilcox(.SD, group_by, y_name),
                                    by='group']
        wilcox_dt[,p_adj:=p.adjust(p_value)]
        wilcox_dt[,p_signif:=symnum(p_adj, cutpoints=symnum$cutpoints,
                                    symbols=symnum$symbols)]


        beeswarm = ggplot(indel_selection,
                          aes_string(x='group', y=y_name, color=group_by)) +
        			   geom_quasirandom(dodge.width=1) +
                       scale_color_manual(values=color) +
                       stat_summary(aes_string(group=group_by), fun.y=median,
                                    fun.ymin = median, fun.ymax = median,
                                    geom="crossbar", width = 0.5,
                                    color='red',
                                    position= position_dodge(width =1)) +
                       geom_text(data=wilcox_dt, aes(x=group, y=0.8, label=p_signif),
                                 inherit.aes=F)
    }
    beeswarm = beeswarm +
    			  theme_bw() +
    			  geom_vline(xintercept=seq(1.5, length(unique(bix.dt$group))-0.5, 1),
    			             lwd=0.5, colour="grey90") +
    			  theme(axis.title=element_blank(),
    			        axis.text.x=element_blank(),
      				    panel.grid.major.x = element_blank(),
    				    axis.ticks.x=element_blank())


    count = data.frame(domain=gsub('_domain', '', domains),
    	               count=colSums(bix.dt[,..domains], na.rm=T))

    count$domain = factor(count$domain, levels=domain_levels)


    histogram = ggplot(count, aes(x=1, y=domain, label = count)) +
    				geom_label() +
    				theme_bw() +
    				theme(plot.background = element_blank(),
        				  panel.grid.major = element_blank(),
        				  panel.grid.minor = element_blank(),
        				  panel.border = element_blank(),
    					  axis.title=element_blank(),
    	  			      axis.text.x=element_blank(),
    	  				  axis.ticks=element_blank())

    empty = ggplot(data=data.frame(1,1),aes(x=1,y=1)) +
                geom_point() +
                theme_void()

    count_beeswarm = plot_grid(beeswarm, count_plot, ncol=1, rel_heights=c(20,1),
                               align='v', axis='lr')
    return(plot_grid(empty, beeswarm, empty, count_plot,
                     histogram, dots, nrow=3,
                     rel_widths=c(1,10), rel_heights=c(15,1,5), align='vh',
                     axis='tlbr'))
}

```

# Data import
```{r import, message=FALSE, warnings=FALSE}
load(paste0(in.dir, "RSTP2_Indel_Chromatin_2kb.RData"))
load(paste0(in.dir, "RSTP2_IndelRatios_Chromatin_2kb.RData"))
```


## Data loading

```{r preprocess experiments}
trip_tib_mut <- copy(trip_tib_2000)
trip_tib_2000 <- trip_tib_2000 %>% distinct(barcode, exp, .keep_all = TRUE)

trip_tib <- trip_tib_2000 %>%
  filter(exp_pool  %in% c("mean_RSTP2_2000_LBR2_64",
               "mean_RSTP2_2000_LBR2_BIX01294_64",
               "mean_RSTP2_2000_LBR2_DMSOGSK_64",
               "mean_RSTP2_2000_LBR2_GSK126_64"))

trip_tib_mut <- trip_tib_mut_2000 %>%
  filter(exp_pool %in% c("mean_RSTP2_2000_LBR2_64",
               "mean_RSTP2_2000_LBR2_BIX01294_64",
               "mean_RSTP2_2000_LBR2_DMSOGSK_64",
               "mean_RSTP2_2000_LBR2_GSK126_64"))

mean_trip_tib_2000 <- trip_tib_2000 %>%
  gather(NHEJ, MMEJ, other_indels, key = "pathway", value = 'ratios') %>%
  distinct(barcode, exp, pathway, .keep_all = TRUE)

mean_trip_tib_2000$pathway <- factor(mean_trip_tib_2000$pathway, levels=c("other_indels", "MMEJ", "NHEJ"))
mean_trip_tib_2000 <- mean_trip_tib_2000[!is.na(mean_trip_tib_2000$ratios), ]
```
# Figures
## Main Figures
### Panel A

Panel A is the scheme of the ssODN. That has been made.
Add loading here I guess.

### Panel B

### Panel A

### GSK/BIX / control per heterochomatin domains.

```{r ratio DRUG/CTRL per het domain, fig.height=7, fig.width=12}
domain_levels = c('late_replicating', 'LAD', 'H3K9me2', 'H3K27me3')

domains = paste0(domain_levels, '_domain')

indel.dt = trip_tib %>%
  dplyr::select(barcode, drug, MMEJ_MMEJNHEJ, domains) %>%
  spread(drug, MMEJ_MMEJNHEJ) %>%
  filter(!is.na(BIX01294),!is.na(`-`), !is.na(GSK126), !is.na(DMSO_GSK)) %>%
  mutate(BIX_ratio = BIX01294/`-`,
         GSK_ratio = GSK126/DMSO_GSK,) %>%
  distinct(barcode, .keep_all = TRUE) %>%
  data.table()


# indel.dt = unique(tib.dt[exp_pool == "indel_mean_LBR2_ssODN_DMSO_64h",  
#                          c('exp_pool', 'barcode', 'SSTR', 'SSTR_MMEJSSTR',
#                            domains), with=F])


indel.dt[indel.dt=='iDomain'] <- 0
indel.dt[indel.dt=='Domain'] <- 1

indel.dt[is.na(LAD_domain), LAD_domain:=0]

indel.dt[, (domains) := lapply(.SD, as.numeric), .SDcols = domains]

colnames(indel.dt) = gsub('_domain', '', colnames(indel.dt))

opt_df = get_combinations(domain_levels)


# Ratio of SSTR vs all indels
plot_comb_grid_beeswarm(indel.dt, opt_df,
                   domain_levels, 'log2(BIX_ratio)', min_count = 25)

# Ratio of SSTR vs MMEJ
plot_comb_grid_beeswarm(indel.dt, opt_df,
                   domain_levels, 'log2(GSK_ratio)', min_count = 25)


# # BIX with general control
# bix.dt = melt(indel.dt, measure.vars=c('-', 'BIX01294'),
#               variable.name='condition', value.name='MMEJ_MMEJNHEJ')
# bix.dt = bix.dt[late_replicating == 0 &
#                 LAD == 0 &
#                 H3K9me2 == 1 &
#                 H3K27me3 == 0,,
#                 ]
# ggpaired(bix.dt, x = "condition", y = "MMEJ_MMEJNHEJ", 
#          order = c("-", "BIX01294"),
#          ylab = "MMEJ ratio", xlab = "treatment")
# 
# 
# gsk.dt = melt(indel.dt, measure.vars=c('DMSO_GSK', 'GSK126'),
#               variable.name='condition', value.name='MMEJ_MMEJNHEJ')
# 
# 
# color_vec = c('DMSO_GSK' = 'green', 'GSK126' = 'blue')
# 
# plot_comb_grid_beeswarm(gsk.dt, opt_df, domain_levels, 'MMEJ_MMEJNHEJ',
#                min_count = 25, group_by='condition', color=color_vec)
# 
# 
# gsk.dt = gsk.dt[late_replicating == 0 &
#                 LAD == 0 &
#                 H3K9me2 == 0 &
#                 H3K27me3 == 1,
#                 ]
# 
# ggpaired(gsk.dt, x = "condition", y = "MMEJ_MMEJNHEJ", 
#          order = c("DMSO_GSK", "GSK126"),
#          ylab = "MMEJ ratio", xlab = "treatment")

# ggplot(gsk.dt, aes(condition, MMEJ_MMEJNHEJ)) +
#   stat_summary(fun.y=median,
#                fun.ymin = median, fun.ymax = median,
#                geom="crossbar", width = 0.5,
#                color='red') +
#   geom_signif(comparisons = combn(levels(gsk.dt$condition),2, simplify = F), 
#               test = 'pairwise.wilcox.test', 
#               map_signif_level = TRUE) +
#   geom_quasirandom() + 
#   theme_bw(base_size = 16)
# 
# 
# dff <- gsk.dt %>% spread(condition, MMEJ_MMEJNHEJ)
# wilcox.test(dff$DMSO_GSK, dff$GSK126, paired = T)
# 
# df.bix <- trip_tib_2000 %>% filter(exp_pool %in% c("mean_RSTP2_2000_LBR2_64", "mean_RSTP2_2000_LBR2_BIX01294_64")) %>%
#   dplyr::select(barcode, drug, MMEJ_MMEJNHEJ, domains) %>%
#   spread(drug, MMEJ_MMEJNHEJ) %>%
#   filter(!is.na(BIX01294), !is.na(`-`))
# 
# wilcox.test(df.bix$`-`, df.bix$BIX01294, paired = T)
# 
# df.bix <- trip_tib_2000 %>% filter(exp_pool %in% c("mean_RSTP2_2000_LBR2_64", "mean_RSTP2_2000_LBR2_BIX01294_64"),
#          late_replicating_domain == "iDomain",
#          LAD_domain == "iDomain",
#          H3K9me2_domain == "Domain", 
#          H3K27me3_domain == "iDomain") %>%
#   dplyr::select(barcode, drug, MMEJ_MMEJNHEJ, domains) %>%
#   spread(drug, MMEJ_MMEJNHEJ) %>%
#   filter(!is.na(BIX01294), !is.na(`-`))
# 
# wilcox.test(df.bix$`-`, df.bix$BIX01294, paired = T)
```

```{r all indels LMNA and LBR ko sep replicates, fig.height= 12, fig.width= 12}
trip_tib_mut = trip_tib_mut_2000 %>%
  filter(replicate %in% c("LMNALBRKORep1", "LMNALBRKORep2", "LMNALBRKORep3"))

ggplot(trip_tib_mut,
       aes(mutation, freq, color = color)) +
  geom_quasirandom() +
  theme_bw(base_size = 16) +
  ylim(0, 1) +
  scale_color_manual("legend", values = colore) +
  facet_wrap(replicate ~ drug + plasmid, nrow = 5)
```



```{r all indels LMNA and LBR KO means, fig.height= 12, fig.width= 12}
trip_tib_mut = trip_tib_mut_2000 %>%
  filter(exp_pool %in% c("mean_RSTP2_clone5_LBR2_LBR_KO_1_64", "mean_RSTP2_clone5_LBR2_LBR_KO_2_64", 
                         "mean_RSTP2_clone5_LBR2_LBR_KO_3_64", "mean_RSTP2_clone5_LBR2_LBR_KO_4_64", 
                         "mean_RSTP2_clone5_LBR2_LMNA_KO_1_64", "mean_RSTP2_clone5_LBR2_LMNA_KO_2_64", 
                         "mean_RSTP2_clone5_LBR2_LMNA_KO_3_64", "mean_RSTP2_clone5_LBR2_LMNA_KO_4_64", 
                         "mean_RSTP2_clone5_LBR2_64", 
                         "mean_RSTP2_clone5_LBR2_LMNA_KOold_1_72", 
                         "mean_RSTP2_clone5_LBR2_LMNA_KOold_2_72", 
                         "mean_RSTP2_clone5_LBR2_LBR_KOold_1_72",
                         "mean_RSTP2_clone5_LBR2_LBR_KOold_2_72",
                         "mean_RSTP2_clone5_GFP_64", "mean_RSTP2_clone5_LBR2_DMSO_64",
                         "mean_RSTP2_clone5_LBR2_KOCTRL1_64", "mean_RSTP2_clone5_LBR2_KOCTRL2_64", "mean_RSTP2_clone5_LBR2_KOCTRL3_64", 
                         "mean_RSTP2_clone5_GFP_KOCTRL_64"),
         barcode %in% clone5barcodes)

ggplot(trip_tib_mut,
       aes(mutation, freq, color = color)) +
  geom_quasirandom() +
  theme_bw(base_size = 16) +
  ylim(0, 1) +
  scale_color_manual("legend", values = colore) +
  facet_wrap( ~ exp_pool)


```

```{r}
# pdf("/DATA/projects/DSBrepair/scratch/rs20200306_indelspectrum_indivbarcodes_ko.pdf", useDingbats = FALSE)
# for (barcodeset in clone5barcodes){
# 
# NHEJratio <- filter(trip_tib, barcode %in% barcodeset) %>% 
#   distinct(barcode, exp_pool, .keep_all = T) %>% 
#   dplyr::select(barcode, NHEJ_MMEJ, efficiency, exp_pool) %>% 
#   mutate(label_efficiency = paste0("Efficiency = ", round(efficiency*100, 1),"%"), 
#          label_ratio = paste0("Ratio = ", round(NHEJ_MMEJ, 1)),
#          label = gsub(".*LBR2_(.*)_.*", "\\1", exp_pool))
# 
# colore <- c("wt" = "#808184", "other" = "black", "NHEJ" = "#e1251b", "MMEJ" = "#26419a", "SSTR" = "#007A4C")
# 
# bc_mut_tib <- trip_tib_mut %>% 
#   filter(barcode %in% barcodeset) %>% dplyr::select(barcode, exp_pool, freq, mutation, color) %>%
#   left_join(NHEJratio, by = c("barcode", "exp_pool")) %>% filter(!is.na(label))
# 
# p <- ggplot(bc_mut_tib, 
#             aes(mutation, freq)) + 
#   geom_bar(stat = "identity", aes(fill = color)) + 
#   scale_fill_manual("legend", values = colore) + 
#   theme_bw() +
#   geom_text(data = NHEJratio, aes(x = 7, y = .7, label = label_efficiency)) +
#   geom_text(data = NHEJratio, aes(x = 7, y = .6, label = label_ratio)) +
#   scale_y_continuous(name="Frequency") +
#   scale_x_discrete(name="Indel size", breaks = c("<-15", -7, 0, 1, ">3"))  + 
#   facet_wrap( ~ label, ncol = 3) + 
#   ggtitle(barcodeset)
# 
# print(p)
# }
# dev.off()
```
```{r}
pdf("/DATA/projects/DSBrepair/scratch/rs20200306_eff_mmej_correlation_barcodes.pdf", useDingbats = FALSE)

bc_mut_tib <- trip_tib %>% dplyr::select(barcode, exp_pool, efficiency, MMEJ_MMEJNHEJ, drug) # %>%
  # left_join(NHEJratio, by = c("barcode", "exp_pool")) %>% filter(!is.na(label))

p <- ggplot(bc_mut_tib, 
            aes(efficiency, MMEJ_MMEJNHEJ)) + 
  geom_point() +
  geom_smooth(method= "lm") +
  theme_bw(base_size = 16) +
  stat_cor(method = "pearson") +
  ylab("KO MMEJ ratio / CTRL MMEJ ratio") +
  facet_wrap( ~ drug, ncol = 3)

print(p)
dev.off()
```

## LMNA and LBR KO in clone 5

### LMNA KO and LBR KO MMEJ ratio vs control in LADs

```{r LMNA LBR het domains}
trip_tib = trip_tib_2000 %>%
  filter(exp_pool %in% c("mean_RSTP2_clone5_LBR2_LBR_KO_1_64", "mean_RSTP2_clone5_LBR2_LBR_KO_2_64", 
                         "mean_RSTP2_clone5_LBR2_LBR_KO_3_64", "mean_RSTP2_clone5_LBR2_LBR_KO_4_64", 
                         "mean_RSTP2_clone5_LBR2_LMNA_KO_1_64", "mean_RSTP2_clone5_LBR2_LMNA_KO_2_64", 
                         "mean_RSTP2_clone5_LBR2_LMNA_KO_3_64", "mean_RSTP2_clone5_LBR2_LMNA_KO_4_64", 
                         "mean_RSTP2_clone5_LBR2_LMNA_KOold_1_72", "mean_RSTP2_clone5_LBR2_LMNA_KOold_2_72", 
                         "mean_RSTP2_clone5_LBR2_LBR_KOold_1_72", "mean_RSTP2_clone5_LBR2_LBR_KOold_2_72",
                         "mean_RSTP2_clone5_LBR2_KOCTRL1_64", 
                         "mean_RSTP2_clone5_LBR2_KOCTRL2_64", 
                         "mean_RSTP2_clone5_LBR2_KOCTRL3_64"),
         barcode %in% clone5barcodes) %>%
  distinct(barcode, exp_pool, .keep_all = TRUE)

indel.dt = trip_tib %>%
  dplyr::select(barcode, drug, MMEJ_MMEJNHEJ, domains, LMNB1) %>%
  spread(drug, MMEJ_MMEJNHEJ) %>%
  rowwise() %>%
  mutate(KO_CTRL = mean(c(KOCTRL1, KOCTRL2, KOCTRL3))) %>% 
  mutate(LMNA_KO1_old_ratio = LMNA_KOold_1/KO_CTRL,
         LMNA_KO2_old_ratio = LMNA_KOold_2/KO_CTRL,
         LMNA_KO1_ratio = LMNA_KO_1/KO_CTRL,
         LMNA_KO2_ratio = LMNA_KO_2/KO_CTRL,
         LMNA_KO3_ratio = LMNA_KO_3/KO_CTRL,
         LMNA_KO4_ratio = LMNA_KO_4/KO_CTRL,
         LBR_KO1_old_ratio = LBR_KOold_1/KO_CTRL,
         LBR_KO2_old_ratio = LBR_KOold_2/KO_CTRL,
         LBR_KO1_ratio = LBR_KO_1/KO_CTRL,
         LBR_KO2_ratio = LBR_KO_2/KO_CTRL,
         LBR_KO3_ratio = LBR_KO_3/KO_CTRL,
         LBR_KO4_ratio = LBR_KO_4/KO_CTRL) %>%
  distinct(barcode, .keep_all = TRUE) %>%
  data.table()



ggplot(trip_tib, aes(LAD_domain, MMEJ_MMEJNHEJ, fill = drug, color = LMNB1)) +
  geom_boxplot()

# trip_tib = trip_tib_2000 %>%
#   filter(drug %in% c("KO_CTRL", "LMNA_KO_1", "LMNA_KO_2", "LMNA_KO_3", "LMNA_KO_4", "LBR_KO_1", "LBR_KO_2", "LBR_KO_3", "LBR_KO_4"),
#          plasmid == "LBR2",
#          barcode %in% clone5barcodes,
#          sum_reads > 10000) %>%
#   distinct(barcode, exp_pool, .keep_all = TRUE)

indel.dt[indel.dt=='iDomain'] <- 0
indel.dt[indel.dt=='Domain'] <- 1

indel.dt[is.na(LAD_domain), LAD_domain:=0]

indel.dt[, (domains) := lapply(.SD, as.numeric), .SDcols = domains]

colnames(indel.dt) = gsub('_domain', '', colnames(indel.dt))




# trip_tib_2000 %>%
#   filter(pool == "clone5",
#          plasmid == "LBR2",
#          time %in% c(64, 72),
#          ssODN == "-",
#          replicate != "mean") %>%
#   mutate(group = paste(drug, siRNA, sep = "_")) %>%
#   ggplot(., aes(exp, MMEJ_MMEJNHEJ, color = group)) + 
#            geom_quasirandom(size = 3) +
#   theme_bw(base_size = 16) + 
#     stat_summary(fun.y=median, fun.ymin = median,
#           fun.ymax = median, geom="crossbar", width = 0.5,
#           color='red') +
#   facet_wrap(~ replicate)
```


```{r LMNA LBR KO indels, fig.height=14, fig.width=16}
ratio_cols <- colnames(indel.dt)[grep("ratio", colnames(indel.dt))]
KOsamples <- unique(trip_tib$drug)
LBRKOs <- KOsamples[grep("LMNA", KOsamples, invert = T)]
LMNAKOs <- KOsamples[grep("LBR", KOsamples, invert = T)]

LAD_bcs <- c("CATCCACCACACTTCA.B", "TCTTTTGAGGAGCTGA.B", "GTACCTCTCGATAGTG.B")

trip_tib <- trip_tib %>% mutate(strongLAD = ifelse(barcode %in% LAD_bcs, "LAD", "other"))

trip_tib %>% filter(drug %in% LMNAKOs) %>% ggplot(., aes(drug, MMEJ_MMEJNHEJ, color = strongLAD)) +
    geom_beeswarm(size = 3) + theme_bw(base_size = 16)

trip_tib %>% filter(drug %in% LBRKOs) %>% ggplot(., aes(drug, MMEJ_MMEJNHEJ, color = strongLAD)) +
    geom_beeswarm(size = 3) + theme_bw(base_size = 16)
```


```{r}
ratio_cols <- colnames(indel.dt)[grep("ratio", colnames(indel.dt))]

indel.dt %>% tidyr::gather(ratio_cols, key = ratio, value = value) %>% ggplot(., aes(ratio, log2(value), color = as.character(LAD))) + geom_beeswarm()

```



```{r}

ratio_col <- ratio_cols[grep("ratio$", ratio_cols)]
MMEJ_cols <- colnames(indel.dt)[grep("KOCTRL1", colnames(indel.dt)):grep("KO_CTRL", colnames(indel.dt))]

trip_tib_ko_ratios <- indel.dt %>% 
  gather(ratio_col, key = KO_ratio, value = ratio)

trip_tib_ko_MMEJ <- indel.dt %>%
  gather(all_of(MMEJ_cols), key = KO, value = MMEJ_MMEJNHEJ)

ctrls <- c("KOCTRL1", "KOCTRL2", "KOCTRL3")

pdf("/DATA/projects/DSBrepair/scratch/rs20200305_paired_ratio_change_controls.pdf")
for (i in ctrls) {
p <- trip_tib_ko_MMEJ %>% filter(KO %in% c(i, "KO_CTRL")) %>% ggpaired(., x = "KO", y = "MMEJ_MMEJNHEJ", 
                 order = c(i, "KO_CTRL"),
                 ylab = "MMEJ ratio", xlab = "control")
grid.arrange(p)
}
dev.off()



drugs <- unique(trip_tib$drug)[1:(length(unique(trip_tib$drug))-3)]

pdf("/DATA/projects/DSBrepair/scratch/rs20200305_paired_ratio_change.pdf")
for (i in drugs) {
p <- trip_tib_ko_MMEJ %>% filter(KO %in% c(i, "KO_CTRL")) %>% ggpaired(., x = "KO", y = "MMEJ_MMEJNHEJ", 
                 order = c(i, "KO_CTRL"),
                 ylab = "MMEJ ratio", xlab = "treatment")
grid.arrange(p)
}
dev.off()
```


```{r}

indel.ratio.dt = indel.dt[, c("barcode", "LMNB1", "LAD", ratio_col), with = FALSE]

indel.ratio.dt %>% gather(ratio_col, key = ratio, value = value) %>% 
    mutate(ko = gsub("^(.*_KO[1234]).*", "\\1", ratio)) %>% ggplot(., aes(ratio, log2(value), color = ko)) + 
  geom_point() + facet_wrap(~ barcode) + scale_y_continuous(breaks=seq(-1,1,.1)) + geom_abline(intercept = 0, slope = 0)

indel.ratio.dt %>% gather(ratio_col, key = ratio, value = value) %>% ggplot(., aes(barcode, value, color = LAD)) + 
  geom_point() + facet_wrap(~ ratio) + scale_y_continuous(breaks=seq(0,2,.2))

ratio_columns <- ratio_col[grep("old", ratio_col, invert = T)]
indel.ratio.dt = indel.dt[, c("barcode", "LMNB1", "LAD", ratio_columns), with = FALSE]

indel.ratio.dt %>% gather(ratio_columns, key = ratio, value = value) %>% 
  mutate(ko = gsub("^(.*_KO).*_.*", "\\1", ratio)) %>% ggplot(., aes(ratio, value, color = ko)) + geom_point() + 
  facet_wrap(~ barcode) + scale_y_continuous(breaks=seq(0,2,.2))

indel.ratio.dt %>% 
  gather(ratio_columns, key = ratio, value = value) %>% 
  mutate(ko = gsub("^(.*_KO).*_.*", "\\1", ratio)) %>% 
  ggplot(., aes(barcode, value, color = ko)) + geom_point() + scale_y_continuous(breaks=seq(0,2,.2))

indel.ratio.dt %>% 
  gather(ratio_columns, key = ratio, value = value) %>% 
  mutate(ko = gsub("^(.*_KO).*_.*", "\\1", ratio)) %>% 
  ggplot(., aes(ko, value, color = LMNB1)) + geom_quasirandom() + scale_y_continuous(breaks=seq(0,2,.2))


```

### LMNA KO and LBR KO MMEJ ratio difference vs control in LADs
```{r LMNA & LBR KO difference in LADS}

ratio_KO <- indel.dt %>% gather(KO_CTRL, LMNA_KO_1, LMNA_KO_2, LMNA_KO_3, LMNA_KO_4, key = "condition", value = "ratio")

for(i in c("LMNA_KO_1",
           "LMNA_KO_2",
           "LMNA_KO_3",
           "LMNA_KO_4")) {
  pvals <- ratio_KO %>%
    filter(condition %in% c("KO_CTRL", i)) %>%
    do(broom::tidy(t.test(ratio ~ condition, data = .))) %>%
    ungroup() %>%
    mutate(p.adj = p.adjust(p.value, method = "fdr")) # %>%
    # dplyr::select(condition,conf.low, conf.high, p.value, p.adj) %>%
    # arrange(p.adj)
  print(i)
  print(pvals)
}

ratio_KO <- indel.dt %>% gather(KO_CTRL, LBR_KO_1, LBR_KO_2, LBR_KO_3, LBR_KO_4, key = "condition", value = "ratio")

for(i in c("LBR_KO_1",
           "LBR_KO_2",
           "LBR_KO_3",
           "LBR_KO_4")) {
  pvals <- ratio_KO %>%
    filter(condition %in% c("KO_CTRL", i)) %>%
    do(broom::tidy(t.test(ratio ~ condition, data = .))) %>%
    ungroup() %>%
    mutate(p.adj = p.adjust(p.value, method = "fdr")) # %>%
    # dplyr::select(condition,conf.low, conf.high, p.value, p.adj) %>%
    # arrange(p.adj)
  print(i)
  print(pvals)
}

indel.dt %>% gather(KO_CTRL,LBR_KO_1, LBR_KO_2, LBR_KO_3, LBR_KO_4, LMNA_KO_1, LMNA_KO_2, LMNA_KO_3, LMNA_KO_4, key = "condition", value = "ratio") %>%
ggplot(., aes(LMNB1, ratio))  +  
  geom_point() +
  geom_smooth(method= "lm") +
  theme_bw(base_size = 16) +
  stat_cor(method = "pearson") +
  facet_wrap( ~ condition, ncol = 3)

```

```{r ratios}
indel.dt %>% gather(LMNA_KO1_ratio,
                    LMNA_KO2_ratio,
                    LMNA_KO3_ratio,
                    LMNA_KO4_ratio,
                    LBR_KO1_ratio,
                    LBR_KO2_ratio,
                    LBR_KO3_ratio,
                    LBR_KO4_ratio,
                    key = "condition",
                    value = "ratio") %>%
ggplot(., aes(LMNB1, ratio))  +  
  geom_point() +
  geom_smooth(method= "lm") +
  theme_bw(base_size = 16) +
  stat_cor(method = "pearson") +
  ylab("KO MMEJ ratio / CTRL MMEJ ratio") +
  facet_wrap( ~ condition)

indel.dt %>% gather(LMNA_KO1_ratio,
                    LMNA_KO2_ratio,
                    LMNA_KO3_ratio,
                    LMNA_KO4_ratio,
                    LBR_KO1_ratio,
                    LBR_KO2_ratio,
                    LBR_KO3_ratio,
                    LBR_KO4_ratio,
                    key = "condition",
                    value = "ratio") %>%
  group_by(condition) %>%
  do(broom::tidy(t.test(ratio ~ LAD, data = .))) %>%
  ungroup() %>%
  mutate(p.adj = p.adjust(p.value, method = "fdr")) %>%
  dplyr::select(condition,conf.low, conf.high, p.value, p.adj) %>%
  arrange(p.adj)



indel.dt %>%
  gather(LMNA_KO1_old_ratio,
         LMNA_KO2_old_ratio,
         LMNA_KO1_ratio,
         LMNA_KO2_ratio,
         LMNA_KO3_ratio,
         LMNA_KO4_ratio,
         LBR_KO1_old_ratio,
         LBR_KO2_old_ratio,
         LBR_KO1_ratio,
         LBR_KO2_ratio,
         LBR_KO3_ratio,
         LBR_KO4_ratio,
         key = "condition",
         value = "ratio") %>%
  ggplot(., aes(as.character(LAD), ratio))  +  
  geom_boxplot(outlier.colour =  "gray") +
  geom_quasirandom(binaxis='y',
               stackdir='center',
               dotsize=1) +
  theme_bw(base_size = 16) +
  ylab("KO MMEJ ratio / CTRL MMEJ ratio") +
  facet_wrap( ~ condition, nrow = 2)

indel.dt %>%
  gather(LMNA_KO1_ratio,
         LMNA_KO2_ratio,
         LMNA_KO3_ratio,
         LMNA_KO4_ratio,
         LBR_KO1_ratio,
         LBR_KO2_ratio,
         LBR_KO3_ratio,
         LBR_KO4_ratio,
         key = "condition",
         value = "ratio") %>%
  ggplot(., aes(condition, ratio))  +  
  geom_boxplot(outlier.colour =  "gray") +
  geom_quasirandom(binaxis='y',
               stackdir='center',
               dotsize=1) +
  theme_bw(base_size = 16) +
  ylab("MMEJ ratio")
```

```{r difference}
# indel.dt %>% gather(LMNA_KO1_diff, LMNA_KO2_diff, LBR_KO1_diff, LBR_KO2_diff, key = "condition", value = "ratio") %>%
# ggplot(., aes(LMNB1, ratio))  +  
#   geom_point() +
#   geom_smooth(method= "lm") +
#   theme_bw(base_size = 16) +
#   stat_cor(method = "pearson") +
#   ylab("KO MMEJ ratio - CTRL MMEJ ratio") +
#   facet_wrap( ~ condition)
# 
# indel.dt %>% gather(LMNA_KO1_diff, LMNA_KO2_diff, LBR_KO1_diff, LBR_KO2_diff, key = "condition", value = "ratio") %>%
#     group_by(condition) %>%
#     do(broom::tidy(t.test(ratio ~ LAD, data = .))) %>%
#     ungroup() %>%
#     mutate(p.adj = p.adjust(p.value, method = "fdr")) %>%
#     dplyr::select(condition,conf.low, conf.high, p.value, p.adj) %>%
#     arrange(p.adj)
# 
# indel.dt %>% gather(LMNA_KO1_diff, LMNA_KO2_diff, LBR_KO1_diff, LBR_KO2_diff, key = "condition", value = "ratio") %>%
# ggplot(., aes(as.character(LAD), ratio))  +  
#   geom_boxplot() +
#   geom_quasirandom(binaxis='y',
#                stackdir='center',
#                dotsize=1) +
#   theme_bw(base_size = 16) +
#   ylab("KO MMEJ ratio - CTRL MMEJ ratio") +
#   facet_wrap( ~ condition)
```

```{r LMNA LBR KO ratio in LADS}
rep_vec = c("LBR2PerturbClone5Rep1", "LBR2PerturbClone5Rep2")

si_mean_vec = c("siNT", "siMll3", "siMll4", "siSetD7", "siSetD1A", "siSetD1B")
si_single_vec = c("siPolQ", "siPool" , "siSETD2", "siRif1", "siIWS1", "siMRE11" )


indel_dt = data.table(trip_tib_2000)

no_col_vec = c('mutation', 'count', 'norm', 'sd_norm', 'ratio_indels',
               'freq', 'color', 'color_LBR1')

indel_mean_dt = unique(indel_dt[((grepl('mean', exp_pool) & siRNA %in% si_mean_vec) |
                                 siRNA %in% si_single_vec) & pool=='clone5' &
                                 drug=='-' & plasmid == 'LBR2',
                                -..no_col_vec])



indel_mean_dt[, MMEJ_change := log2(MMEJ_MMEJNHEJ/MMEJ_MMEJNHEJ[siRNA=='siNT']),
              by=c('barcode', 'time')]

indel_mean_dt[,siRNA:=factor(siRNA, levels=c(si_mean_vec, si_single_vec))]

ggplot(indel_mean_dt[time == 72,], aes(x=siRNA, y=MMEJ_MMEJNHEJ, color = H3K4me1)) +
    geom_quasirandom() +
    stat_summary(fun.y=median, fun.ymin = median,
          fun.ymax = median, geom="crossbar", width = 0.5,
          color='red') +
    stat_compare_means(aes(label = ..p.signif..),
                symnum.args = symnum,
                ref.group='siNT', paired=T)

ggplot(indel_mean_dt[time == 72,], aes(x=siRNA, y=MMEJ_change, color = H3K4me1)) +
    geom_quasirandom() +
    stat_summary(fun.y=median, fun.ymin = median,
          fun.ymax = median, geom="crossbar", width = 0.5,
          color='red') +
    stat_compare_means(aes(label = ..p.signif..),
                symnum.args = symnum,
                ref.group='siNT', paired=T)
```

```{r siRNA}
rep_vec = c("LBR2PerturbClone5Rep1", "LBR2PerturbClone5Rep2")

si_mean_vec = c("siNT", "siMll3", "siMll4", "siSetD7", "siSetD1A", "siSetD1B")
si_single_vec = c("siPolQ", "siPool") #, "siSETD2", "siRif1", "siIWS1", "siMRE11" )

trip_tib_K4si <- trip_tib_2000 %>% 
  filter(cell_line == "RSTP2_clone5", 
         plasmid == "LBR2", 
         ssODN == "-", 
         siRNA %in% c(si_mean_vec, si_single_vec), 
         time %in% c(72),
         replicate %in% c(rep_vec, "mean")) %>% 
  filter(!(siRNA %in% si_mean_vec & replicate != "mean")) %>%
  mutate(siRNA = factor(.$siRNA, levels=c("siNT", "siMll3", "siMll4", "siSetD7", "siSetD1A", "siSetD1B", "siPool", "siPolQ")))


ggplot(trip_tib_K4si, aes(siRNA, MMEJ_MMEJNHEJ, color = H3K4me1)) + geom_quasirandom(size = 3) +
  theme_bw(base_size = 16) + 
  geom_vline(xintercept=seq(1.5, length(unique(trip_tib_K4si$siRNA))-0.5, 1),
    			   lwd=2, colour="grey90")  +
    stat_summary(fun.y=median, fun.ymin = median,
          fun.ymax = median, geom="crossbar", width = 0.5,
          color='red')
ggplot(trip_tib_K4si, aes(siRNA, efficiency, color = H3K4me1)) + geom_quasirandom(size = 3) +
  theme_bw(base_size = 16) + 
  geom_vline(xintercept=seq(1.5, length(unique(trip_tib_K4si$siRNA))-0.5, 1),
    			   lwd=2, colour="grey90")  +
    stat_summary(fun.y=median, fun.ymin = median,
          fun.ymax = median, geom="crossbar", width = 0.5,
          color='red')


indel_mean_dt = as.data.table(trip_tib_K4si)
indel_mean_dt[, MMEJ_change := log2(MMEJ_MMEJNHEJ/MMEJ_MMEJNHEJ[siRNA=='siNT']),
              by=c('barcode', 'time')]

# indel_mean_dt[,siRNA:=factor(siRNA, levels=c(si_mean_vec, si_single_vec))]

ggplot(indel_mean_dt, aes(x=siRNA, y=MMEJ_change, color = H3K4me1)) +
    geom_quasirandom() +
    stat_summary(fun.y=median, fun.ymin = median,
          fun.ymax = median, geom="crossbar", width = 0.5,
          color='red') +
    stat_compare_means(aes(label = ..p.signif..),
                symnum.args = symnum,
                ref.group='siNT', paired=T)
```

```{r SETD2 Rif1}
si_single_vec = c("siNT", "siPolQ", "siSETD2", "siRif1")

trip_tib_K4si <- trip_tib_2000 %>% 
  filter(cell_line == "RSTP2_clone5", 
         plasmid == "LBR2", 
         ssODN == "-", 
         siRNA %in% si_single_vec, 
         time %in% c(72),
         replicate %in% c(rep_vec, "mean")) %>% 
  filter(!(siRNA %in% si_mean_vec & replicate != "mean")) %>%
  mutate(siRNA = factor(.$siRNA, levels=c("siNT", "siPolQ", "siSETD2", "siRif1")))


ggplot(trip_tib_K4si, aes(siRNA, MMEJ_MMEJNHEJ)) + geom_quasirandom(size = 3) +
  theme_bw(base_size = 16) + 
  geom_vline(xintercept=seq(1.5, length(unique(trip_tib_K4si$siRNA))-0.5, 1),
    			   lwd=2, colour="grey90")  +
    stat_summary(fun.y=median, fun.ymin = median,
          fun.ymax = median, geom="crossbar", width = 0.5,
          color='red')


# trip_tib_K4si %>% filter(siRNA != "siRif1") %>%
# ggplot(., aes(siRNA, MMEJ_MMEJNHEJ, color = H3K36me3)) + geom_quasirandom(size = 3) +
#   theme_bw(base_size = 16) + 
#   geom_vline(xintercept=seq(1.5, length(unique(trip_tib_txn$drug))-0.5, 1),
#     			   lwd=2, colour="grey90")  +
#     stat_summary(fun.y=median, fun.ymin = median,
#           fun.ymax = median, geom="crossbar", width = 0.5,
#           color='red')
# 
# trip_tib_K4si %>% filter(siRNA != "siSETD2") %>%
# ggplot(., aes(siRNA, MMEJ_MMEJNHEJ, color = late_replicating)) + geom_quasirandom(size = 3) +
#   theme_bw(base_size = 16) + 
#   geom_vline(xintercept=seq(1.5, length(unique(trip_tib_txn$drug))-0.5, 1),
#     			   lwd=2, colour="grey90")  +
#     stat_summary(fun.y=median, fun.ymin = median,
#           fun.ymax = median, geom="crossbar", width = 0.5,
#           color='red')


indel_mean_dt = as.data.table(trip_tib_K4si)
indel_mean_dt[, MMEJ_change := log2(MMEJ_MMEJNHEJ/MMEJ_MMEJNHEJ[siRNA=='siNT']),
              by=c('barcode', 'time')]

# indel_mean_dt[,siRNA:=factor(siRNA, levels=c(si_mean_vec, si_single_vec))]

ggplot(indel_mean_dt, aes(x=siRNA, y=MMEJ_change, color = H3K4me1)) +
    geom_quasirandom() +
    stat_summary(fun.y=median, fun.ymin = median,
          fun.ymax = median, geom="crossbar", width = 0.5,
          color='red') +
    stat_compare_means(aes(label = ..p.signif..),
                symnum.args = symnum,
                ref.group='siNT', paired=T)
```

## Supplementary Figures
### Panel xx
```{r seperate replicates, fig.height=12, fig.width=12}
colores <- c("wt" = "#808184", "other_indels" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "SSTR" = "#007A4C")

pathway_distribution <- trip_tib_2000 %>%
  filter(cell_line == "RSTP2_clone5", plasmid == "LBR2", time %in% c(64,72), drug %in% c("DMSO", "-"), replicate == "mean") %>%
  gather(NHEJ, MMEJ, SSTR, other_indels,
         key = "pathway", value = 'ratios') %>%
  arrange(dplyr::desc(ratios))
barcode_order <- pathway_distribution %>%
  filter(pathway == "NHEJ")  %>%
  arrange(dplyr::desc(ratios)) %>%
  dplyr::select(barcode) %>% unique()

pathway_distribution$barcode <- factor(pathway_distribution$barcode, levels=barcode_order$barcode)
pathway_distribution$pathway <- factor(pathway_distribution$pathway, levels=c("SSTR", "MMEJ", "other_indels",  "NHEJ"))

ggplot(pathway_distribution,
       aes(barcode, ratios, width=1)) +
  geom_bar(stat = "identity", aes(fill = pathway)) +
  theme_bw(base_size = 16) +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  ylab("Pathway Contribution") +
  scale_fill_manual("legend", values = colores) +
  ggtitle("Pathway contribution per barcode") +
  facet_wrap(~ exp_pool)


pathway_distribution <- trip_tib_2000 %>%
  filter(cell_line == "RSTP2_clone5", plasmid == "LBR2", time %in% c(64,72), siRNA %in% c("siNT", "-"), sum_reads > 4000, replicate != "mean") %>%
  gather(NHEJ, MMEJ, SSTR, other_indels,
         key = "pathway", value = 'ratios') %>%
  arrange(dplyr::desc(ratios))
barcode_order <- pathway_distribution %>%
  filter(pathway == "NHEJ")  %>%
  arrange(dplyr::desc(ratios)) %>%
  dplyr::select(barcode)  %>% unique()

pathway_distribution$barcode <- factor(pathway_distribution$barcode, levels=barcode_order$barcode)
pathway_distribution$pathway <- factor(pathway_distribution$pathway, levels=c("SSTR", "MMEJ", "other_indels",  "NHEJ"))

ggplot(pathway_distribution,
       aes(barcode, ratios, width=1)) +
  geom_bar(stat = "identity", position = "fill", aes(fill = pathway)) +
  theme_bw(base_size = 16) +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  ylab("Pathway Contribution") +
  scale_fill_manual("legend", values = colores) +
  ggtitle("Pathway contribution per barcode") +
  facet_wrap(drug + ssODN ~ siRNA)
```


### Panel H
```{r scatter NHEJ with and without ssODN, fig.height=6, fig.width=10}
txn_drugs <- unique(trip_tib_2000$drug[grepl("DRB|Trip|DMSO_", trip_tib_2000$drug)]) # get drug names with triptolide, BRB and DMSO
trip_tib_txn <- trip_tib_2000 %>% 
  filter(drug %in% txn_drugs, 
         cell_line == "RSTP2_clone5", 
         plasmid == "LBR2", 
         ssODN == "-", 
         siRNA == "-", 
         time %in% c(72)) %>% 
  mutate(drug = factor(.$drug, 
                                    levels=c("DMSO_low", "DMSO_mid", "DMSO_high",
                                             "DRB_35uM", "DRB_50uM", "DRB_70uM", 
                                             "Triptolide_5uM", "Triptolide_10uM", "Triptolide_20uM")))

```

```{r txn drugs}
pathway_distribution <- trip_tib_txn %>% 
  mutate(wt = 1-efficiency) %>% 
  gather(NHEJ, MMEJ, SSTR, other_indels, wt,
         key = "pathway", value = 'ratios') %>%
  arrange(dplyr::desc(ratios))
barcode_order <- pathway_distribution %>%
  filter(pathway == "wt", drug == "DMSO_low")  %>%
  arrange(dplyr::desc(ratios)) %>%
  dplyr::select(barcode)  %>% unique()

pathway_distribution$barcode <- factor(pathway_distribution$barcode, 
                                       levels=barcode_order$barcode)
pathway_distribution$pathway <- factor(pathway_distribution$pathway, 
                                       levels=c("SSTR", "MMEJ", "other_indels",  "NHEJ", "wt"))
pathway_distribution$drug <- factor(pathway_distribution$drug, 
                                    levels=c("DMSO_low", "DMSO_mid", "DMSO_high",
                                             "DRB_35uM", "DRB_50uM", "DRB_70uM", 
                                             "Triptolide_5uM", "Triptolide_10uM", "Triptolide_20uM"))

colores <- c("other_indels" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "SSTR" = "#007A4C", "wt" = "#808184")


ggplot(pathway_distribution,
       aes(barcode, ratios, width=1)) +
  geom_bar(stat = "identity", position = "fill", aes(fill = pathway)) +
  theme_bw(base_size = 16) +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  ylab("Pathway Contribution") +
  scale_fill_manual("legend", values = colores) +
  ggtitle("Pathway contribution per barcode") +
  facet_grid(time ~ drug)
```

```{r txn drugs again}
ggplot(trip_tib_txn, aes(drug, MMEJ_MMEJNHEJ, color = TTseq)) + geom_quasirandom(size = 3) +
  theme_bw(base_size = 16) + 
  geom_vline(xintercept=seq(1.5, length(unique(trip_tib_txn$drug))-0.5, 1),
    			   lwd=2, colour="grey90")  +
    stat_summary(fun.y=median, fun.ymin = median,
          fun.ymax = median, geom="crossbar", width = 0.5,
          color='red')
ggplot(trip_tib_txn, aes(drug, efficiency, color = TTseq)) + geom_quasirandom(size = 3) +
  theme_bw(base_size = 16) + 
  geom_vline(xintercept=seq(1.5, length(unique(trip_tib_txn$drug))-0.5, 1),
    			   lwd=2, colour="grey90")  +
    stat_summary(fun.y=median, fun.ymin = median,
          fun.ymax = median, geom="crossbar", width = 0.5,
          color='red')
```


```{r scatter NHEJ with and without ssODN 2, fig.height=6, fig.width=10}
txn_drugs <- unique(trip_tib_2000$drug[grepl("PFI|LLY|DMSO_", trip_tib_2000$drug)]) # get drug names with triptolide, BRB and DMSO
trip_tib_txn <- trip_tib_2000 %>% 
  filter(drug %in% txn_drugs, 
         cell_line == "RSTP2_clone5", 
         plasmid == "LBR2", 
         ssODN == "-", 
         siRNA == "-", 
         time %in% c(72)) %>% 
  mutate(drug = factor(.$drug,
                       levels=c("DMSO_low",  "DMSO_mid", "DMSO_high",
                                "LLY-507_500nM","LLY-507_1uM",  "LLY-507_2uM",
                                "PFI-2S_2nM", "PFI-2S_5nM", "PFI-2S_10nM", 
                                "PFI-2R_2nM","PFI-2R_5nM", "PFI-2R_10nM")))

```

```{r perturb2}
pathway_distribution <- trip_tib_txn %>% 
  mutate(wt = 1-efficiency) %>% 
  gather(NHEJ, MMEJ, SSTR, other_indels, wt,
         key = "pathway", value = 'ratios') %>%
  arrange(dplyr::desc(ratios))
barcode_order <- pathway_distribution %>%
  filter(pathway == "wt", drug == "DMSO_low")  %>%
  arrange(dplyr::desc(ratios)) %>%
  dplyr::select(barcode)  %>% unique()

pathway_distribution$barcode <- factor(pathway_distribution$barcode, 
                                       levels=barcode_order$barcode)
pathway_distribution$pathway <- factor(pathway_distribution$pathway, 
                                       levels=c("SSTR", "MMEJ", "other_indels",  "NHEJ", "wt"))
pathway_distribution$drug <- factor(pathway_distribution$drug, 
                                    levels=c("DMSO_low",  "DMSO_mid", "DMSO_high",
                                             "LLY-507_500nM","LLY-507_1uM",  "LLY-507_2uM",
                                             "PFI-2S_2nM", "PFI-2S_5nM", "PFI-2S_10nM", 
                                             "PFI-2R_2nM","PFI-2R_5nM", "PFI-2R_10nM"))

colores <- c("other_indels" = "black", "NHEJ" = "#E1251B", "MMEJ" = "#223DA0", "SSTR" = "#007A4C", "wt" = "#808184")


ggplot(pathway_distribution,
       aes(barcode, ratios, width=1)) +
  geom_bar(stat = "identity", position = "fill", aes(fill = pathway)) +
  theme_bw(base_size = 16) +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  ylab("Pathway Contribution") +
  scale_fill_manual("legend", values = colores) +
  ggtitle("Pathway contribution per barcode") +
  facet_grid(time ~ drug)
```

```{r perturb3}
ggplot(trip_tib_txn, aes(drug, MMEJ_MMEJNHEJ, color = H3K4me1)) + 
  geom_quasirandom(size = 3) +
  theme_bw(base_size = 16) + 
  geom_vline(xintercept=seq(1.5, length(unique(trip_tib_txn$drug))-0.5, 1),
    			   lwd=2, colour="grey90") +
    stat_summary(fun.y=median, fun.ymin = median,
          fun.ymax = median, geom="crossbar", width = 0.5,
          color='red')
ggplot(trip_tib_txn, aes(drug, efficiency, color = H3K4me1)) + 
  geom_quasirandom(size = 3) +
  theme_bw(base_size = 16) + 
  geom_vline(xintercept=seq(1.5, length(unique(trip_tib_txn$drug))-0.5, 1),
    			   lwd=2, colour="grey90")  +
    stat_summary(fun.y=median, fun.ymin = median,
          fun.ymax = median, geom="crossbar", width = 0.5,
          color='red')
```

# Session Info
```{r session info}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```
